// waltid-sd-jwt/kotlinx-serialization-kotlinx-serialization-core.js
(function (root, factory) {
  if (typeof define === 'function' && define.amd)
    define(['exports', './kotlin-kotlin-stdlib.js'], factory);
  else if (typeof exports === 'object')
    factory(module.exports, require('./kotlin-kotlin-stdlib.js'));
  else {
    if (typeof this['kotlin-kotlin-stdlib'] === 'undefined') {
      throw new Error("Error loading module 'kotlinx-serialization-kotlinx-serialization-core'. Its dependency 'kotlin-kotlin-stdlib' was not found. Please, check whether 'kotlin-kotlin-stdlib' is loaded prior to 'kotlinx-serialization-kotlinx-serialization-core'.");
    }
    root['kotlinx-serialization-kotlinx-serialization-core'] = factory(typeof this['kotlinx-serialization-kotlinx-serialization-core'] === 'undefined' ? {} : this['kotlinx-serialization-kotlinx-serialization-core'], this['kotlin-kotlin-stdlib']);
  }
}(this, function (_, kotlin_kotlin) {
  'use strict';
  //region block: imports
  var imul = Math.imul;
  var protoOf = kotlin_kotlin.$_$.s7;
  var interfaceMeta = kotlin_kotlin.$_$.b7;
  var setMetadataFor = kotlin_kotlin.$_$.t7;
  var VOID = kotlin_kotlin.$_$.c;
  var IllegalArgumentException_init_$Init$ = kotlin_kotlin.$_$.y;
  var objectCreate = kotlin_kotlin.$_$.q7;
  var captureStack = kotlin_kotlin.$_$.n6;
  var IllegalArgumentException_init_$Init$_0 = kotlin_kotlin.$_$.a1;
  var IllegalArgumentException_init_$Init$_1 = kotlin_kotlin.$_$.c1;
  var IllegalArgumentException = kotlin_kotlin.$_$.l9;
  var classMeta = kotlin_kotlin.$_$.t6;
  var THROW_CCE = kotlin_kotlin.$_$.p9;
  var isInterface = kotlin_kotlin.$_$.i7;
  var emptyList = kotlin_kotlin.$_$.m4;
  var isBlank = kotlin_kotlin.$_$.n8;
  var toString = kotlin_kotlin.$_$.x7;
  var IllegalArgumentException_init_$Create$ = kotlin_kotlin.$_$.b1;
  var equals = kotlin_kotlin.$_$.v6;
  var toList = kotlin_kotlin.$_$.o5;
  var ArrayList_init_$Create$ = kotlin_kotlin.$_$.h;
  var HashSet_init_$Create$ = kotlin_kotlin.$_$.m;
  var Unit_instance = kotlin_kotlin.$_$.o3;
  var toHashSet = kotlin_kotlin.$_$.m5;
  var copyToArray = kotlin_kotlin.$_$.l4;
  var toBooleanArray = kotlin_kotlin.$_$.l5;
  var withIndex = kotlin_kotlin.$_$.s5;
  var collectionSizeOrDefault = kotlin_kotlin.$_$.x3;
  var ArrayList_init_$Create$_0 = kotlin_kotlin.$_$.g;
  var to = kotlin_kotlin.$_$.oa;
  var toMap = kotlin_kotlin.$_$.p5;
  var lazy = kotlin_kotlin.$_$.ja;
  var contentEquals = kotlin_kotlin.$_$.y3;
  var until = kotlin_kotlin.$_$.e8;
  var joinToString = kotlin_kotlin.$_$.t4;
  var KProperty1 = kotlin_kotlin.$_$.g8;
  var getPropertyCallableRef = kotlin_kotlin.$_$.y6;
  var objectMeta = kotlin_kotlin.$_$.r7;
  var getKClassFromExpression = kotlin_kotlin.$_$.a;
  var ensureNotNull = kotlin_kotlin.$_$.ea;
  var getStringHashCode = kotlin_kotlin.$_$.z6;
  var Long = kotlin_kotlin.$_$.m9;
  var Char = kotlin_kotlin.$_$.f9;
  var Companion_getInstance = kotlin_kotlin.$_$.g3;
  var Duration = kotlin_kotlin.$_$.e9;
  var toIntOrNull = kotlin_kotlin.$_$.w8;
  var hashCode = kotlin_kotlin.$_$.a7;
  var IllegalStateException_init_$Create$ = kotlin_kotlin.$_$.d1;
  var ArrayList = kotlin_kotlin.$_$.p3;
  var ArrayList_init_$Create$_1 = kotlin_kotlin.$_$.i;
  var List = kotlin_kotlin.$_$.s3;
  var LinkedHashMap_init_$Create$ = kotlin_kotlin.$_$.o;
  var LinkedHashMap = kotlin_kotlin.$_$.r3;
  var LinkedHashMap_init_$Create$_0 = kotlin_kotlin.$_$.p;
  var Map = kotlin_kotlin.$_$.t3;
  var step = kotlin_kotlin.$_$.d8;
  var MutableMap = kotlin_kotlin.$_$.u3;
  var getValue = kotlin_kotlin.$_$.q4;
  var longArray = kotlin_kotlin.$_$.l7;
  var Companion_getInstance_0 = kotlin_kotlin.$_$.i3;
  var get_lastIndex = kotlin_kotlin.$_$.v4;
  var countTrailingZeroBits = kotlin_kotlin.$_$.ca;
  var fillArrayVal = kotlin_kotlin.$_$.w6;
  var LazyThreadSafetyMode_PUBLICATION_getInstance = kotlin_kotlin.$_$.d;
  var lazy_0 = kotlin_kotlin.$_$.ia;
  var HashSet_init_$Create$_0 = kotlin_kotlin.$_$.l;
  var HashMap_init_$Create$ = kotlin_kotlin.$_$.k;
  var booleanArray = kotlin_kotlin.$_$.m6;
  var emptyMap = kotlin_kotlin.$_$.n4;
  var contentHashCode = kotlin_kotlin.$_$.b4;
  var Companion_getInstance_1 = kotlin_kotlin.$_$.h3;
  var isCharArray = kotlin_kotlin.$_$.e7;
  var charArray = kotlin_kotlin.$_$.p6;
  var DoubleCompanionObject_instance = kotlin_kotlin.$_$.a3;
  var isDoubleArray = kotlin_kotlin.$_$.f7;
  var FloatCompanionObject_instance = kotlin_kotlin.$_$.b3;
  var isFloatArray = kotlin_kotlin.$_$.g7;
  var isLongArray = kotlin_kotlin.$_$.j7;
  var Companion_getInstance_2 = kotlin_kotlin.$_$.m3;
  var ULongArray = kotlin_kotlin.$_$.v9;
  var _ULongArray___init__impl__twm1l3 = kotlin_kotlin.$_$.k2;
  var _ULong___init__impl__c78o9k = kotlin_kotlin.$_$.i2;
  var IntCompanionObject_instance = kotlin_kotlin.$_$.c3;
  var isIntArray = kotlin_kotlin.$_$.h7;
  var Companion_getInstance_3 = kotlin_kotlin.$_$.l3;
  var UIntArray = kotlin_kotlin.$_$.t9;
  var _UIntArray___init__impl__ghjpc6 = kotlin_kotlin.$_$.d2;
  var _UInt___init__impl__l7qpdl = kotlin_kotlin.$_$.b2;
  var ShortCompanionObject_instance = kotlin_kotlin.$_$.d3;
  var isShortArray = kotlin_kotlin.$_$.k7;
  var Companion_getInstance_4 = kotlin_kotlin.$_$.n3;
  var UShortArray = kotlin_kotlin.$_$.x9;
  var _UShortArray___init__impl__9b26ef = kotlin_kotlin.$_$.r2;
  var _UShort___init__impl__jigrne = kotlin_kotlin.$_$.p2;
  var ByteCompanionObject_instance = kotlin_kotlin.$_$.z2;
  var isByteArray = kotlin_kotlin.$_$.d7;
  var Companion_getInstance_5 = kotlin_kotlin.$_$.k3;
  var UByteArray = kotlin_kotlin.$_$.r9;
  var _UByteArray___init__impl__ip4y9n = kotlin_kotlin.$_$.x1;
  var _UByte___init__impl__g9hnc4 = kotlin_kotlin.$_$.u1;
  var BooleanCompanionObject_instance = kotlin_kotlin.$_$.y2;
  var isBooleanArray = kotlin_kotlin.$_$.c7;
  var coerceAtLeast = kotlin_kotlin.$_$.a8;
  var copyOf = kotlin_kotlin.$_$.f4;
  var copyOf_0 = kotlin_kotlin.$_$.h4;
  var copyOf_1 = kotlin_kotlin.$_$.i4;
  var copyOf_2 = kotlin_kotlin.$_$.d4;
  var _ULongArray___get_size__impl__ju6dtr = kotlin_kotlin.$_$.n2;
  var _ULongArray___get_storage__impl__28e64j = kotlin_kotlin.$_$.o2;
  var _ULongArray___init__impl__twm1l3_0 = kotlin_kotlin.$_$.l2;
  var ULongArray__set_impl_z19mvh = kotlin_kotlin.$_$.m2;
  var copyOf_3 = kotlin_kotlin.$_$.k4;
  var _UIntArray___get_size__impl__r6l8ci = kotlin_kotlin.$_$.g2;
  var _UIntArray___get_storage__impl__92a0v0 = kotlin_kotlin.$_$.h2;
  var _UIntArray___init__impl__ghjpc6_0 = kotlin_kotlin.$_$.e2;
  var UIntArray__set_impl_7f2zu2 = kotlin_kotlin.$_$.f2;
  var copyOf_4 = kotlin_kotlin.$_$.c4;
  var _UShortArray___get_size__impl__jqto1b = kotlin_kotlin.$_$.u2;
  var _UShortArray___get_storage__impl__t2jpv5 = kotlin_kotlin.$_$.v2;
  var _UShortArray___init__impl__9b26ef_0 = kotlin_kotlin.$_$.s2;
  var UShortArray__set_impl_6d8whp = kotlin_kotlin.$_$.t2;
  var copyOf_5 = kotlin_kotlin.$_$.g4;
  var _UByteArray___get_size__impl__h6pkdv = kotlin_kotlin.$_$.z1;
  var _UByteArray___get_storage__impl__d4kctt = kotlin_kotlin.$_$.a2;
  var _UByteArray___init__impl__ip4y9n_0 = kotlin_kotlin.$_$.w1;
  var UByteArray__set_impl_jvcicn = kotlin_kotlin.$_$.y1;
  var copyOf_6 = kotlin_kotlin.$_$.e4;
  var trimIndent = kotlin_kotlin.$_$.c9;
  var equals_0 = kotlin_kotlin.$_$.l8;
  var charSequenceLength = kotlin_kotlin.$_$.r6;
  var charSequenceGet = kotlin_kotlin.$_$.q6;
  var toString_0 = kotlin_kotlin.$_$.q1;
  var titlecase = kotlin_kotlin.$_$.u8;
  var isLowerCase = kotlin_kotlin.$_$.o8;
  var PrimitiveClasses_getInstance = kotlin_kotlin.$_$.f3;
  var StringCompanionObject_instance = kotlin_kotlin.$_$.e3;
  var getKClass = kotlin_kotlin.$_$.b;
  var ULong = kotlin_kotlin.$_$.w9;
  var UInt = kotlin_kotlin.$_$.u9;
  var UShort = kotlin_kotlin.$_$.y9;
  var UByte = kotlin_kotlin.$_$.s9;
  var Unit = kotlin_kotlin.$_$.z9;
  var mapOf = kotlin_kotlin.$_$.c5;
  var lastOrNull = kotlin_kotlin.$_$.x4;
  var get_lastIndex_0 = kotlin_kotlin.$_$.w4;
  var get_indices = kotlin_kotlin.$_$.s4;
  var IndexOutOfBoundsException_init_$Create$ = kotlin_kotlin.$_$.f1;
  var get_indices_0 = kotlin_kotlin.$_$.r4;
  //endregion
  //region block: pre-declaration
  setMetadataFor(DeserializationStrategy, 'DeserializationStrategy', interfaceMeta);
  setMetadataFor(KSerializer, 'KSerializer', interfaceMeta, VOID, [DeserializationStrategy]);
  setMetadataFor(SerializationException, 'SerializationException', classMeta, IllegalArgumentException, VOID, SerializationException_init_$Create$);
  setMetadataFor(UnknownFieldException, 'UnknownFieldException', classMeta, SerializationException);
  setMetadataFor(MissingFieldException, 'MissingFieldException', classMeta, SerializationException);
  function get_isNullable() {
    return false;
  }
  function get_isInline() {
    return false;
  }
  function get_annotations() {
    return emptyList();
  }
  setMetadataFor(SerialDescriptor, 'SerialDescriptor', interfaceMeta);
  setMetadataFor(ContextDescriptor, 'ContextDescriptor', classMeta, VOID, [SerialDescriptor]);
  setMetadataFor(elementDescriptors$1$1, VOID, classMeta);
  setMetadataFor(_no_name_provided__qut3iv, VOID, classMeta);
  setMetadataFor(elementNames$1$1, VOID, classMeta);
  setMetadataFor(_no_name_provided__qut3iv_0, VOID, classMeta);
  setMetadataFor(ClassSerialDescriptorBuilder, 'ClassSerialDescriptorBuilder', classMeta);
  setMetadataFor(CachedNames, 'CachedNames', interfaceMeta);
  setMetadataFor(SerialDescriptorImpl, 'SerialDescriptorImpl', classMeta, VOID, [SerialDescriptor, CachedNames]);
  setMetadataFor(SerialKind, 'SerialKind', classMeta);
  setMetadataFor(ENUM, 'ENUM', objectMeta, SerialKind);
  setMetadataFor(CONTEXTUAL, 'CONTEXTUAL', objectMeta, SerialKind);
  setMetadataFor(PrimitiveKind, 'PrimitiveKind', classMeta, SerialKind);
  setMetadataFor(BOOLEAN, 'BOOLEAN', objectMeta, PrimitiveKind);
  setMetadataFor(BYTE, 'BYTE', objectMeta, PrimitiveKind);
  setMetadataFor(CHAR, 'CHAR', objectMeta, PrimitiveKind);
  setMetadataFor(SHORT, 'SHORT', objectMeta, PrimitiveKind);
  setMetadataFor(INT, 'INT', objectMeta, PrimitiveKind);
  setMetadataFor(LONG, 'LONG', objectMeta, PrimitiveKind);
  setMetadataFor(FLOAT, 'FLOAT', objectMeta, PrimitiveKind);
  setMetadataFor(DOUBLE, 'DOUBLE', objectMeta, PrimitiveKind);
  setMetadataFor(STRING, 'STRING', objectMeta, PrimitiveKind);
  setMetadataFor(StructureKind, 'StructureKind', classMeta, SerialKind);
  setMetadataFor(CLASS, 'CLASS', objectMeta, StructureKind);
  setMetadataFor(LIST, 'LIST', objectMeta, StructureKind);
  setMetadataFor(MAP, 'MAP', objectMeta, StructureKind);
  setMetadataFor(OBJECT, 'OBJECT', objectMeta, StructureKind);
  setMetadataFor(PolymorphicKind, 'PolymorphicKind', classMeta, SerialKind);
  setMetadataFor(SEALED, 'SEALED', objectMeta, PolymorphicKind);
  function decodeSerializableValue(deserializer) {
    return deserializer.gi(this);
  }
  setMetadataFor(Decoder, 'Decoder', interfaceMeta);
  function decodeSequentially() {
    return false;
  }
  function decodeCollectionSize(descriptor) {
    return -1;
  }
  function decodeSerializableElement$default(descriptor, index, deserializer, previousValue, $super) {
    previousValue = previousValue === VOID ? null : previousValue;
    return $super === VOID ? this.cl(descriptor, index, deserializer, previousValue) : $super.cl.call(this, descriptor, index, deserializer, previousValue);
  }
  setMetadataFor(CompositeDecoder, 'CompositeDecoder', interfaceMeta);
  setMetadataFor(AbstractDecoder, 'AbstractDecoder', classMeta, VOID, [Decoder, CompositeDecoder]);
  setMetadataFor(Companion, 'Companion', objectMeta);
  setMetadataFor(AbstractPolymorphicSerializer, 'AbstractPolymorphicSerializer', classMeta, VOID, [KSerializer]);
  setMetadataFor(NothingSerializer_0, 'NothingSerializer', objectMeta, VOID, [KSerializer]);
  setMetadataFor(DurationSerializer, 'DurationSerializer', objectMeta, VOID, [KSerializer]);
  setMetadataFor(ListLikeDescriptor, 'ListLikeDescriptor', classMeta, VOID, [SerialDescriptor]);
  setMetadataFor(ArrayListClassDesc, 'ArrayListClassDesc', classMeta, ListLikeDescriptor);
  setMetadataFor(MapLikeDescriptor, 'MapLikeDescriptor', classMeta, VOID, [SerialDescriptor]);
  setMetadataFor(LinkedHashMapClassDesc, 'LinkedHashMapClassDesc', classMeta, MapLikeDescriptor);
  setMetadataFor(PrimitiveArrayDescriptor, 'PrimitiveArrayDescriptor', classMeta, ListLikeDescriptor);
  setMetadataFor(AbstractCollectionSerializer, 'AbstractCollectionSerializer', classMeta, VOID, [KSerializer]);
  setMetadataFor(CollectionLikeSerializer, 'CollectionLikeSerializer', classMeta, AbstractCollectionSerializer);
  setMetadataFor(CollectionSerializer, 'CollectionSerializer', classMeta, CollectionLikeSerializer);
  setMetadataFor(ArrayListSerializer, 'ArrayListSerializer', classMeta, CollectionSerializer);
  setMetadataFor(MapLikeSerializer, 'MapLikeSerializer', classMeta, AbstractCollectionSerializer);
  setMetadataFor(LinkedHashMapSerializer, 'LinkedHashMapSerializer', classMeta, MapLikeSerializer);
  setMetadataFor(PrimitiveArraySerializer, 'PrimitiveArraySerializer', classMeta, CollectionLikeSerializer);
  setMetadataFor(PrimitiveArrayBuilder, 'PrimitiveArrayBuilder', classMeta);
  setMetadataFor(Companion_0, 'Companion', objectMeta);
  setMetadataFor(ElementMarker, 'ElementMarker', classMeta);
  setMetadataFor(EnumSerializer, 'EnumSerializer', classMeta, VOID, [KSerializer]);
  setMetadataFor(PluginGeneratedSerialDescriptor, 'PluginGeneratedSerialDescriptor', classMeta, VOID, [SerialDescriptor, CachedNames]);
  setMetadataFor(EnumDescriptor, 'EnumDescriptor', classMeta, PluginGeneratedSerialDescriptor);
  setMetadataFor(InlineClassDescriptor, 'InlineClassDescriptor', classMeta, PluginGeneratedSerialDescriptor);
  function typeParametersSerializers() {
    return get_EMPTY_SERIALIZER_ARRAY();
  }
  setMetadataFor(GeneratedSerializer, 'GeneratedSerializer', interfaceMeta, VOID, [KSerializer]);
  setMetadataFor(InlinePrimitiveDescriptor$1, VOID, classMeta, VOID, [GeneratedSerializer]);
  setMetadataFor(NothingSerialDescriptor, 'NothingSerialDescriptor', objectMeta, VOID, [SerialDescriptor]);
  setMetadataFor(NullableSerializer, 'NullableSerializer', classMeta, VOID, [KSerializer]);
  setMetadataFor(SerialDescriptorForNullable, 'SerialDescriptorForNullable', classMeta, VOID, [SerialDescriptor, CachedNames]);
  setMetadataFor(ObjectSerializer, 'ObjectSerializer', classMeta, VOID, [KSerializer]);
  setMetadataFor(CharArraySerializer_0, 'CharArraySerializer', objectMeta, PrimitiveArraySerializer, [KSerializer, PrimitiveArraySerializer]);
  setMetadataFor(DoubleArraySerializer_0, 'DoubleArraySerializer', objectMeta, PrimitiveArraySerializer, [KSerializer, PrimitiveArraySerializer]);
  setMetadataFor(FloatArraySerializer_0, 'FloatArraySerializer', objectMeta, PrimitiveArraySerializer, [KSerializer, PrimitiveArraySerializer]);
  setMetadataFor(LongArraySerializer_0, 'LongArraySerializer', objectMeta, PrimitiveArraySerializer, [KSerializer, PrimitiveArraySerializer]);
  setMetadataFor(ULongArraySerializer_0, 'ULongArraySerializer', objectMeta, PrimitiveArraySerializer, [KSerializer, PrimitiveArraySerializer]);
  setMetadataFor(IntArraySerializer_0, 'IntArraySerializer', objectMeta, PrimitiveArraySerializer, [KSerializer, PrimitiveArraySerializer]);
  setMetadataFor(UIntArraySerializer_0, 'UIntArraySerializer', objectMeta, PrimitiveArraySerializer, [KSerializer, PrimitiveArraySerializer]);
  setMetadataFor(ShortArraySerializer_0, 'ShortArraySerializer', objectMeta, PrimitiveArraySerializer, [KSerializer, PrimitiveArraySerializer]);
  setMetadataFor(UShortArraySerializer_0, 'UShortArraySerializer', objectMeta, PrimitiveArraySerializer, [KSerializer, PrimitiveArraySerializer]);
  setMetadataFor(ByteArraySerializer_0, 'ByteArraySerializer', objectMeta, PrimitiveArraySerializer, [KSerializer, PrimitiveArraySerializer]);
  setMetadataFor(UByteArraySerializer_0, 'UByteArraySerializer', objectMeta, PrimitiveArraySerializer, [KSerializer, PrimitiveArraySerializer]);
  setMetadataFor(BooleanArraySerializer_0, 'BooleanArraySerializer', objectMeta, PrimitiveArraySerializer, [KSerializer, PrimitiveArraySerializer]);
  setMetadataFor(CharArrayBuilder, 'CharArrayBuilder', classMeta, PrimitiveArrayBuilder);
  setMetadataFor(DoubleArrayBuilder, 'DoubleArrayBuilder', classMeta, PrimitiveArrayBuilder);
  setMetadataFor(FloatArrayBuilder, 'FloatArrayBuilder', classMeta, PrimitiveArrayBuilder);
  setMetadataFor(LongArrayBuilder, 'LongArrayBuilder', classMeta, PrimitiveArrayBuilder);
  setMetadataFor(ULongArrayBuilder, 'ULongArrayBuilder', classMeta, PrimitiveArrayBuilder);
  setMetadataFor(IntArrayBuilder, 'IntArrayBuilder', classMeta, PrimitiveArrayBuilder);
  setMetadataFor(UIntArrayBuilder, 'UIntArrayBuilder', classMeta, PrimitiveArrayBuilder);
  setMetadataFor(ShortArrayBuilder, 'ShortArrayBuilder', classMeta, PrimitiveArrayBuilder);
  setMetadataFor(UShortArrayBuilder, 'UShortArrayBuilder', classMeta, PrimitiveArrayBuilder);
  setMetadataFor(ByteArrayBuilder, 'ByteArrayBuilder', classMeta, PrimitiveArrayBuilder);
  setMetadataFor(UByteArrayBuilder, 'UByteArrayBuilder', classMeta, PrimitiveArrayBuilder);
  setMetadataFor(BooleanArrayBuilder, 'BooleanArrayBuilder', classMeta, PrimitiveArrayBuilder);
  setMetadataFor(StringSerializer, 'StringSerializer', objectMeta, VOID, [KSerializer]);
  setMetadataFor(CharSerializer, 'CharSerializer', objectMeta, VOID, [KSerializer]);
  setMetadataFor(DoubleSerializer, 'DoubleSerializer', objectMeta, VOID, [KSerializer]);
  setMetadataFor(FloatSerializer, 'FloatSerializer', objectMeta, VOID, [KSerializer]);
  setMetadataFor(LongSerializer, 'LongSerializer', objectMeta, VOID, [KSerializer]);
  setMetadataFor(IntSerializer, 'IntSerializer', objectMeta, VOID, [KSerializer]);
  setMetadataFor(ShortSerializer, 'ShortSerializer', objectMeta, VOID, [KSerializer]);
  setMetadataFor(ByteSerializer, 'ByteSerializer', objectMeta, VOID, [KSerializer]);
  setMetadataFor(BooleanSerializer, 'BooleanSerializer', objectMeta, VOID, [KSerializer]);
  setMetadataFor(UnitSerializer, 'UnitSerializer', objectMeta, VOID, [KSerializer]);
  setMetadataFor(PrimitiveSerialDescriptor_0, 'PrimitiveSerialDescriptor', classMeta, VOID, [SerialDescriptor]);
  setMetadataFor(TaggedDecoder, 'TaggedDecoder', classMeta, VOID, [Decoder, CompositeDecoder]);
  setMetadataFor(NamedValueDecoder, 'NamedValueDecoder', classMeta, TaggedDecoder);
  setMetadataFor(ULongSerializer, 'ULongSerializer', objectMeta, VOID, [KSerializer]);
  setMetadataFor(UIntSerializer, 'UIntSerializer', objectMeta, VOID, [KSerializer]);
  setMetadataFor(UShortSerializer, 'UShortSerializer', objectMeta, VOID, [KSerializer]);
  setMetadataFor(UByteSerializer, 'UByteSerializer', objectMeta, VOID, [KSerializer]);
  setMetadataFor(SerializersModule, 'SerializersModule', classMeta);
  setMetadataFor(SerialModuleImpl, 'SerialModuleImpl', classMeta, SerializersModule);
  setMetadataFor(SerializableWith, 'SerializableWith', classMeta, VOID, VOID, VOID, 0);
  //endregion
  function KSerializer() {
  }
  function DeserializationStrategy() {
  }
  function SerializationException_init_$Init$($this) {
    IllegalArgumentException_init_$Init$($this);
    SerializationException.call($this);
    return $this;
  }
  function SerializationException_init_$Create$() {
    var tmp = SerializationException_init_$Init$(objectCreate(protoOf(SerializationException)));
    captureStack(tmp, SerializationException_init_$Create$);
    return tmp;
  }
  function SerializationException_init_$Init$_0(message, $this) {
    IllegalArgumentException_init_$Init$_0(message, $this);
    SerializationException.call($this);
    return $this;
  }
  function SerializationException_init_$Create$_0(message) {
    var tmp = SerializationException_init_$Init$_0(message, objectCreate(protoOf(SerializationException)));
    captureStack(tmp, SerializationException_init_$Create$_0);
    return tmp;
  }
  function SerializationException_init_$Init$_1(message, cause, $this) {
    IllegalArgumentException_init_$Init$_1(message, cause, $this);
    SerializationException.call($this);
    return $this;
  }
  function SerializationException() {
    captureStack(this, SerializationException);
  }
  function UnknownFieldException_init_$Init$(index, $this) {
    UnknownFieldException.call($this, 'An unknown field for index ' + index);
    return $this;
  }
  function UnknownFieldException_init_$Create$(index) {
    var tmp = UnknownFieldException_init_$Init$(index, objectCreate(protoOf(UnknownFieldException)));
    captureStack(tmp, UnknownFieldException_init_$Create$);
    return tmp;
  }
  function UnknownFieldException(message) {
    SerializationException_init_$Init$_0(message, this);
    captureStack(this, UnknownFieldException);
  }
  function MissingFieldException_init_$Init$(missingFields, serialName, $this) {
    MissingFieldException.call($this, missingFields, missingFields.u() === 1 ? "Field '" + missingFields.b1(0) + "' is required for type with serial name '" + serialName + "', but it was missing" : 'Fields ' + missingFields + " are required for type with serial name '" + serialName + "', but they were missing", null);
    return $this;
  }
  function MissingFieldException_init_$Create$(missingFields, serialName) {
    var tmp = MissingFieldException_init_$Init$(missingFields, serialName, objectCreate(protoOf(MissingFieldException)));
    captureStack(tmp, MissingFieldException_init_$Create$);
    return tmp;
  }
  function MissingFieldException(missingFields, message, cause) {
    SerializationException_init_$Init$_1(message, cause, this);
    captureStack(this, MissingFieldException);
    this.hi_1 = missingFields;
  }
  function get_nullable(_this__u8e3s4) {
    var tmp;
    if (_this__u8e3s4.fi().ii()) {
      tmp = isInterface(_this__u8e3s4, KSerializer) ? _this__u8e3s4 : THROW_CCE();
    } else {
      tmp = new NullableSerializer(_this__u8e3s4);
    }
    return tmp;
  }
  function serializer(_this__u8e3s4) {
    return StringSerializer_getInstance();
  }
  function serializer_0(_this__u8e3s4) {
    return CharSerializer_getInstance();
  }
  function CharArraySerializer() {
    return CharArraySerializer_getInstance();
  }
  function serializer_1(_this__u8e3s4) {
    return DoubleSerializer_getInstance();
  }
  function DoubleArraySerializer() {
    return DoubleArraySerializer_getInstance();
  }
  function serializer_2(_this__u8e3s4) {
    return FloatSerializer_getInstance();
  }
  function FloatArraySerializer() {
    return FloatArraySerializer_getInstance();
  }
  function serializer_3(_this__u8e3s4) {
    return LongSerializer_getInstance();
  }
  function LongArraySerializer() {
    return LongArraySerializer_getInstance();
  }
  function serializer_4(_this__u8e3s4) {
    return ULongSerializer_getInstance();
  }
  function ULongArraySerializer() {
    return ULongArraySerializer_getInstance();
  }
  function serializer_5(_this__u8e3s4) {
    return IntSerializer_getInstance();
  }
  function IntArraySerializer() {
    return IntArraySerializer_getInstance();
  }
  function serializer_6(_this__u8e3s4) {
    return UIntSerializer_getInstance();
  }
  function UIntArraySerializer() {
    return UIntArraySerializer_getInstance();
  }
  function serializer_7(_this__u8e3s4) {
    return ShortSerializer_getInstance();
  }
  function ShortArraySerializer() {
    return ShortArraySerializer_getInstance();
  }
  function serializer_8(_this__u8e3s4) {
    return UShortSerializer_getInstance();
  }
  function UShortArraySerializer() {
    return UShortArraySerializer_getInstance();
  }
  function serializer_9(_this__u8e3s4) {
    return ByteSerializer_getInstance();
  }
  function ByteArraySerializer() {
    return ByteArraySerializer_getInstance();
  }
  function serializer_10(_this__u8e3s4) {
    return UByteSerializer_getInstance();
  }
  function UByteArraySerializer() {
    return UByteArraySerializer_getInstance();
  }
  function serializer_11(_this__u8e3s4) {
    return BooleanSerializer_getInstance();
  }
  function BooleanArraySerializer() {
    return BooleanArraySerializer_getInstance();
  }
  function serializer_12(_this__u8e3s4) {
    return UnitSerializer_getInstance();
  }
  function NothingSerializer() {
    return NothingSerializer_getInstance();
  }
  function serializer_13(_this__u8e3s4) {
    return DurationSerializer_getInstance();
  }
  function MapSerializer(keySerializer, valueSerializer) {
    return new LinkedHashMapSerializer(keySerializer, valueSerializer);
  }
  function ListSerializer(elementSerializer) {
    return new ArrayListSerializer(elementSerializer);
  }
  function ContextDescriptor() {
  }
  function getContextualDescriptor(_this__u8e3s4, descriptor) {
    var tmp0_safe_receiver = get_capturedKClass(descriptor);
    var tmp;
    if (tmp0_safe_receiver == null) {
      tmp = null;
    } else {
      // Inline function 'kotlin.let' call
      // Inline function 'kotlin.contracts.contract' call
      // Inline function 'kotlinx.serialization.descriptors.getContextualDescriptor.<anonymous>' call
      var tmp0_safe_receiver_0 = _this__u8e3s4.ji(tmp0_safe_receiver);
      tmp = tmp0_safe_receiver_0 == null ? null : tmp0_safe_receiver_0.fi();
    }
    return tmp;
  }
  function get_capturedKClass(_this__u8e3s4) {
    var tmp;
    if (_this__u8e3s4 instanceof ContextDescriptor) {
      tmp = _this__u8e3s4.ni_1;
    } else {
      if (_this__u8e3s4 instanceof SerialDescriptorForNullable) {
        tmp = get_capturedKClass(_this__u8e3s4.ki_1);
      } else {
        tmp = null;
      }
    }
    return tmp;
  }
  function SerialDescriptor() {
  }
  function get_elementDescriptors(_this__u8e3s4) {
    // Inline function 'kotlin.collections.Iterable' call
    return new _no_name_provided__qut3iv(_this__u8e3s4);
  }
  function get_elementNames(_this__u8e3s4) {
    // Inline function 'kotlin.collections.Iterable' call
    return new _no_name_provided__qut3iv_0(_this__u8e3s4);
  }
  function elementDescriptors$1$1($this_elementDescriptors) {
    this.zi_1 = $this_elementDescriptors;
    this.yi_1 = $this_elementDescriptors.ri();
  }
  protoOf(elementDescriptors$1$1).s = function () {
    return this.yi_1 > 0;
  };
  protoOf(elementDescriptors$1$1).t = function () {
    var tmp = this.zi_1.ri();
    var tmp1 = this.yi_1;
    this.yi_1 = tmp1 - 1 | 0;
    return this.zi_1.wi(tmp - tmp1 | 0);
  };
  function _no_name_provided__qut3iv($this_elementDescriptors) {
    this.aj_1 = $this_elementDescriptors;
  }
  protoOf(_no_name_provided__qut3iv).r = function () {
    // Inline function 'kotlinx.serialization.descriptors.<get-elementDescriptors>.<anonymous>' call
    return new elementDescriptors$1$1(this.aj_1);
  };
  function elementNames$1$1($this_elementNames) {
    this.cj_1 = $this_elementNames;
    this.bj_1 = $this_elementNames.ri();
  }
  protoOf(elementNames$1$1).s = function () {
    return this.bj_1 > 0;
  };
  protoOf(elementNames$1$1).t = function () {
    var tmp = this.cj_1.ri();
    var tmp1 = this.bj_1;
    this.bj_1 = tmp1 - 1 | 0;
    return this.cj_1.ti(tmp - tmp1 | 0);
  };
  function _no_name_provided__qut3iv_0($this_elementNames) {
    this.dj_1 = $this_elementNames;
  }
  protoOf(_no_name_provided__qut3iv_0).r = function () {
    // Inline function 'kotlinx.serialization.descriptors.<get-elementNames>.<anonymous>' call
    return new elementNames$1$1(this.dj_1);
  };
  function buildSerialDescriptor(serialName, kind, typeParameters, builder) {
    var tmp;
    if (builder === VOID) {
      tmp = buildSerialDescriptor$lambda;
    } else {
      tmp = builder;
    }
    builder = tmp;
    // Inline function 'kotlin.require' call
    // Inline function 'kotlin.text.isNotBlank' call
    // Inline function 'kotlin.contracts.contract' call
    if (!!isBlank(serialName)) {
      // Inline function 'kotlinx.serialization.descriptors.buildSerialDescriptor.<anonymous>' call
      var message = 'Blank serial names are prohibited';
      throw IllegalArgumentException_init_$Create$(toString(message));
    }
    // Inline function 'kotlin.require' call
    // Inline function 'kotlin.contracts.contract' call
    if (!!equals(kind, CLASS_getInstance())) {
      // Inline function 'kotlinx.serialization.descriptors.buildSerialDescriptor.<anonymous>' call
      var message_0 = "For StructureKind.CLASS please use 'buildClassSerialDescriptor' instead";
      throw IllegalArgumentException_init_$Create$(toString(message_0));
    }
    var sdBuilder = new ClassSerialDescriptorBuilder(serialName);
    builder(sdBuilder);
    return new SerialDescriptorImpl(serialName, kind, sdBuilder.hj_1.u(), toList(typeParameters), sdBuilder);
  }
  function ClassSerialDescriptorBuilder(serialName) {
    this.ej_1 = serialName;
    this.fj_1 = false;
    this.gj_1 = emptyList();
    this.hj_1 = ArrayList_init_$Create$();
    this.ij_1 = HashSet_init_$Create$();
    this.jj_1 = ArrayList_init_$Create$();
    this.kj_1 = ArrayList_init_$Create$();
    this.lj_1 = ArrayList_init_$Create$();
  }
  protoOf(ClassSerialDescriptorBuilder).mj = function (elementName, descriptor, annotations, isOptional) {
    // Inline function 'kotlin.require' call
    // Inline function 'kotlin.contracts.contract' call
    if (!this.ij_1.o(elementName)) {
      // Inline function 'kotlinx.serialization.descriptors.ClassSerialDescriptorBuilder.element.<anonymous>' call
      var message = "Element with name '" + elementName + "' is already registered in " + this.ej_1;
      throw IllegalArgumentException_init_$Create$(toString(message));
    }
    // Inline function 'kotlin.collections.plusAssign' call
    this.hj_1.o(elementName);
    // Inline function 'kotlin.collections.plusAssign' call
    this.jj_1.o(descriptor);
    // Inline function 'kotlin.collections.plusAssign' call
    this.kj_1.o(annotations);
    // Inline function 'kotlin.collections.plusAssign' call
    this.lj_1.o(isOptional);
  };
  protoOf(ClassSerialDescriptorBuilder).nj = function (elementName, descriptor, annotations, isOptional, $super) {
    annotations = annotations === VOID ? emptyList() : annotations;
    isOptional = isOptional === VOID ? false : isOptional;
    var tmp;
    if ($super === VOID) {
      this.mj(elementName, descriptor, annotations, isOptional);
      tmp = Unit_instance;
    } else {
      tmp = $super.mj.call(this, elementName, descriptor, annotations, isOptional);
    }
    return tmp;
  };
  function _get__hashCode__tgwhef($this) {
    // Inline function 'kotlin.getValue' call
    var this_0 = $this.zj_1;
    _hashCode$factory();
    return this_0.v1();
  }
  function SerialDescriptorImpl$_hashCode$delegate$lambda(this$0) {
    return function () {
      return hashCodeImpl(this$0, this$0.yj_1);
    };
  }
  function SerialDescriptorImpl$toString$lambda(this$0) {
    return function (it) {
      return this$0.ti(it) + ': ' + this$0.wi(it).oi();
    };
  }
  function SerialDescriptorImpl(serialName, kind, elementsCount, typeParameters, builder) {
    this.oj_1 = serialName;
    this.pj_1 = kind;
    this.qj_1 = elementsCount;
    this.rj_1 = builder.gj_1;
    this.sj_1 = toHashSet(builder.hj_1);
    var tmp = this;
    // Inline function 'kotlin.collections.toTypedArray' call
    var this_0 = builder.hj_1;
    tmp.tj_1 = copyToArray(this_0);
    this.uj_1 = compactArray(builder.jj_1);
    var tmp_0 = this;
    // Inline function 'kotlin.collections.toTypedArray' call
    var this_1 = builder.kj_1;
    tmp_0.vj_1 = copyToArray(this_1);
    this.wj_1 = toBooleanArray(builder.lj_1);
    var tmp_1 = this;
    // Inline function 'kotlin.collections.map' call
    var this_2 = withIndex(this.tj_1);
    // Inline function 'kotlin.collections.mapTo' call
    var destination = ArrayList_init_$Create$_0(collectionSizeOrDefault(this_2, 10));
    var tmp0_iterator = this_2.r();
    while (tmp0_iterator.s()) {
      var item = tmp0_iterator.t();
      // Inline function 'kotlinx.serialization.descriptors.SerialDescriptorImpl.name2Index.<anonymous>' call
      var tmp$ret$2 = to(item.ed_1, item.dd_1);
      destination.o(tmp$ret$2);
    }
    tmp_1.xj_1 = toMap(destination);
    this.yj_1 = compactArray(typeParameters);
    var tmp_2 = this;
    tmp_2.zj_1 = lazy(SerialDescriptorImpl$_hashCode$delegate$lambda(this));
  }
  protoOf(SerialDescriptorImpl).oi = function () {
    return this.oj_1;
  };
  protoOf(SerialDescriptorImpl).pi = function () {
    return this.pj_1;
  };
  protoOf(SerialDescriptorImpl).ri = function () {
    return this.qj_1;
  };
  protoOf(SerialDescriptorImpl).si = function () {
    return this.rj_1;
  };
  protoOf(SerialDescriptorImpl).ak = function () {
    return this.sj_1;
  };
  protoOf(SerialDescriptorImpl).ti = function (index) {
    return getChecked(this.tj_1, index);
  };
  protoOf(SerialDescriptorImpl).ui = function (name) {
    var tmp0_elvis_lhs = this.xj_1.b2(name);
    var tmp;
    if (tmp0_elvis_lhs == null) {
      tmp = -3;
    } else {
      tmp = tmp0_elvis_lhs;
    }
    return tmp;
  };
  protoOf(SerialDescriptorImpl).vi = function (index) {
    return getChecked(this.vj_1, index);
  };
  protoOf(SerialDescriptorImpl).wi = function (index) {
    return getChecked(this.uj_1, index);
  };
  protoOf(SerialDescriptorImpl).xi = function (index) {
    return getChecked_0(this.wj_1, index);
  };
  protoOf(SerialDescriptorImpl).equals = function (other) {
    var tmp$ret$0;
    $l$block_5: {
      // Inline function 'kotlinx.serialization.internal.equalsImpl' call
      if (this === other) {
        tmp$ret$0 = true;
        break $l$block_5;
      }
      if (!(other instanceof SerialDescriptorImpl)) {
        tmp$ret$0 = false;
        break $l$block_5;
      }
      if (!(this.oi() === other.oi())) {
        tmp$ret$0 = false;
        break $l$block_5;
      }
      // Inline function 'kotlinx.serialization.descriptors.SerialDescriptorImpl.equals.<anonymous>' call
      if (!contentEquals(this.yj_1, other.yj_1)) {
        tmp$ret$0 = false;
        break $l$block_5;
      }
      if (!(this.ri() === other.ri())) {
        tmp$ret$0 = false;
        break $l$block_5;
      }
      var inductionVariable = 0;
      var last = this.ri();
      if (inductionVariable < last)
        do {
          var index = inductionVariable;
          inductionVariable = inductionVariable + 1 | 0;
          if (!(this.wi(index).oi() === other.wi(index).oi())) {
            tmp$ret$0 = false;
            break $l$block_5;
          }
          if (!equals(this.wi(index).pi(), other.wi(index).pi())) {
            tmp$ret$0 = false;
            break $l$block_5;
          }
        }
         while (inductionVariable < last);
      tmp$ret$0 = true;
    }
    return tmp$ret$0;
  };
  protoOf(SerialDescriptorImpl).hashCode = function () {
    return _get__hashCode__tgwhef(this);
  };
  protoOf(SerialDescriptorImpl).toString = function () {
    var tmp = until(0, this.qj_1);
    var tmp_0 = this.oj_1 + '(';
    return joinToString(tmp, ', ', tmp_0, ')', VOID, VOID, SerialDescriptorImpl$toString$lambda(this));
  };
  function PrimitiveSerialDescriptor(serialName, kind) {
    // Inline function 'kotlin.require' call
    // Inline function 'kotlin.text.isNotBlank' call
    // Inline function 'kotlin.contracts.contract' call
    if (!!isBlank(serialName)) {
      // Inline function 'kotlinx.serialization.descriptors.PrimitiveSerialDescriptor.<anonymous>' call
      var message = 'Blank serial names are prohibited';
      throw IllegalArgumentException_init_$Create$(toString(message));
    }
    return PrimitiveDescriptorSafe(serialName, kind);
  }
  function buildSerialDescriptor$lambda($this$null) {
    return Unit_instance;
  }
  function _hashCode$factory() {
    return getPropertyCallableRef('_hashCode', 1, KProperty1, function (receiver) {
      return _get__hashCode__tgwhef(receiver);
    }, null);
  }
  function ENUM() {
    ENUM_instance = this;
    SerialKind.call(this);
  }
  var ENUM_instance;
  function ENUM_getInstance() {
    if (ENUM_instance == null)
      new ENUM();
    return ENUM_instance;
  }
  function CONTEXTUAL() {
    CONTEXTUAL_instance = this;
    SerialKind.call(this);
  }
  var CONTEXTUAL_instance;
  function CONTEXTUAL_getInstance() {
    if (CONTEXTUAL_instance == null)
      new CONTEXTUAL();
    return CONTEXTUAL_instance;
  }
  function SerialKind() {
  }
  protoOf(SerialKind).toString = function () {
    return ensureNotNull(getKClassFromExpression(this).c6());
  };
  protoOf(SerialKind).hashCode = function () {
    return getStringHashCode(this.toString());
  };
  function BOOLEAN() {
    BOOLEAN_instance = this;
    PrimitiveKind.call(this);
  }
  var BOOLEAN_instance;
  function BOOLEAN_getInstance() {
    if (BOOLEAN_instance == null)
      new BOOLEAN();
    return BOOLEAN_instance;
  }
  function BYTE() {
    BYTE_instance = this;
    PrimitiveKind.call(this);
  }
  var BYTE_instance;
  function BYTE_getInstance() {
    if (BYTE_instance == null)
      new BYTE();
    return BYTE_instance;
  }
  function CHAR() {
    CHAR_instance = this;
    PrimitiveKind.call(this);
  }
  var CHAR_instance;
  function CHAR_getInstance() {
    if (CHAR_instance == null)
      new CHAR();
    return CHAR_instance;
  }
  function SHORT() {
    SHORT_instance = this;
    PrimitiveKind.call(this);
  }
  var SHORT_instance;
  function SHORT_getInstance() {
    if (SHORT_instance == null)
      new SHORT();
    return SHORT_instance;
  }
  function INT() {
    INT_instance = this;
    PrimitiveKind.call(this);
  }
  var INT_instance;
  function INT_getInstance() {
    if (INT_instance == null)
      new INT();
    return INT_instance;
  }
  function LONG() {
    LONG_instance = this;
    PrimitiveKind.call(this);
  }
  var LONG_instance;
  function LONG_getInstance() {
    if (LONG_instance == null)
      new LONG();
    return LONG_instance;
  }
  function FLOAT() {
    FLOAT_instance = this;
    PrimitiveKind.call(this);
  }
  var FLOAT_instance;
  function FLOAT_getInstance() {
    if (FLOAT_instance == null)
      new FLOAT();
    return FLOAT_instance;
  }
  function DOUBLE() {
    DOUBLE_instance = this;
    PrimitiveKind.call(this);
  }
  var DOUBLE_instance;
  function DOUBLE_getInstance() {
    if (DOUBLE_instance == null)
      new DOUBLE();
    return DOUBLE_instance;
  }
  function STRING() {
    STRING_instance = this;
    PrimitiveKind.call(this);
  }
  var STRING_instance;
  function STRING_getInstance() {
    if (STRING_instance == null)
      new STRING();
    return STRING_instance;
  }
  function PrimitiveKind() {
    SerialKind.call(this);
  }
  function CLASS() {
    CLASS_instance = this;
    StructureKind.call(this);
  }
  var CLASS_instance;
  function CLASS_getInstance() {
    if (CLASS_instance == null)
      new CLASS();
    return CLASS_instance;
  }
  function LIST() {
    LIST_instance = this;
    StructureKind.call(this);
  }
  var LIST_instance;
  function LIST_getInstance() {
    if (LIST_instance == null)
      new LIST();
    return LIST_instance;
  }
  function MAP() {
    MAP_instance = this;
    StructureKind.call(this);
  }
  var MAP_instance;
  function MAP_getInstance() {
    if (MAP_instance == null)
      new MAP();
    return MAP_instance;
  }
  function OBJECT() {
    OBJECT_instance = this;
    StructureKind.call(this);
  }
  var OBJECT_instance;
  function OBJECT_getInstance() {
    if (OBJECT_instance == null)
      new OBJECT();
    return OBJECT_instance;
  }
  function StructureKind() {
    SerialKind.call(this);
  }
  function SEALED() {
    SEALED_instance = this;
    PolymorphicKind.call(this);
  }
  var SEALED_instance;
  function SEALED_getInstance() {
    if (SEALED_instance == null)
      new SEALED();
    return SEALED_instance;
  }
  function PolymorphicKind() {
    SerialKind.call(this);
  }
  function AbstractDecoder() {
  }
  protoOf(AbstractDecoder).bk = function () {
    throw SerializationException_init_$Create$_0('' + getKClassFromExpression(this) + " can't retrieve untyped values");
  };
  protoOf(AbstractDecoder).ck = function () {
    return true;
  };
  protoOf(AbstractDecoder).dk = function () {
    return null;
  };
  protoOf(AbstractDecoder).ek = function () {
    var tmp = this.bk();
    return typeof tmp === 'boolean' ? tmp : THROW_CCE();
  };
  protoOf(AbstractDecoder).fk = function () {
    var tmp = this.bk();
    return typeof tmp === 'number' ? tmp : THROW_CCE();
  };
  protoOf(AbstractDecoder).gk = function () {
    var tmp = this.bk();
    return typeof tmp === 'number' ? tmp : THROW_CCE();
  };
  protoOf(AbstractDecoder).hk = function () {
    var tmp = this.bk();
    return typeof tmp === 'number' ? tmp : THROW_CCE();
  };
  protoOf(AbstractDecoder).ik = function () {
    var tmp = this.bk();
    return tmp instanceof Long ? tmp : THROW_CCE();
  };
  protoOf(AbstractDecoder).jk = function () {
    var tmp = this.bk();
    return typeof tmp === 'number' ? tmp : THROW_CCE();
  };
  protoOf(AbstractDecoder).kk = function () {
    var tmp = this.bk();
    return typeof tmp === 'number' ? tmp : THROW_CCE();
  };
  protoOf(AbstractDecoder).lk = function () {
    var tmp = this.bk();
    return tmp instanceof Char ? tmp.h9_1 : THROW_CCE();
  };
  protoOf(AbstractDecoder).mk = function () {
    var tmp = this.bk();
    return typeof tmp === 'string' ? tmp : THROW_CCE();
  };
  protoOf(AbstractDecoder).nk = function (enumDescriptor) {
    var tmp = this.bk();
    return typeof tmp === 'number' ? tmp : THROW_CCE();
  };
  protoOf(AbstractDecoder).ok = function (descriptor) {
    return this;
  };
  protoOf(AbstractDecoder).pk = function (deserializer, previousValue) {
    return this.qk(deserializer);
  };
  protoOf(AbstractDecoder).rk = function (descriptor) {
    return this;
  };
  protoOf(AbstractDecoder).sk = function (descriptor) {
  };
  protoOf(AbstractDecoder).tk = function (descriptor, index) {
    return this.ek();
  };
  protoOf(AbstractDecoder).uk = function (descriptor, index) {
    return this.fk();
  };
  protoOf(AbstractDecoder).vk = function (descriptor, index) {
    return this.gk();
  };
  protoOf(AbstractDecoder).wk = function (descriptor, index) {
    return this.hk();
  };
  protoOf(AbstractDecoder).xk = function (descriptor, index) {
    return this.ik();
  };
  protoOf(AbstractDecoder).yk = function (descriptor, index) {
    return this.jk();
  };
  protoOf(AbstractDecoder).zk = function (descriptor, index) {
    return this.kk();
  };
  protoOf(AbstractDecoder).al = function (descriptor, index) {
    return this.lk();
  };
  protoOf(AbstractDecoder).bl = function (descriptor, index) {
    return this.ok(descriptor.wi(index));
  };
  protoOf(AbstractDecoder).cl = function (descriptor, index, deserializer, previousValue) {
    return this.pk(deserializer, previousValue);
  };
  protoOf(AbstractDecoder).el = function (descriptor, index, deserializer, previousValue) {
    var isNullabilitySupported = deserializer.fi().ii();
    return (isNullabilitySupported ? true : this.ck()) ? this.pk(deserializer, previousValue) : this.dk();
  };
  function Decoder() {
  }
  function Companion() {
    this.jl_1 = -1;
    this.kl_1 = -3;
  }
  var Companion_instance;
  function Companion_getInstance_6() {
    return Companion_instance;
  }
  function CompositeDecoder() {
  }
  function AbstractPolymorphicSerializer() {
  }
  protoOf(AbstractPolymorphicSerializer).ml = function (decoder, klassName) {
    return decoder.fl().nl(this.ll(), klassName);
  };
  function NothingSerializer_0() {
    NothingSerializer_instance = this;
    this.ol_1 = NothingSerialDescriptor_getInstance();
  }
  protoOf(NothingSerializer_0).fi = function () {
    return this.ol_1;
  };
  protoOf(NothingSerializer_0).gi = function (decoder) {
    throw SerializationException_init_$Create$_0("'kotlin.Nothing' does not have instances");
  };
  var NothingSerializer_instance;
  function NothingSerializer_getInstance() {
    if (NothingSerializer_instance == null)
      new NothingSerializer_0();
    return NothingSerializer_instance;
  }
  function DurationSerializer() {
    DurationSerializer_instance = this;
    this.pl_1 = new PrimitiveSerialDescriptor_0('kotlin.time.Duration', STRING_getInstance());
  }
  protoOf(DurationSerializer).fi = function () {
    return this.pl_1;
  };
  protoOf(DurationSerializer).ql = function (decoder) {
    return Companion_getInstance().tf(decoder.mk());
  };
  protoOf(DurationSerializer).gi = function (decoder) {
    return new Duration(this.ql(decoder));
  };
  var DurationSerializer_instance;
  function DurationSerializer_getInstance() {
    if (DurationSerializer_instance == null)
      new DurationSerializer();
    return DurationSerializer_instance;
  }
  function CachedNames() {
  }
  function ArrayListClassDesc(elementDesc) {
    ListLikeDescriptor.call(this, elementDesc);
  }
  protoOf(ArrayListClassDesc).oi = function () {
    return 'kotlin.collections.ArrayList';
  };
  function LinkedHashMapClassDesc(keyDesc, valueDesc) {
    MapLikeDescriptor.call(this, 'kotlin.collections.LinkedHashMap', keyDesc, valueDesc);
  }
  function ListLikeDescriptor(elementDescriptor) {
    this.tl_1 = elementDescriptor;
    this.ul_1 = 1;
  }
  protoOf(ListLikeDescriptor).pi = function () {
    return LIST_getInstance();
  };
  protoOf(ListLikeDescriptor).ri = function () {
    return this.ul_1;
  };
  protoOf(ListLikeDescriptor).ti = function (index) {
    return index.toString();
  };
  protoOf(ListLikeDescriptor).ui = function (name) {
    var tmp0_elvis_lhs = toIntOrNull(name);
    var tmp;
    if (tmp0_elvis_lhs == null) {
      throw IllegalArgumentException_init_$Create$(name + ' is not a valid list index');
    } else {
      tmp = tmp0_elvis_lhs;
    }
    return tmp;
  };
  protoOf(ListLikeDescriptor).xi = function (index) {
    // Inline function 'kotlin.require' call
    // Inline function 'kotlin.contracts.contract' call
    if (!(index >= 0)) {
      // Inline function 'kotlinx.serialization.internal.ListLikeDescriptor.isElementOptional.<anonymous>' call
      var message = 'Illegal index ' + index + ', ' + this.oi() + ' expects only non-negative indices';
      throw IllegalArgumentException_init_$Create$(toString(message));
    }
    return false;
  };
  protoOf(ListLikeDescriptor).vi = function (index) {
    // Inline function 'kotlin.require' call
    // Inline function 'kotlin.contracts.contract' call
    if (!(index >= 0)) {
      // Inline function 'kotlinx.serialization.internal.ListLikeDescriptor.getElementAnnotations.<anonymous>' call
      var message = 'Illegal index ' + index + ', ' + this.oi() + ' expects only non-negative indices';
      throw IllegalArgumentException_init_$Create$(toString(message));
    }
    return emptyList();
  };
  protoOf(ListLikeDescriptor).wi = function (index) {
    // Inline function 'kotlin.require' call
    // Inline function 'kotlin.contracts.contract' call
    if (!(index >= 0)) {
      // Inline function 'kotlinx.serialization.internal.ListLikeDescriptor.getElementDescriptor.<anonymous>' call
      var message = 'Illegal index ' + index + ', ' + this.oi() + ' expects only non-negative indices';
      throw IllegalArgumentException_init_$Create$(toString(message));
    }
    return this.tl_1;
  };
  protoOf(ListLikeDescriptor).equals = function (other) {
    if (this === other)
      return true;
    if (!(other instanceof ListLikeDescriptor))
      return false;
    if (equals(this.tl_1, other.tl_1) ? this.oi() === other.oi() : false)
      return true;
    return false;
  };
  protoOf(ListLikeDescriptor).hashCode = function () {
    return imul(hashCode(this.tl_1), 31) + getStringHashCode(this.oi()) | 0;
  };
  protoOf(ListLikeDescriptor).toString = function () {
    return this.oi() + '(' + this.tl_1 + ')';
  };
  function MapLikeDescriptor(serialName, keyDescriptor, valueDescriptor) {
    this.vl_1 = serialName;
    this.wl_1 = keyDescriptor;
    this.xl_1 = valueDescriptor;
    this.yl_1 = 2;
  }
  protoOf(MapLikeDescriptor).oi = function () {
    return this.vl_1;
  };
  protoOf(MapLikeDescriptor).pi = function () {
    return MAP_getInstance();
  };
  protoOf(MapLikeDescriptor).ri = function () {
    return this.yl_1;
  };
  protoOf(MapLikeDescriptor).ti = function (index) {
    return index.toString();
  };
  protoOf(MapLikeDescriptor).ui = function (name) {
    var tmp0_elvis_lhs = toIntOrNull(name);
    var tmp;
    if (tmp0_elvis_lhs == null) {
      throw IllegalArgumentException_init_$Create$(name + ' is not a valid map index');
    } else {
      tmp = tmp0_elvis_lhs;
    }
    return tmp;
  };
  protoOf(MapLikeDescriptor).xi = function (index) {
    // Inline function 'kotlin.require' call
    // Inline function 'kotlin.contracts.contract' call
    if (!(index >= 0)) {
      // Inline function 'kotlinx.serialization.internal.MapLikeDescriptor.isElementOptional.<anonymous>' call
      var message = 'Illegal index ' + index + ', ' + this.oi() + ' expects only non-negative indices';
      throw IllegalArgumentException_init_$Create$(toString(message));
    }
    return false;
  };
  protoOf(MapLikeDescriptor).vi = function (index) {
    // Inline function 'kotlin.require' call
    // Inline function 'kotlin.contracts.contract' call
    if (!(index >= 0)) {
      // Inline function 'kotlinx.serialization.internal.MapLikeDescriptor.getElementAnnotations.<anonymous>' call
      var message = 'Illegal index ' + index + ', ' + this.oi() + ' expects only non-negative indices';
      throw IllegalArgumentException_init_$Create$(toString(message));
    }
    return emptyList();
  };
  protoOf(MapLikeDescriptor).wi = function (index) {
    // Inline function 'kotlin.require' call
    // Inline function 'kotlin.contracts.contract' call
    if (!(index >= 0)) {
      // Inline function 'kotlinx.serialization.internal.MapLikeDescriptor.getElementDescriptor.<anonymous>' call
      var message = 'Illegal index ' + index + ', ' + this.oi() + ' expects only non-negative indices';
      throw IllegalArgumentException_init_$Create$(toString(message));
    }
    var tmp;
    switch (index % 2 | 0) {
      case 0:
        tmp = this.wl_1;
        break;
      case 1:
        tmp = this.xl_1;
        break;
      default:
        var message_0 = 'Unreached';
        throw IllegalStateException_init_$Create$(toString(message_0));
    }
    return tmp;
  };
  protoOf(MapLikeDescriptor).equals = function (other) {
    if (this === other)
      return true;
    if (!(other instanceof MapLikeDescriptor))
      return false;
    if (!(this.oi() === other.oi()))
      return false;
    if (!equals(this.wl_1, other.wl_1))
      return false;
    if (!equals(this.xl_1, other.xl_1))
      return false;
    return true;
  };
  protoOf(MapLikeDescriptor).hashCode = function () {
    var result = getStringHashCode(this.oi());
    result = imul(31, result) + hashCode(this.wl_1) | 0;
    result = imul(31, result) + hashCode(this.xl_1) | 0;
    return result;
  };
  protoOf(MapLikeDescriptor).toString = function () {
    return this.oi() + '(' + this.wl_1 + ', ' + this.xl_1 + ')';
  };
  function PrimitiveArrayDescriptor(primitive) {
    ListLikeDescriptor.call(this, primitive);
    this.bm_1 = primitive.oi() + 'Array';
  }
  protoOf(PrimitiveArrayDescriptor).oi = function () {
    return this.bm_1;
  };
  function ArrayListSerializer(element) {
    CollectionSerializer.call(this, element);
    this.dm_1 = new ArrayListClassDesc(element.fi());
  }
  protoOf(ArrayListSerializer).fi = function () {
    return this.dm_1;
  };
  protoOf(ArrayListSerializer).em = function () {
    // Inline function 'kotlin.collections.arrayListOf' call
    return ArrayList_init_$Create$();
  };
  protoOf(ArrayListSerializer).fm = function (_this__u8e3s4) {
    return _this__u8e3s4.u();
  };
  protoOf(ArrayListSerializer).gm = function (_this__u8e3s4) {
    return this.fm(_this__u8e3s4 instanceof ArrayList ? _this__u8e3s4 : THROW_CCE());
  };
  protoOf(ArrayListSerializer).hm = function (_this__u8e3s4) {
    return _this__u8e3s4;
  };
  protoOf(ArrayListSerializer).im = function (_this__u8e3s4) {
    return this.hm(_this__u8e3s4 instanceof ArrayList ? _this__u8e3s4 : THROW_CCE());
  };
  protoOf(ArrayListSerializer).jm = function (_this__u8e3s4) {
    var tmp0_elvis_lhs = _this__u8e3s4 instanceof ArrayList ? _this__u8e3s4 : null;
    return tmp0_elvis_lhs == null ? ArrayList_init_$Create$_1(_this__u8e3s4) : tmp0_elvis_lhs;
  };
  protoOf(ArrayListSerializer).km = function (_this__u8e3s4) {
    return this.jm((!(_this__u8e3s4 == null) ? isInterface(_this__u8e3s4, List) : false) ? _this__u8e3s4 : THROW_CCE());
  };
  protoOf(ArrayListSerializer).lm = function (_this__u8e3s4, size) {
    return _this__u8e3s4.k2(size);
  };
  protoOf(ArrayListSerializer).mm = function (_this__u8e3s4, size) {
    return this.lm(_this__u8e3s4 instanceof ArrayList ? _this__u8e3s4 : THROW_CCE(), size);
  };
  protoOf(ArrayListSerializer).nm = function (_this__u8e3s4, index, element) {
    _this__u8e3s4.f1(index, element);
  };
  protoOf(ArrayListSerializer).om = function (_this__u8e3s4, index, element) {
    var tmp = _this__u8e3s4 instanceof ArrayList ? _this__u8e3s4 : THROW_CCE();
    return this.nm(tmp, index, (element == null ? true : !(element == null)) ? element : THROW_CCE());
  };
  function LinkedHashMapSerializer(kSerializer, vSerializer) {
    MapLikeSerializer.call(this, kSerializer, vSerializer);
    this.wm_1 = new LinkedHashMapClassDesc(kSerializer.fi(), vSerializer.fi());
  }
  protoOf(LinkedHashMapSerializer).fi = function () {
    return this.wm_1;
  };
  protoOf(LinkedHashMapSerializer).em = function () {
    return LinkedHashMap_init_$Create$();
  };
  protoOf(LinkedHashMapSerializer).xm = function (_this__u8e3s4) {
    return imul(_this__u8e3s4.u(), 2);
  };
  protoOf(LinkedHashMapSerializer).gm = function (_this__u8e3s4) {
    return this.xm(_this__u8e3s4 instanceof LinkedHashMap ? _this__u8e3s4 : THROW_CCE());
  };
  protoOf(LinkedHashMapSerializer).ym = function (_this__u8e3s4) {
    return _this__u8e3s4;
  };
  protoOf(LinkedHashMapSerializer).im = function (_this__u8e3s4) {
    return this.ym(_this__u8e3s4 instanceof LinkedHashMap ? _this__u8e3s4 : THROW_CCE());
  };
  protoOf(LinkedHashMapSerializer).zm = function (_this__u8e3s4) {
    var tmp0_elvis_lhs = _this__u8e3s4 instanceof LinkedHashMap ? _this__u8e3s4 : null;
    return tmp0_elvis_lhs == null ? LinkedHashMap_init_$Create$_0(_this__u8e3s4) : tmp0_elvis_lhs;
  };
  protoOf(LinkedHashMapSerializer).km = function (_this__u8e3s4) {
    return this.zm((!(_this__u8e3s4 == null) ? isInterface(_this__u8e3s4, Map) : false) ? _this__u8e3s4 : THROW_CCE());
  };
  protoOf(LinkedHashMapSerializer).an = function (_this__u8e3s4, size) {
  };
  protoOf(LinkedHashMapSerializer).mm = function (_this__u8e3s4, size) {
    return this.an(_this__u8e3s4 instanceof LinkedHashMap ? _this__u8e3s4 : THROW_CCE(), size);
  };
  function CollectionSerializer(element) {
    CollectionLikeSerializer.call(this, element);
  }
  function MapLikeSerializer(keySerializer, valueSerializer) {
    AbstractCollectionSerializer.call(this);
    this.bn_1 = keySerializer;
    this.cn_1 = valueSerializer;
  }
  protoOf(MapLikeSerializer).dn = function (decoder, builder, startIndex, size) {
    // Inline function 'kotlin.require' call
    // Inline function 'kotlin.contracts.contract' call
    if (!(size >= 0)) {
      // Inline function 'kotlinx.serialization.internal.MapLikeSerializer.readAll.<anonymous>' call
      var message = 'Size must be known in advance when using READ_ALL';
      throw IllegalArgumentException_init_$Create$(toString(message));
    }
    var progression = step(until(0, imul(size, 2)), 2);
    var inductionVariable = progression.a9_1;
    var last = progression.b9_1;
    var step_0 = progression.c9_1;
    if ((step_0 > 0 ? inductionVariable <= last : false) ? true : step_0 < 0 ? last <= inductionVariable : false)
      do {
        var index = inductionVariable;
        inductionVariable = inductionVariable + step_0 | 0;
        this.en(decoder, startIndex + index | 0, builder, false);
      }
       while (!(index === last));
  };
  protoOf(MapLikeSerializer).qm = function (decoder, builder, startIndex, size) {
    return this.dn(decoder, (!(builder == null) ? isInterface(builder, MutableMap) : false) ? builder : THROW_CCE(), startIndex, size);
  };
  protoOf(MapLikeSerializer).en = function (decoder, index, builder, checkIndex) {
    var key = decoder.dl(this.fi(), index, this.bn_1);
    var tmp;
    if (checkIndex) {
      // Inline function 'kotlin.also' call
      var this_0 = decoder.hl(this.fi());
      // Inline function 'kotlin.contracts.contract' call
      // Inline function 'kotlinx.serialization.internal.MapLikeSerializer.readElement.<anonymous>' call
      // Inline function 'kotlin.require' call
      // Inline function 'kotlin.contracts.contract' call
      if (!(this_0 === (index + 1 | 0))) {
        // Inline function 'kotlinx.serialization.internal.MapLikeSerializer.readElement.<anonymous>.<anonymous>' call
        var message = 'Value must follow key in a map, index for key: ' + index + ', returned index for value: ' + this_0;
        throw IllegalArgumentException_init_$Create$(toString(message));
      }
      tmp = this_0;
    } else {
      tmp = index + 1 | 0;
    }
    var vIndex = tmp;
    var tmp_0;
    var tmp_1;
    if (builder.y1(key)) {
      var tmp_2 = this.cn_1.fi().pi();
      tmp_1 = !(tmp_2 instanceof PrimitiveKind);
    } else {
      tmp_1 = false;
    }
    if (tmp_1) {
      tmp_0 = decoder.cl(this.fi(), vIndex, this.cn_1, getValue(builder, key));
    } else {
      tmp_0 = decoder.dl(this.fi(), vIndex, this.cn_1);
    }
    var value = tmp_0;
    // Inline function 'kotlin.collections.set' call
    builder.r1(key, value);
  };
  protoOf(MapLikeSerializer).rm = function (decoder, index, builder, checkIndex) {
    return this.en(decoder, index, (!(builder == null) ? isInterface(builder, MutableMap) : false) ? builder : THROW_CCE(), checkIndex);
  };
  function CollectionLikeSerializer(elementSerializer) {
    AbstractCollectionSerializer.call(this);
    this.pm_1 = elementSerializer;
  }
  protoOf(CollectionLikeSerializer).qm = function (decoder, builder, startIndex, size) {
    // Inline function 'kotlin.require' call
    // Inline function 'kotlin.contracts.contract' call
    if (!(size >= 0)) {
      // Inline function 'kotlinx.serialization.internal.CollectionLikeSerializer.readAll.<anonymous>' call
      var message = 'Size must be known in advance when using READ_ALL';
      throw IllegalArgumentException_init_$Create$(toString(message));
    }
    var inductionVariable = 0;
    if (inductionVariable < size)
      do {
        var index = inductionVariable;
        inductionVariable = inductionVariable + 1 | 0;
        this.rm(decoder, startIndex + index | 0, builder, false);
      }
       while (inductionVariable < size);
  };
  protoOf(CollectionLikeSerializer).rm = function (decoder, index, builder, checkIndex) {
    this.om(builder, index, decoder.dl(this.fi(), index, this.pm_1));
  };
  function readSize($this, decoder, builder) {
    var size = decoder.il($this.fi());
    $this.mm(builder, size);
    return size;
  }
  function AbstractCollectionSerializer() {
  }
  protoOf(AbstractCollectionSerializer).tm = function (decoder, previous) {
    var tmp1_elvis_lhs = previous == null ? null : this.km(previous);
    var builder = tmp1_elvis_lhs == null ? this.em() : tmp1_elvis_lhs;
    var startIndex = this.gm(builder);
    var compositeDecoder = decoder.rk(this.fi());
    if (compositeDecoder.gl()) {
      this.qm(compositeDecoder, builder, startIndex, readSize(this, compositeDecoder, builder));
    } else {
      $l$loop: while (true) {
        var index = compositeDecoder.hl(this.fi());
        if (index === -1)
          break $l$loop;
        this.sm(compositeDecoder, startIndex + index | 0, builder);
      }
    }
    compositeDecoder.sk(this.fi());
    return this.im(builder);
  };
  protoOf(AbstractCollectionSerializer).gi = function (decoder) {
    return this.tm(decoder, null);
  };
  protoOf(AbstractCollectionSerializer).sm = function (decoder, index, builder, checkIndex, $super) {
    checkIndex = checkIndex === VOID ? true : checkIndex;
    var tmp;
    if ($super === VOID) {
      this.rm(decoder, index, builder, checkIndex);
      tmp = Unit_instance;
    } else {
      tmp = $super.rm.call(this, decoder, index, builder, checkIndex);
    }
    return tmp;
  };
  function PrimitiveArraySerializer(primitiveSerializer) {
    CollectionLikeSerializer.call(this, primitiveSerializer);
    this.gn_1 = new PrimitiveArrayDescriptor(primitiveSerializer.fi());
  }
  protoOf(PrimitiveArraySerializer).fi = function () {
    return this.gn_1;
  };
  protoOf(PrimitiveArraySerializer).hn = function (_this__u8e3s4) {
    return _this__u8e3s4.in();
  };
  protoOf(PrimitiveArraySerializer).gm = function (_this__u8e3s4) {
    return this.hn(_this__u8e3s4 instanceof PrimitiveArrayBuilder ? _this__u8e3s4 : THROW_CCE());
  };
  protoOf(PrimitiveArraySerializer).jn = function (_this__u8e3s4) {
    return _this__u8e3s4.kn();
  };
  protoOf(PrimitiveArraySerializer).im = function (_this__u8e3s4) {
    return this.jn(_this__u8e3s4 instanceof PrimitiveArrayBuilder ? _this__u8e3s4 : THROW_CCE());
  };
  protoOf(PrimitiveArraySerializer).ln = function (_this__u8e3s4, size) {
    return _this__u8e3s4.k2(size);
  };
  protoOf(PrimitiveArraySerializer).mm = function (_this__u8e3s4, size) {
    return this.ln(_this__u8e3s4 instanceof PrimitiveArrayBuilder ? _this__u8e3s4 : THROW_CCE(), size);
  };
  protoOf(PrimitiveArraySerializer).mn = function (_this__u8e3s4, index, element) {
    var message = 'This method lead to boxing and must not be used, use Builder.append instead';
    throw IllegalStateException_init_$Create$(toString(message));
  };
  protoOf(PrimitiveArraySerializer).om = function (_this__u8e3s4, index, element) {
    var tmp = _this__u8e3s4 instanceof PrimitiveArrayBuilder ? _this__u8e3s4 : THROW_CCE();
    return this.mn(tmp, index, (element == null ? true : !(element == null)) ? element : THROW_CCE());
  };
  protoOf(PrimitiveArraySerializer).em = function () {
    return this.km(this.nn());
  };
  protoOf(PrimitiveArraySerializer).gi = function (decoder) {
    return this.tm(decoder, null);
  };
  function PrimitiveArrayBuilder() {
  }
  protoOf(PrimitiveArrayBuilder).pn = function (requiredCapacity, $super) {
    requiredCapacity = requiredCapacity === VOID ? this.in() + 1 | 0 : requiredCapacity;
    var tmp;
    if ($super === VOID) {
      this.k2(requiredCapacity);
      tmp = Unit_instance;
    } else {
      tmp = $super.k2.call(this, requiredCapacity);
    }
    return tmp;
  };
  function Companion_0() {
    Companion_instance_0 = this;
    this.qn_1 = longArray(0);
  }
  var Companion_instance_0;
  function Companion_getInstance_7() {
    if (Companion_instance_0 == null)
      new Companion_0();
    return Companion_instance_0;
  }
  function prepareHighMarksArray($this, elementsCount) {
    var slotsCount = (elementsCount - 1 | 0) >>> 6 | 0;
    Companion_getInstance_0();
    var elementsInLastSlot = elementsCount & (64 - 1 | 0);
    var highMarks = longArray(slotsCount);
    if (!(elementsInLastSlot === 0)) {
      highMarks[get_lastIndex(highMarks)] = (new Long(-1, -1)).ga(elementsCount);
    }
    return highMarks;
  }
  function markHigh($this, index) {
    var slot = (index >>> 6 | 0) - 1 | 0;
    Companion_getInstance_0();
    var offsetInSlot = index & (64 - 1 | 0);
    $this.un_1[slot] = $this.un_1[slot].ka((new Long(1, 0)).ga(offsetInSlot));
  }
  function nextUnmarkedHighIndex($this) {
    var inductionVariable = 0;
    var last = $this.un_1.length - 1 | 0;
    if (inductionVariable <= last)
      do {
        var slot = inductionVariable;
        inductionVariable = inductionVariable + 1 | 0;
        var tmp = slot + 1 | 0;
        Companion_getInstance_0();
        var slotOffset = imul(tmp, 64);
        var slotMarks = $this.un_1[slot];
        while (!slotMarks.equals(new Long(-1, -1))) {
          var indexInSlot = countTrailingZeroBits(slotMarks.fa());
          slotMarks = slotMarks.ka((new Long(1, 0)).ga(indexInSlot));
          var index = slotOffset + indexInSlot | 0;
          if ($this.sn_1($this.rn_1, index)) {
            $this.un_1[slot] = slotMarks;
            return index;
          }
        }
        $this.un_1[slot] = slotMarks;
      }
       while (inductionVariable <= last);
    return -1;
  }
  function ElementMarker(descriptor, readIfAbsent) {
    Companion_getInstance_7();
    this.rn_1 = descriptor;
    this.sn_1 = readIfAbsent;
    var elementsCount = this.rn_1.ri();
    Companion_getInstance_0();
    if (elementsCount <= 64) {
      var tmp = this;
      var tmp_0;
      Companion_getInstance_0();
      if (elementsCount === 64) {
        tmp_0 = new Long(0, 0);
      } else {
        tmp_0 = (new Long(-1, -1)).ga(elementsCount);
      }
      tmp.tn_1 = tmp_0;
      this.un_1 = Companion_getInstance_7().qn_1;
    } else {
      this.tn_1 = new Long(0, 0);
      this.un_1 = prepareHighMarksArray(this, elementsCount);
    }
  }
  protoOf(ElementMarker).vn = function (index) {
    Companion_getInstance_0();
    if (index < 64) {
      this.tn_1 = this.tn_1.ka((new Long(1, 0)).ga(index));
    } else {
      markHigh(this, index);
    }
  };
  protoOf(ElementMarker).wn = function () {
    var elementsCount = this.rn_1.ri();
    while (!this.tn_1.equals(new Long(-1, -1))) {
      var index = countTrailingZeroBits(this.tn_1.fa());
      this.tn_1 = this.tn_1.ka((new Long(1, 0)).ga(index));
      if (this.sn_1(this.rn_1, index)) {
        return index;
      }
    }
    Companion_getInstance_0();
    if (elementsCount > 64) {
      return nextUnmarkedHighIndex(this);
    }
    return -1;
  };
  function createSimpleEnumSerializer(serialName, values) {
    return new EnumSerializer(serialName, values);
  }
  function createUnmarkedDescriptor($this, serialName) {
    var d = new EnumDescriptor(serialName, $this.xn_1.length);
    // Inline function 'kotlin.collections.forEach' call
    var indexedObject = $this.xn_1;
    var inductionVariable = 0;
    var last = indexedObject.length;
    while (inductionVariable < last) {
      var element = indexedObject[inductionVariable];
      inductionVariable = inductionVariable + 1 | 0;
      // Inline function 'kotlinx.serialization.internal.EnumSerializer.createUnmarkedDescriptor.<anonymous>' call
      d.mo(element.s8_1);
    }
    return d;
  }
  function EnumSerializer$descriptor$delegate$lambda(this$0, $serialName) {
    return function () {
      var tmp0_elvis_lhs = this$0.yn_1;
      return tmp0_elvis_lhs == null ? createUnmarkedDescriptor(this$0, $serialName) : tmp0_elvis_lhs;
    };
  }
  function EnumSerializer(serialName, values) {
    this.xn_1 = values;
    this.yn_1 = null;
    var tmp = this;
    tmp.zn_1 = lazy(EnumSerializer$descriptor$delegate$lambda(this, serialName));
  }
  protoOf(EnumSerializer).fi = function () {
    // Inline function 'kotlin.getValue' call
    var this_0 = this.zn_1;
    descriptor$factory();
    return this_0.v1();
  };
  protoOf(EnumSerializer).gi = function (decoder) {
    var index = decoder.nk(this.fi());
    if (!(0 <= index ? index <= (this.xn_1.length - 1 | 0) : false)) {
      throw SerializationException_init_$Create$_0('' + index + ' is not among valid ' + this.fi().oi() + ' enum values, ' + ('values size is ' + this.xn_1.length));
    }
    return this.xn_1[index];
  };
  protoOf(EnumSerializer).toString = function () {
    return 'kotlinx.serialization.internal.EnumSerializer<' + this.fi().oi() + '>';
  };
  function _get_elementDescriptors__y23q9p($this) {
    // Inline function 'kotlin.getValue' call
    var this_0 = $this.ap_1;
    elementDescriptors$factory();
    return this_0.v1();
  }
  function EnumDescriptor$elementDescriptors$delegate$lambda($elementsCount, $name, this$0) {
    return function () {
      var tmp = 0;
      var tmp_0 = $elementsCount;
      // Inline function 'kotlin.arrayOfNulls' call
      var tmp_1 = fillArrayVal(Array(tmp_0), null);
      while (tmp < tmp_0) {
        var tmp_2 = tmp;
        tmp_1[tmp_2] = buildSerialDescriptor($name + '.' + this$0.ti(tmp_2), OBJECT_getInstance(), []);
        tmp = tmp + 1 | 0;
      }
      return tmp_1;
    };
  }
  function EnumDescriptor(name, elementsCount) {
    PluginGeneratedSerialDescriptor.call(this, name, VOID, elementsCount);
    this.zo_1 = ENUM_getInstance();
    var tmp = this;
    tmp.ap_1 = lazy(EnumDescriptor$elementDescriptors$delegate$lambda(elementsCount, name, this));
  }
  protoOf(EnumDescriptor).pi = function () {
    return this.zo_1;
  };
  protoOf(EnumDescriptor).wi = function (index) {
    return getChecked(_get_elementDescriptors__y23q9p(this), index);
  };
  protoOf(EnumDescriptor).equals = function (other) {
    if (this === other)
      return true;
    if (other == null)
      return false;
    if (!(!(other == null) ? isInterface(other, SerialDescriptor) : false))
      return false;
    if (!(other.pi() === ENUM_getInstance()))
      return false;
    if (!(this.oi() === other.oi()))
      return false;
    if (!equals(cachedSerialNames(this), cachedSerialNames(other)))
      return false;
    return true;
  };
  protoOf(EnumDescriptor).toString = function () {
    return joinToString(get_elementNames(this), ', ', this.oi() + '(', ')');
  };
  protoOf(EnumDescriptor).hashCode = function () {
    var result = getStringHashCode(this.oi());
    // Inline function 'kotlinx.serialization.internal.elementsHashCodeBy' call
    // Inline function 'kotlin.collections.fold' call
    var accumulator = 1;
    var tmp0_iterator = get_elementNames(this).r();
    while (tmp0_iterator.s()) {
      var element = tmp0_iterator.t();
      // Inline function 'kotlinx.serialization.internal.elementsHashCodeBy.<anonymous>' call
      var hash = accumulator;
      var tmp = imul(31, hash);
      // Inline function 'kotlin.hashCode' call
      // Inline function 'kotlinx.serialization.internal.EnumDescriptor.hashCode.<anonymous>' call
      var tmp1_elvis_lhs = element == null ? null : hashCode(element);
      accumulator = tmp + (tmp1_elvis_lhs == null ? 0 : tmp1_elvis_lhs) | 0;
    }
    var elementsHashCode = accumulator;
    result = imul(31, result) + elementsHashCode | 0;
    return result;
  };
  function descriptor$factory() {
    return getPropertyCallableRef('descriptor', 1, KProperty1, function (receiver) {
      return receiver.fi();
    }, null);
  }
  function elementDescriptors$factory() {
    return getPropertyCallableRef('elementDescriptors', 1, KProperty1, function (receiver) {
      return _get_elementDescriptors__y23q9p(receiver);
    }, null);
  }
  function InlinePrimitiveDescriptor(name, primitiveSerializer) {
    return new InlineClassDescriptor(name, new InlinePrimitiveDescriptor$1(primitiveSerializer));
  }
  function InlineClassDescriptor(name, generatedSerializer) {
    PluginGeneratedSerialDescriptor.call(this, name, generatedSerializer, 1);
    this.pp_1 = true;
  }
  protoOf(InlineClassDescriptor).qi = function () {
    return this.pp_1;
  };
  protoOf(InlineClassDescriptor).hashCode = function () {
    return imul(protoOf(PluginGeneratedSerialDescriptor).hashCode.call(this), 31);
  };
  protoOf(InlineClassDescriptor).equals = function (other) {
    var tmp$ret$0;
    $l$block_5: {
      // Inline function 'kotlinx.serialization.internal.equalsImpl' call
      if (this === other) {
        tmp$ret$0 = true;
        break $l$block_5;
      }
      if (!(other instanceof InlineClassDescriptor)) {
        tmp$ret$0 = false;
        break $l$block_5;
      }
      if (!(this.oi() === other.oi())) {
        tmp$ret$0 = false;
        break $l$block_5;
      }
      // Inline function 'kotlinx.serialization.internal.InlineClassDescriptor.equals.<anonymous>' call
      if (!(other.pp_1 ? contentEquals(this.bp(), other.bp()) : false)) {
        tmp$ret$0 = false;
        break $l$block_5;
      }
      if (!(this.ri() === other.ri())) {
        tmp$ret$0 = false;
        break $l$block_5;
      }
      var inductionVariable = 0;
      var last = this.ri();
      if (inductionVariable < last)
        do {
          var index = inductionVariable;
          inductionVariable = inductionVariable + 1 | 0;
          if (!(this.wi(index).oi() === other.wi(index).oi())) {
            tmp$ret$0 = false;
            break $l$block_5;
          }
          if (!equals(this.wi(index).pi(), other.wi(index).pi())) {
            tmp$ret$0 = false;
            break $l$block_5;
          }
        }
         while (inductionVariable < last);
      tmp$ret$0 = true;
    }
    return tmp$ret$0;
  };
  function InlinePrimitiveDescriptor$1($primitiveSerializer) {
    this.qp_1 = $primitiveSerializer;
  }
  protoOf(InlinePrimitiveDescriptor$1).rp = function () {
    // Inline function 'kotlin.arrayOf' call
    // Inline function 'kotlin.js.unsafeCast' call
    // Inline function 'kotlin.js.asDynamic' call
    return [this.qp_1];
  };
  protoOf(InlinePrimitiveDescriptor$1).fi = function () {
    var message = 'unsupported';
    throw IllegalStateException_init_$Create$(toString(message));
  };
  protoOf(InlinePrimitiveDescriptor$1).gi = function (decoder) {
    // Inline function 'kotlin.error' call
    var message = 'unsupported';
    throw IllegalStateException_init_$Create$(toString(message));
  };
  function jsonCachedSerialNames(_this__u8e3s4) {
    return cachedSerialNames(_this__u8e3s4);
  }
  function error($this) {
    throw IllegalStateException_init_$Create$('Descriptor for type `kotlin.Nothing` does not have elements');
  }
  function NothingSerialDescriptor() {
    NothingSerialDescriptor_instance = this;
    this.tp_1 = OBJECT_getInstance();
    this.up_1 = 'kotlin.Nothing';
  }
  protoOf(NothingSerialDescriptor).pi = function () {
    return this.tp_1;
  };
  protoOf(NothingSerialDescriptor).oi = function () {
    return this.up_1;
  };
  protoOf(NothingSerialDescriptor).ri = function () {
    return 0;
  };
  protoOf(NothingSerialDescriptor).ti = function (index) {
    error(this);
  };
  protoOf(NothingSerialDescriptor).ui = function (name) {
    error(this);
  };
  protoOf(NothingSerialDescriptor).xi = function (index) {
    error(this);
  };
  protoOf(NothingSerialDescriptor).wi = function (index) {
    error(this);
  };
  protoOf(NothingSerialDescriptor).vi = function (index) {
    error(this);
  };
  protoOf(NothingSerialDescriptor).toString = function () {
    return 'NothingSerialDescriptor';
  };
  protoOf(NothingSerialDescriptor).equals = function (other) {
    return this === other;
  };
  protoOf(NothingSerialDescriptor).hashCode = function () {
    return getStringHashCode(this.up_1) + imul(31, this.tp_1.hashCode()) | 0;
  };
  var NothingSerialDescriptor_instance;
  function NothingSerialDescriptor_getInstance() {
    if (NothingSerialDescriptor_instance == null)
      new NothingSerialDescriptor();
    return NothingSerialDescriptor_instance;
  }
  function NullableSerializer(serializer) {
    this.vp_1 = serializer;
    this.wp_1 = new SerialDescriptorForNullable(this.vp_1.fi());
  }
  protoOf(NullableSerializer).fi = function () {
    return this.wp_1;
  };
  protoOf(NullableSerializer).gi = function (decoder) {
    return decoder.ck() ? decoder.qk(this.vp_1) : decoder.dk();
  };
  protoOf(NullableSerializer).equals = function (other) {
    if (this === other)
      return true;
    if (other == null ? true : !getKClassFromExpression(this).equals(getKClassFromExpression(other)))
      return false;
    if (!(other instanceof NullableSerializer))
      THROW_CCE();
    if (!equals(this.vp_1, other.vp_1))
      return false;
    return true;
  };
  protoOf(NullableSerializer).hashCode = function () {
    return hashCode(this.vp_1);
  };
  function SerialDescriptorForNullable(original) {
    this.ki_1 = original;
    this.li_1 = this.ki_1.oi() + '?';
    this.mi_1 = cachedSerialNames(this.ki_1);
  }
  protoOf(SerialDescriptorForNullable).si = function () {
    return this.ki_1.si();
  };
  protoOf(SerialDescriptorForNullable).ri = function () {
    return this.ki_1.ri();
  };
  protoOf(SerialDescriptorForNullable).qi = function () {
    return this.ki_1.qi();
  };
  protoOf(SerialDescriptorForNullable).pi = function () {
    return this.ki_1.pi();
  };
  protoOf(SerialDescriptorForNullable).vi = function (index) {
    return this.ki_1.vi(index);
  };
  protoOf(SerialDescriptorForNullable).wi = function (index) {
    return this.ki_1.wi(index);
  };
  protoOf(SerialDescriptorForNullable).ui = function (name) {
    return this.ki_1.ui(name);
  };
  protoOf(SerialDescriptorForNullable).ti = function (index) {
    return this.ki_1.ti(index);
  };
  protoOf(SerialDescriptorForNullable).xi = function (index) {
    return this.ki_1.xi(index);
  };
  protoOf(SerialDescriptorForNullable).oi = function () {
    return this.li_1;
  };
  protoOf(SerialDescriptorForNullable).ak = function () {
    return this.mi_1;
  };
  protoOf(SerialDescriptorForNullable).ii = function () {
    return true;
  };
  protoOf(SerialDescriptorForNullable).equals = function (other) {
    if (this === other)
      return true;
    if (!(other instanceof SerialDescriptorForNullable))
      return false;
    if (!equals(this.ki_1, other.ki_1))
      return false;
    return true;
  };
  protoOf(SerialDescriptorForNullable).toString = function () {
    return '' + this.ki_1 + '?';
  };
  protoOf(SerialDescriptorForNullable).hashCode = function () {
    return imul(hashCode(this.ki_1), 31);
  };
  function ObjectSerializer$descriptor$delegate$lambda$lambda(this$0) {
    return function ($this$buildSerialDescriptor) {
      $this$buildSerialDescriptor.gj_1 = this$0.yp_1;
      return Unit_instance;
    };
  }
  function ObjectSerializer$descriptor$delegate$lambda($serialName, this$0) {
    return function () {
      var tmp = OBJECT_getInstance();
      return buildSerialDescriptor($serialName, tmp, [], ObjectSerializer$descriptor$delegate$lambda$lambda(this$0));
    };
  }
  function ObjectSerializer(serialName, objectInstance) {
    this.xp_1 = objectInstance;
    this.yp_1 = emptyList();
    var tmp = this;
    var tmp_0 = LazyThreadSafetyMode_PUBLICATION_getInstance();
    tmp.zp_1 = lazy_0(tmp_0, ObjectSerializer$descriptor$delegate$lambda(serialName, this));
  }
  protoOf(ObjectSerializer).fi = function () {
    // Inline function 'kotlin.getValue' call
    var this_0 = this.zp_1;
    descriptor$factory_0();
    return this_0.v1();
  };
  protoOf(ObjectSerializer).gi = function (decoder) {
    // Inline function 'kotlinx.serialization.encoding.decodeStructure' call
    var descriptor = this.fi();
    var composite = decoder.rk(descriptor);
    var tmp$ret$0;
    $l$block_0: {
      // Inline function 'kotlinx.serialization.internal.ObjectSerializer.deserialize.<anonymous>' call
      if (composite.gl()) {
        tmp$ret$0 = Unit_instance;
        break $l$block_0;
      }
      var index = composite.hl(this.fi());
      if (index === -1) {
        tmp$ret$0 = Unit_instance;
        break $l$block_0;
      } else {
        throw SerializationException_init_$Create$_0('Unexpected index ' + index);
      }
    }
    var result = tmp$ret$0;
    composite.sk(descriptor);
    return this.xp_1;
  };
  function descriptor$factory_0() {
    return getPropertyCallableRef('descriptor', 1, KProperty1, function (receiver) {
      return receiver.fi();
    }, null);
  }
  function get_EMPTY_DESCRIPTOR_ARRAY() {
    _init_properties_Platform_common_kt__3qzecs();
    return EMPTY_DESCRIPTOR_ARRAY;
  }
  var EMPTY_DESCRIPTOR_ARRAY;
  function cachedSerialNames(_this__u8e3s4) {
    _init_properties_Platform_common_kt__3qzecs();
    if (isInterface(_this__u8e3s4, CachedNames))
      return _this__u8e3s4.ak();
    var result = HashSet_init_$Create$_0(_this__u8e3s4.ri());
    var inductionVariable = 0;
    var last = _this__u8e3s4.ri();
    if (inductionVariable < last)
      do {
        var i = inductionVariable;
        inductionVariable = inductionVariable + 1 | 0;
        // Inline function 'kotlin.collections.plusAssign' call
        var element = _this__u8e3s4.ti(i);
        result.o(element);
      }
       while (inductionVariable < last);
    return result;
  }
  function compactArray(_this__u8e3s4) {
    _init_properties_Platform_common_kt__3qzecs();
    // Inline function 'kotlin.takeUnless' call
    // Inline function 'kotlin.contracts.contract' call
    var tmp;
    // Inline function 'kotlinx.serialization.internal.compactArray.<anonymous>' call
    // Inline function 'kotlin.collections.isNullOrEmpty' call
    // Inline function 'kotlin.contracts.contract' call
    if (!(_this__u8e3s4 == null ? true : _this__u8e3s4.x())) {
      tmp = _this__u8e3s4;
    } else {
      tmp = null;
    }
    var tmp0_safe_receiver = tmp;
    var tmp_0;
    if (tmp0_safe_receiver == null) {
      tmp_0 = null;
    } else {
      // Inline function 'kotlin.collections.toTypedArray' call
      tmp_0 = copyToArray(tmp0_safe_receiver);
    }
    var tmp1_elvis_lhs = tmp_0;
    return tmp1_elvis_lhs == null ? get_EMPTY_DESCRIPTOR_ARRAY() : tmp1_elvis_lhs;
  }
  var properties_initialized_Platform_common_kt_i7q4ty;
  function _init_properties_Platform_common_kt__3qzecs() {
    if (!properties_initialized_Platform_common_kt_i7q4ty) {
      properties_initialized_Platform_common_kt_i7q4ty = true;
      // Inline function 'kotlin.arrayOf' call
      // Inline function 'kotlin.js.unsafeCast' call
      // Inline function 'kotlin.js.asDynamic' call
      EMPTY_DESCRIPTOR_ARRAY = [];
    }
  }
  function throwMissingFieldException(seen, goldenMask, descriptor) {
    // Inline function 'kotlin.collections.mutableListOf' call
    var missingFields = ArrayList_init_$Create$();
    var missingFieldsBits = goldenMask & ~seen;
    var inductionVariable = 0;
    if (inductionVariable < 32)
      do {
        var i = inductionVariable;
        inductionVariable = inductionVariable + 1 | 0;
        if (!((missingFieldsBits & 1) === 0)) {
          // Inline function 'kotlin.collections.plusAssign' call
          var element = descriptor.ti(i);
          missingFields.o(element);
        }
        missingFieldsBits = missingFieldsBits >>> 1 | 0;
      }
       while (inductionVariable < 32);
    throw MissingFieldException_init_$Create$(missingFields, descriptor.oi());
  }
  function _get_childSerializers__7vnyfa($this) {
    // Inline function 'kotlin.getValue' call
    var this_0 = $this.jo_1;
    childSerializers$factory();
    return this_0.v1();
  }
  function _get__hashCode__tgwhef_0($this) {
    // Inline function 'kotlin.getValue' call
    var this_0 = $this.lo_1;
    _hashCode$factory_0();
    return this_0.v1();
  }
  function buildIndices($this) {
    var indices = HashMap_init_$Create$();
    var inductionVariable = 0;
    var last = $this.eo_1.length - 1 | 0;
    if (inductionVariable <= last)
      do {
        var i = inductionVariable;
        inductionVariable = inductionVariable + 1 | 0;
        // Inline function 'kotlin.collections.set' call
        var key = $this.eo_1[i];
        indices.r1(key, i);
      }
       while (inductionVariable <= last);
    return indices;
  }
  function PluginGeneratedSerialDescriptor$childSerializers$delegate$lambda(this$0) {
    return function () {
      var tmp0_safe_receiver = this$0.bo_1;
      var tmp1_elvis_lhs = tmp0_safe_receiver == null ? null : tmp0_safe_receiver.rp();
      return tmp1_elvis_lhs == null ? get_EMPTY_SERIALIZER_ARRAY() : tmp1_elvis_lhs;
    };
  }
  function PluginGeneratedSerialDescriptor$typeParameterDescriptors$delegate$lambda(this$0) {
    return function () {
      var tmp0_safe_receiver = this$0.bo_1;
      var tmp1_safe_receiver = tmp0_safe_receiver == null ? null : tmp0_safe_receiver.sp();
      var tmp;
      if (tmp1_safe_receiver == null) {
        tmp = null;
      } else {
        // Inline function 'kotlin.collections.map' call
        // Inline function 'kotlin.collections.mapTo' call
        var destination = ArrayList_init_$Create$_0(tmp1_safe_receiver.length);
        var inductionVariable = 0;
        var last = tmp1_safe_receiver.length;
        while (inductionVariable < last) {
          var item = tmp1_safe_receiver[inductionVariable];
          inductionVariable = inductionVariable + 1 | 0;
          // Inline function 'kotlinx.serialization.internal.PluginGeneratedSerialDescriptor.typeParameterDescriptors$delegate.<anonymous>.<anonymous>' call
          var tmp$ret$0 = item.fi();
          destination.o(tmp$ret$0);
        }
        tmp = destination;
      }
      return compactArray(tmp);
    };
  }
  function PluginGeneratedSerialDescriptor$_hashCode$delegate$lambda(this$0) {
    return function () {
      return hashCodeImpl(this$0, this$0.bp());
    };
  }
  function PluginGeneratedSerialDescriptor$toString$lambda(this$0) {
    return function (i) {
      return this$0.ti(i) + ': ' + this$0.wi(i).oi();
    };
  }
  function PluginGeneratedSerialDescriptor(serialName, generatedSerializer, elementsCount) {
    generatedSerializer = generatedSerializer === VOID ? null : generatedSerializer;
    this.ao_1 = serialName;
    this.bo_1 = generatedSerializer;
    this.co_1 = elementsCount;
    this.do_1 = -1;
    var tmp = this;
    var tmp_0 = 0;
    var tmp_1 = this.co_1;
    // Inline function 'kotlin.arrayOfNulls' call
    var tmp_2 = fillArrayVal(Array(tmp_1), null);
    while (tmp_0 < tmp_1) {
      tmp_2[tmp_0] = '[UNINITIALIZED]';
      tmp_0 = tmp_0 + 1 | 0;
    }
    tmp.eo_1 = tmp_2;
    var tmp_3 = this;
    // Inline function 'kotlin.arrayOfNulls' call
    var size = this.co_1;
    tmp_3.fo_1 = fillArrayVal(Array(size), null);
    this.go_1 = null;
    this.ho_1 = booleanArray(this.co_1);
    this.io_1 = emptyMap();
    var tmp_4 = this;
    var tmp_5 = LazyThreadSafetyMode_PUBLICATION_getInstance();
    tmp_4.jo_1 = lazy_0(tmp_5, PluginGeneratedSerialDescriptor$childSerializers$delegate$lambda(this));
    var tmp_6 = this;
    var tmp_7 = LazyThreadSafetyMode_PUBLICATION_getInstance();
    tmp_6.ko_1 = lazy_0(tmp_7, PluginGeneratedSerialDescriptor$typeParameterDescriptors$delegate$lambda(this));
    var tmp_8 = this;
    var tmp_9 = LazyThreadSafetyMode_PUBLICATION_getInstance();
    tmp_8.lo_1 = lazy_0(tmp_9, PluginGeneratedSerialDescriptor$_hashCode$delegate$lambda(this));
  }
  protoOf(PluginGeneratedSerialDescriptor).oi = function () {
    return this.ao_1;
  };
  protoOf(PluginGeneratedSerialDescriptor).ri = function () {
    return this.co_1;
  };
  protoOf(PluginGeneratedSerialDescriptor).pi = function () {
    return CLASS_getInstance();
  };
  protoOf(PluginGeneratedSerialDescriptor).si = function () {
    var tmp0_elvis_lhs = this.go_1;
    return tmp0_elvis_lhs == null ? emptyList() : tmp0_elvis_lhs;
  };
  protoOf(PluginGeneratedSerialDescriptor).ak = function () {
    return this.io_1.p1();
  };
  protoOf(PluginGeneratedSerialDescriptor).bp = function () {
    // Inline function 'kotlin.getValue' call
    var this_0 = this.ko_1;
    typeParameterDescriptors$factory();
    return this_0.v1();
  };
  protoOf(PluginGeneratedSerialDescriptor).cp = function (name, isOptional) {
    this.do_1 = this.do_1 + 1 | 0;
    this.eo_1[this.do_1] = name;
    this.ho_1[this.do_1] = isOptional;
    this.fo_1[this.do_1] = null;
    if (this.do_1 === (this.co_1 - 1 | 0)) {
      this.io_1 = buildIndices(this);
    }
  };
  protoOf(PluginGeneratedSerialDescriptor).mo = function (name, isOptional, $super) {
    isOptional = isOptional === VOID ? false : isOptional;
    var tmp;
    if ($super === VOID) {
      this.cp(name, isOptional);
      tmp = Unit_instance;
    } else {
      tmp = $super.cp.call(this, name, isOptional);
    }
    return tmp;
  };
  protoOf(PluginGeneratedSerialDescriptor).wi = function (index) {
    return getChecked(_get_childSerializers__7vnyfa(this), index).fi();
  };
  protoOf(PluginGeneratedSerialDescriptor).xi = function (index) {
    return getChecked_0(this.ho_1, index);
  };
  protoOf(PluginGeneratedSerialDescriptor).vi = function (index) {
    var tmp0_elvis_lhs = getChecked(this.fo_1, index);
    return tmp0_elvis_lhs == null ? emptyList() : tmp0_elvis_lhs;
  };
  protoOf(PluginGeneratedSerialDescriptor).ti = function (index) {
    return getChecked(this.eo_1, index);
  };
  protoOf(PluginGeneratedSerialDescriptor).ui = function (name) {
    var tmp0_elvis_lhs = this.io_1.b2(name);
    var tmp;
    if (tmp0_elvis_lhs == null) {
      tmp = -3;
    } else {
      tmp = tmp0_elvis_lhs;
    }
    return tmp;
  };
  protoOf(PluginGeneratedSerialDescriptor).equals = function (other) {
    var tmp$ret$0;
    $l$block_5: {
      // Inline function 'kotlinx.serialization.internal.equalsImpl' call
      if (this === other) {
        tmp$ret$0 = true;
        break $l$block_5;
      }
      if (!(other instanceof PluginGeneratedSerialDescriptor)) {
        tmp$ret$0 = false;
        break $l$block_5;
      }
      if (!(this.oi() === other.oi())) {
        tmp$ret$0 = false;
        break $l$block_5;
      }
      // Inline function 'kotlinx.serialization.internal.PluginGeneratedSerialDescriptor.equals.<anonymous>' call
      if (!contentEquals(this.bp(), other.bp())) {
        tmp$ret$0 = false;
        break $l$block_5;
      }
      if (!(this.ri() === other.ri())) {
        tmp$ret$0 = false;
        break $l$block_5;
      }
      var inductionVariable = 0;
      var last = this.ri();
      if (inductionVariable < last)
        do {
          var index = inductionVariable;
          inductionVariable = inductionVariable + 1 | 0;
          if (!(this.wi(index).oi() === other.wi(index).oi())) {
            tmp$ret$0 = false;
            break $l$block_5;
          }
          if (!equals(this.wi(index).pi(), other.wi(index).pi())) {
            tmp$ret$0 = false;
            break $l$block_5;
          }
        }
         while (inductionVariable < last);
      tmp$ret$0 = true;
    }
    return tmp$ret$0;
  };
  protoOf(PluginGeneratedSerialDescriptor).hashCode = function () {
    return _get__hashCode__tgwhef_0(this);
  };
  protoOf(PluginGeneratedSerialDescriptor).toString = function () {
    var tmp = until(0, this.co_1);
    var tmp_0 = this.oi() + '(';
    return joinToString(tmp, ', ', tmp_0, ')', VOID, VOID, PluginGeneratedSerialDescriptor$toString$lambda(this));
  };
  function hashCodeImpl(_this__u8e3s4, typeParams) {
    var result = getStringHashCode(_this__u8e3s4.oi());
    result = imul(31, result) + contentHashCode(typeParams) | 0;
    var elementDescriptors = get_elementDescriptors(_this__u8e3s4);
    // Inline function 'kotlinx.serialization.internal.elementsHashCodeBy' call
    // Inline function 'kotlin.collections.fold' call
    var accumulator = 1;
    var tmp0_iterator = elementDescriptors.r();
    while (tmp0_iterator.s()) {
      var element = tmp0_iterator.t();
      // Inline function 'kotlinx.serialization.internal.elementsHashCodeBy.<anonymous>' call
      var hash = accumulator;
      var tmp = imul(31, hash);
      // Inline function 'kotlin.hashCode' call
      // Inline function 'kotlinx.serialization.internal.hashCodeImpl.<anonymous>' call
      var tmp0_safe_receiver = element.oi();
      var tmp1_elvis_lhs = tmp0_safe_receiver == null ? null : hashCode(tmp0_safe_receiver);
      accumulator = tmp + (tmp1_elvis_lhs == null ? 0 : tmp1_elvis_lhs) | 0;
    }
    var namesHash = accumulator;
    // Inline function 'kotlinx.serialization.internal.elementsHashCodeBy' call
    // Inline function 'kotlin.collections.fold' call
    var accumulator_0 = 1;
    var tmp0_iterator_0 = elementDescriptors.r();
    while (tmp0_iterator_0.s()) {
      var element_0 = tmp0_iterator_0.t();
      // Inline function 'kotlinx.serialization.internal.elementsHashCodeBy.<anonymous>' call
      var hash_0 = accumulator_0;
      var tmp_0 = imul(31, hash_0);
      // Inline function 'kotlin.hashCode' call
      // Inline function 'kotlinx.serialization.internal.hashCodeImpl.<anonymous>' call
      var tmp0_safe_receiver_0 = element_0.pi();
      var tmp1_elvis_lhs_0 = tmp0_safe_receiver_0 == null ? null : hashCode(tmp0_safe_receiver_0);
      accumulator_0 = tmp_0 + (tmp1_elvis_lhs_0 == null ? 0 : tmp1_elvis_lhs_0) | 0;
    }
    var kindHash = accumulator_0;
    result = imul(31, result) + namesHash | 0;
    result = imul(31, result) + kindHash | 0;
    return result;
  }
  function childSerializers$factory() {
    return getPropertyCallableRef('childSerializers', 1, KProperty1, function (receiver) {
      return _get_childSerializers__7vnyfa(receiver);
    }, null);
  }
  function typeParameterDescriptors$factory() {
    return getPropertyCallableRef('typeParameterDescriptors', 1, KProperty1, function (receiver) {
      return receiver.bp();
    }, null);
  }
  function _hashCode$factory_0() {
    return getPropertyCallableRef('_hashCode', 1, KProperty1, function (receiver) {
      return _get__hashCode__tgwhef_0(receiver);
    }, null);
  }
  function get_EMPTY_SERIALIZER_ARRAY() {
    _init_properties_PluginHelperInterfaces_kt__xgvzfp();
    return EMPTY_SERIALIZER_ARRAY;
  }
  var EMPTY_SERIALIZER_ARRAY;
  function GeneratedSerializer() {
  }
  var properties_initialized_PluginHelperInterfaces_kt_ap8in1;
  function _init_properties_PluginHelperInterfaces_kt__xgvzfp() {
    if (!properties_initialized_PluginHelperInterfaces_kt_ap8in1) {
      properties_initialized_PluginHelperInterfaces_kt_ap8in1 = true;
      // Inline function 'kotlin.arrayOf' call
      // Inline function 'kotlin.js.unsafeCast' call
      // Inline function 'kotlin.js.asDynamic' call
      EMPTY_SERIALIZER_ARRAY = [];
    }
  }
  function CharArraySerializer_0() {
    CharArraySerializer_instance = this;
    PrimitiveArraySerializer.call(this, serializer_0(Companion_getInstance_1()));
  }
  protoOf(CharArraySerializer_0).cq = function (_this__u8e3s4) {
    return new CharArrayBuilder(_this__u8e3s4);
  };
  protoOf(CharArraySerializer_0).km = function (_this__u8e3s4) {
    return this.cq((!(_this__u8e3s4 == null) ? isCharArray(_this__u8e3s4) : false) ? _this__u8e3s4 : THROW_CCE());
  };
  protoOf(CharArraySerializer_0).nn = function () {
    return charArray(0);
  };
  protoOf(CharArraySerializer_0).dq = function (decoder, index, builder, checkIndex) {
    builder.gq(decoder.al(this.gn_1, index));
  };
  protoOf(CharArraySerializer_0).rm = function (decoder, index, builder, checkIndex) {
    return this.dq(decoder, index, builder instanceof CharArrayBuilder ? builder : THROW_CCE(), checkIndex);
  };
  protoOf(CharArraySerializer_0).on = function (decoder, index, builder, checkIndex) {
    return this.dq(decoder, index, builder instanceof CharArrayBuilder ? builder : THROW_CCE(), checkIndex);
  };
  var CharArraySerializer_instance;
  function CharArraySerializer_getInstance() {
    if (CharArraySerializer_instance == null)
      new CharArraySerializer_0();
    return CharArraySerializer_instance;
  }
  function DoubleArraySerializer_0() {
    DoubleArraySerializer_instance = this;
    PrimitiveArraySerializer.call(this, serializer_1(DoubleCompanionObject_instance));
  }
  protoOf(DoubleArraySerializer_0).jq = function (_this__u8e3s4) {
    return new DoubleArrayBuilder(_this__u8e3s4);
  };
  protoOf(DoubleArraySerializer_0).km = function (_this__u8e3s4) {
    return this.jq((!(_this__u8e3s4 == null) ? isDoubleArray(_this__u8e3s4) : false) ? _this__u8e3s4 : THROW_CCE());
  };
  protoOf(DoubleArraySerializer_0).nn = function () {
    return new Float64Array(0);
  };
  protoOf(DoubleArraySerializer_0).kq = function (decoder, index, builder, checkIndex) {
    builder.nq(decoder.zk(this.gn_1, index));
  };
  protoOf(DoubleArraySerializer_0).rm = function (decoder, index, builder, checkIndex) {
    return this.kq(decoder, index, builder instanceof DoubleArrayBuilder ? builder : THROW_CCE(), checkIndex);
  };
  protoOf(DoubleArraySerializer_0).on = function (decoder, index, builder, checkIndex) {
    return this.kq(decoder, index, builder instanceof DoubleArrayBuilder ? builder : THROW_CCE(), checkIndex);
  };
  var DoubleArraySerializer_instance;
  function DoubleArraySerializer_getInstance() {
    if (DoubleArraySerializer_instance == null)
      new DoubleArraySerializer_0();
    return DoubleArraySerializer_instance;
  }
  function FloatArraySerializer_0() {
    FloatArraySerializer_instance = this;
    PrimitiveArraySerializer.call(this, serializer_2(FloatCompanionObject_instance));
  }
  protoOf(FloatArraySerializer_0).qq = function (_this__u8e3s4) {
    return new FloatArrayBuilder(_this__u8e3s4);
  };
  protoOf(FloatArraySerializer_0).km = function (_this__u8e3s4) {
    return this.qq((!(_this__u8e3s4 == null) ? isFloatArray(_this__u8e3s4) : false) ? _this__u8e3s4 : THROW_CCE());
  };
  protoOf(FloatArraySerializer_0).nn = function () {
    return new Float32Array(0);
  };
  protoOf(FloatArraySerializer_0).rq = function (decoder, index, builder, checkIndex) {
    builder.uq(decoder.yk(this.gn_1, index));
  };
  protoOf(FloatArraySerializer_0).rm = function (decoder, index, builder, checkIndex) {
    return this.rq(decoder, index, builder instanceof FloatArrayBuilder ? builder : THROW_CCE(), checkIndex);
  };
  protoOf(FloatArraySerializer_0).on = function (decoder, index, builder, checkIndex) {
    return this.rq(decoder, index, builder instanceof FloatArrayBuilder ? builder : THROW_CCE(), checkIndex);
  };
  var FloatArraySerializer_instance;
  function FloatArraySerializer_getInstance() {
    if (FloatArraySerializer_instance == null)
      new FloatArraySerializer_0();
    return FloatArraySerializer_instance;
  }
  function LongArraySerializer_0() {
    LongArraySerializer_instance = this;
    PrimitiveArraySerializer.call(this, serializer_3(Companion_getInstance_0()));
  }
  protoOf(LongArraySerializer_0).xq = function (_this__u8e3s4) {
    return new LongArrayBuilder(_this__u8e3s4);
  };
  protoOf(LongArraySerializer_0).km = function (_this__u8e3s4) {
    return this.xq((!(_this__u8e3s4 == null) ? isLongArray(_this__u8e3s4) : false) ? _this__u8e3s4 : THROW_CCE());
  };
  protoOf(LongArraySerializer_0).nn = function () {
    return longArray(0);
  };
  protoOf(LongArraySerializer_0).yq = function (decoder, index, builder, checkIndex) {
    builder.br(decoder.xk(this.gn_1, index));
  };
  protoOf(LongArraySerializer_0).rm = function (decoder, index, builder, checkIndex) {
    return this.yq(decoder, index, builder instanceof LongArrayBuilder ? builder : THROW_CCE(), checkIndex);
  };
  protoOf(LongArraySerializer_0).on = function (decoder, index, builder, checkIndex) {
    return this.yq(decoder, index, builder instanceof LongArrayBuilder ? builder : THROW_CCE(), checkIndex);
  };
  var LongArraySerializer_instance;
  function LongArraySerializer_getInstance() {
    if (LongArraySerializer_instance == null)
      new LongArraySerializer_0();
    return LongArraySerializer_instance;
  }
  function ULongArraySerializer_0() {
    ULongArraySerializer_instance = this;
    PrimitiveArraySerializer.call(this, serializer_4(Companion_getInstance_2()));
  }
  protoOf(ULongArraySerializer_0).er = function (_this__u8e3s4) {
    return new ULongArrayBuilder(_this__u8e3s4);
  };
  protoOf(ULongArraySerializer_0).km = function (_this__u8e3s4) {
    return this.er(_this__u8e3s4 instanceof ULongArray ? _this__u8e3s4.oh_1 : THROW_CCE());
  };
  protoOf(ULongArraySerializer_0).fr = function () {
    return _ULongArray___init__impl__twm1l3(0);
  };
  protoOf(ULongArraySerializer_0).nn = function () {
    return new ULongArray(this.fr());
  };
  protoOf(ULongArraySerializer_0).gr = function (decoder, index, builder, checkIndex) {
    // Inline function 'kotlin.toULong' call
    var this_0 = decoder.bl(this.gn_1, index).ik();
    var tmp$ret$0 = _ULong___init__impl__c78o9k(this_0);
    builder.jr(tmp$ret$0);
  };
  protoOf(ULongArraySerializer_0).rm = function (decoder, index, builder, checkIndex) {
    return this.gr(decoder, index, builder instanceof ULongArrayBuilder ? builder : THROW_CCE(), checkIndex);
  };
  protoOf(ULongArraySerializer_0).on = function (decoder, index, builder, checkIndex) {
    return this.gr(decoder, index, builder instanceof ULongArrayBuilder ? builder : THROW_CCE(), checkIndex);
  };
  var ULongArraySerializer_instance;
  function ULongArraySerializer_getInstance() {
    if (ULongArraySerializer_instance == null)
      new ULongArraySerializer_0();
    return ULongArraySerializer_instance;
  }
  function IntArraySerializer_0() {
    IntArraySerializer_instance = this;
    PrimitiveArraySerializer.call(this, serializer_5(IntCompanionObject_instance));
  }
  protoOf(IntArraySerializer_0).mr = function (_this__u8e3s4) {
    return new IntArrayBuilder(_this__u8e3s4);
  };
  protoOf(IntArraySerializer_0).km = function (_this__u8e3s4) {
    return this.mr((!(_this__u8e3s4 == null) ? isIntArray(_this__u8e3s4) : false) ? _this__u8e3s4 : THROW_CCE());
  };
  protoOf(IntArraySerializer_0).nn = function () {
    return new Int32Array(0);
  };
  protoOf(IntArraySerializer_0).nr = function (decoder, index, builder, checkIndex) {
    builder.qr(decoder.wk(this.gn_1, index));
  };
  protoOf(IntArraySerializer_0).rm = function (decoder, index, builder, checkIndex) {
    return this.nr(decoder, index, builder instanceof IntArrayBuilder ? builder : THROW_CCE(), checkIndex);
  };
  protoOf(IntArraySerializer_0).on = function (decoder, index, builder, checkIndex) {
    return this.nr(decoder, index, builder instanceof IntArrayBuilder ? builder : THROW_CCE(), checkIndex);
  };
  var IntArraySerializer_instance;
  function IntArraySerializer_getInstance() {
    if (IntArraySerializer_instance == null)
      new IntArraySerializer_0();
    return IntArraySerializer_instance;
  }
  function UIntArraySerializer_0() {
    UIntArraySerializer_instance = this;
    PrimitiveArraySerializer.call(this, serializer_6(Companion_getInstance_3()));
  }
  protoOf(UIntArraySerializer_0).tr = function (_this__u8e3s4) {
    return new UIntArrayBuilder(_this__u8e3s4);
  };
  protoOf(UIntArraySerializer_0).km = function (_this__u8e3s4) {
    return this.tr(_this__u8e3s4 instanceof UIntArray ? _this__u8e3s4.dh_1 : THROW_CCE());
  };
  protoOf(UIntArraySerializer_0).ur = function () {
    return _UIntArray___init__impl__ghjpc6(0);
  };
  protoOf(UIntArraySerializer_0).nn = function () {
    return new UIntArray(this.ur());
  };
  protoOf(UIntArraySerializer_0).vr = function (decoder, index, builder, checkIndex) {
    // Inline function 'kotlin.toUInt' call
    var this_0 = decoder.bl(this.gn_1, index).hk();
    var tmp$ret$0 = _UInt___init__impl__l7qpdl(this_0);
    builder.yr(tmp$ret$0);
  };
  protoOf(UIntArraySerializer_0).rm = function (decoder, index, builder, checkIndex) {
    return this.vr(decoder, index, builder instanceof UIntArrayBuilder ? builder : THROW_CCE(), checkIndex);
  };
  protoOf(UIntArraySerializer_0).on = function (decoder, index, builder, checkIndex) {
    return this.vr(decoder, index, builder instanceof UIntArrayBuilder ? builder : THROW_CCE(), checkIndex);
  };
  var UIntArraySerializer_instance;
  function UIntArraySerializer_getInstance() {
    if (UIntArraySerializer_instance == null)
      new UIntArraySerializer_0();
    return UIntArraySerializer_instance;
  }
  function ShortArraySerializer_0() {
    ShortArraySerializer_instance = this;
    PrimitiveArraySerializer.call(this, serializer_7(ShortCompanionObject_instance));
  }
  protoOf(ShortArraySerializer_0).bs = function (_this__u8e3s4) {
    return new ShortArrayBuilder(_this__u8e3s4);
  };
  protoOf(ShortArraySerializer_0).km = function (_this__u8e3s4) {
    return this.bs((!(_this__u8e3s4 == null) ? isShortArray(_this__u8e3s4) : false) ? _this__u8e3s4 : THROW_CCE());
  };
  protoOf(ShortArraySerializer_0).nn = function () {
    return new Int16Array(0);
  };
  protoOf(ShortArraySerializer_0).cs = function (decoder, index, builder, checkIndex) {
    builder.fs(decoder.vk(this.gn_1, index));
  };
  protoOf(ShortArraySerializer_0).rm = function (decoder, index, builder, checkIndex) {
    return this.cs(decoder, index, builder instanceof ShortArrayBuilder ? builder : THROW_CCE(), checkIndex);
  };
  protoOf(ShortArraySerializer_0).on = function (decoder, index, builder, checkIndex) {
    return this.cs(decoder, index, builder instanceof ShortArrayBuilder ? builder : THROW_CCE(), checkIndex);
  };
  var ShortArraySerializer_instance;
  function ShortArraySerializer_getInstance() {
    if (ShortArraySerializer_instance == null)
      new ShortArraySerializer_0();
    return ShortArraySerializer_instance;
  }
  function UShortArraySerializer_0() {
    UShortArraySerializer_instance = this;
    PrimitiveArraySerializer.call(this, serializer_8(Companion_getInstance_4()));
  }
  protoOf(UShortArraySerializer_0).is = function (_this__u8e3s4) {
    return new UShortArrayBuilder(_this__u8e3s4);
  };
  protoOf(UShortArraySerializer_0).km = function (_this__u8e3s4) {
    return this.is(_this__u8e3s4 instanceof UShortArray ? _this__u8e3s4.zh_1 : THROW_CCE());
  };
  protoOf(UShortArraySerializer_0).js = function () {
    return _UShortArray___init__impl__9b26ef(0);
  };
  protoOf(UShortArraySerializer_0).nn = function () {
    return new UShortArray(this.js());
  };
  protoOf(UShortArraySerializer_0).ks = function (decoder, index, builder, checkIndex) {
    // Inline function 'kotlin.toUShort' call
    var this_0 = decoder.bl(this.gn_1, index).gk();
    var tmp$ret$0 = _UShort___init__impl__jigrne(this_0);
    builder.ns(tmp$ret$0);
  };
  protoOf(UShortArraySerializer_0).rm = function (decoder, index, builder, checkIndex) {
    return this.ks(decoder, index, builder instanceof UShortArrayBuilder ? builder : THROW_CCE(), checkIndex);
  };
  protoOf(UShortArraySerializer_0).on = function (decoder, index, builder, checkIndex) {
    return this.ks(decoder, index, builder instanceof UShortArrayBuilder ? builder : THROW_CCE(), checkIndex);
  };
  var UShortArraySerializer_instance;
  function UShortArraySerializer_getInstance() {
    if (UShortArraySerializer_instance == null)
      new UShortArraySerializer_0();
    return UShortArraySerializer_instance;
  }
  function ByteArraySerializer_0() {
    ByteArraySerializer_instance = this;
    PrimitiveArraySerializer.call(this, serializer_9(ByteCompanionObject_instance));
  }
  protoOf(ByteArraySerializer_0).qs = function (_this__u8e3s4) {
    return new ByteArrayBuilder(_this__u8e3s4);
  };
  protoOf(ByteArraySerializer_0).km = function (_this__u8e3s4) {
    return this.qs((!(_this__u8e3s4 == null) ? isByteArray(_this__u8e3s4) : false) ? _this__u8e3s4 : THROW_CCE());
  };
  protoOf(ByteArraySerializer_0).nn = function () {
    return new Int8Array(0);
  };
  protoOf(ByteArraySerializer_0).rs = function (decoder, index, builder, checkIndex) {
    builder.us(decoder.uk(this.gn_1, index));
  };
  protoOf(ByteArraySerializer_0).rm = function (decoder, index, builder, checkIndex) {
    return this.rs(decoder, index, builder instanceof ByteArrayBuilder ? builder : THROW_CCE(), checkIndex);
  };
  protoOf(ByteArraySerializer_0).on = function (decoder, index, builder, checkIndex) {
    return this.rs(decoder, index, builder instanceof ByteArrayBuilder ? builder : THROW_CCE(), checkIndex);
  };
  var ByteArraySerializer_instance;
  function ByteArraySerializer_getInstance() {
    if (ByteArraySerializer_instance == null)
      new ByteArraySerializer_0();
    return ByteArraySerializer_instance;
  }
  function UByteArraySerializer_0() {
    UByteArraySerializer_instance = this;
    PrimitiveArraySerializer.call(this, serializer_10(Companion_getInstance_5()));
  }
  protoOf(UByteArraySerializer_0).xs = function (_this__u8e3s4) {
    return new UByteArrayBuilder(_this__u8e3s4);
  };
  protoOf(UByteArraySerializer_0).km = function (_this__u8e3s4) {
    return this.xs(_this__u8e3s4 instanceof UByteArray ? _this__u8e3s4.sg_1 : THROW_CCE());
  };
  protoOf(UByteArraySerializer_0).ys = function () {
    return _UByteArray___init__impl__ip4y9n(0);
  };
  protoOf(UByteArraySerializer_0).nn = function () {
    return new UByteArray(this.ys());
  };
  protoOf(UByteArraySerializer_0).zs = function (decoder, index, builder, checkIndex) {
    // Inline function 'kotlin.toUByte' call
    var this_0 = decoder.bl(this.gn_1, index).fk();
    var tmp$ret$0 = _UByte___init__impl__g9hnc4(this_0);
    builder.ct(tmp$ret$0);
  };
  protoOf(UByteArraySerializer_0).rm = function (decoder, index, builder, checkIndex) {
    return this.zs(decoder, index, builder instanceof UByteArrayBuilder ? builder : THROW_CCE(), checkIndex);
  };
  protoOf(UByteArraySerializer_0).on = function (decoder, index, builder, checkIndex) {
    return this.zs(decoder, index, builder instanceof UByteArrayBuilder ? builder : THROW_CCE(), checkIndex);
  };
  var UByteArraySerializer_instance;
  function UByteArraySerializer_getInstance() {
    if (UByteArraySerializer_instance == null)
      new UByteArraySerializer_0();
    return UByteArraySerializer_instance;
  }
  function BooleanArraySerializer_0() {
    BooleanArraySerializer_instance = this;
    PrimitiveArraySerializer.call(this, serializer_11(BooleanCompanionObject_instance));
  }
  protoOf(BooleanArraySerializer_0).ft = function (_this__u8e3s4) {
    return new BooleanArrayBuilder(_this__u8e3s4);
  };
  protoOf(BooleanArraySerializer_0).km = function (_this__u8e3s4) {
    return this.ft((!(_this__u8e3s4 == null) ? isBooleanArray(_this__u8e3s4) : false) ? _this__u8e3s4 : THROW_CCE());
  };
  protoOf(BooleanArraySerializer_0).nn = function () {
    return booleanArray(0);
  };
  protoOf(BooleanArraySerializer_0).gt = function (decoder, index, builder, checkIndex) {
    builder.jt(decoder.tk(this.gn_1, index));
  };
  protoOf(BooleanArraySerializer_0).rm = function (decoder, index, builder, checkIndex) {
    return this.gt(decoder, index, builder instanceof BooleanArrayBuilder ? builder : THROW_CCE(), checkIndex);
  };
  protoOf(BooleanArraySerializer_0).on = function (decoder, index, builder, checkIndex) {
    return this.gt(decoder, index, builder instanceof BooleanArrayBuilder ? builder : THROW_CCE(), checkIndex);
  };
  var BooleanArraySerializer_instance;
  function BooleanArraySerializer_getInstance() {
    if (BooleanArraySerializer_instance == null)
      new BooleanArraySerializer_0();
    return BooleanArraySerializer_instance;
  }
  function CharArrayBuilder(bufferWithData) {
    PrimitiveArrayBuilder.call(this);
    this.eq_1 = bufferWithData;
    this.fq_1 = bufferWithData.length;
    this.k2(10);
  }
  protoOf(CharArrayBuilder).in = function () {
    return this.fq_1;
  };
  protoOf(CharArrayBuilder).k2 = function (requiredCapacity) {
    if (this.eq_1.length < requiredCapacity)
      this.eq_1 = copyOf(this.eq_1, coerceAtLeast(requiredCapacity, imul(this.eq_1.length, 2)));
  };
  protoOf(CharArrayBuilder).gq = function (c) {
    this.pn();
    var tmp = this.eq_1;
    var tmp1 = this.fq_1;
    this.fq_1 = tmp1 + 1 | 0;
    tmp[tmp1] = c;
  };
  protoOf(CharArrayBuilder).kn = function () {
    return copyOf(this.eq_1, this.fq_1);
  };
  function DoubleArrayBuilder(bufferWithData) {
    PrimitiveArrayBuilder.call(this);
    this.lq_1 = bufferWithData;
    this.mq_1 = bufferWithData.length;
    this.k2(10);
  }
  protoOf(DoubleArrayBuilder).in = function () {
    return this.mq_1;
  };
  protoOf(DoubleArrayBuilder).k2 = function (requiredCapacity) {
    if (this.lq_1.length < requiredCapacity)
      this.lq_1 = copyOf_0(this.lq_1, coerceAtLeast(requiredCapacity, imul(this.lq_1.length, 2)));
  };
  protoOf(DoubleArrayBuilder).nq = function (c) {
    this.pn();
    var tmp = this.lq_1;
    var tmp1 = this.mq_1;
    this.mq_1 = tmp1 + 1 | 0;
    tmp[tmp1] = c;
  };
  protoOf(DoubleArrayBuilder).kn = function () {
    return copyOf_0(this.lq_1, this.mq_1);
  };
  function FloatArrayBuilder(bufferWithData) {
    PrimitiveArrayBuilder.call(this);
    this.sq_1 = bufferWithData;
    this.tq_1 = bufferWithData.length;
    this.k2(10);
  }
  protoOf(FloatArrayBuilder).in = function () {
    return this.tq_1;
  };
  protoOf(FloatArrayBuilder).k2 = function (requiredCapacity) {
    if (this.sq_1.length < requiredCapacity)
      this.sq_1 = copyOf_1(this.sq_1, coerceAtLeast(requiredCapacity, imul(this.sq_1.length, 2)));
  };
  protoOf(FloatArrayBuilder).uq = function (c) {
    this.pn();
    var tmp = this.sq_1;
    var tmp1 = this.tq_1;
    this.tq_1 = tmp1 + 1 | 0;
    tmp[tmp1] = c;
  };
  protoOf(FloatArrayBuilder).kn = function () {
    return copyOf_1(this.sq_1, this.tq_1);
  };
  function LongArrayBuilder(bufferWithData) {
    PrimitiveArrayBuilder.call(this);
    this.zq_1 = bufferWithData;
    this.ar_1 = bufferWithData.length;
    this.k2(10);
  }
  protoOf(LongArrayBuilder).in = function () {
    return this.ar_1;
  };
  protoOf(LongArrayBuilder).k2 = function (requiredCapacity) {
    if (this.zq_1.length < requiredCapacity)
      this.zq_1 = copyOf_2(this.zq_1, coerceAtLeast(requiredCapacity, imul(this.zq_1.length, 2)));
  };
  protoOf(LongArrayBuilder).br = function (c) {
    this.pn();
    var tmp = this.zq_1;
    var tmp1 = this.ar_1;
    this.ar_1 = tmp1 + 1 | 0;
    tmp[tmp1] = c;
  };
  protoOf(LongArrayBuilder).kn = function () {
    return copyOf_2(this.zq_1, this.ar_1);
  };
  function ULongArrayBuilder(bufferWithData) {
    PrimitiveArrayBuilder.call(this);
    this.hr_1 = bufferWithData;
    this.ir_1 = _ULongArray___get_size__impl__ju6dtr(bufferWithData);
    this.k2(10);
  }
  protoOf(ULongArrayBuilder).in = function () {
    return this.ir_1;
  };
  protoOf(ULongArrayBuilder).k2 = function (requiredCapacity) {
    if (_ULongArray___get_size__impl__ju6dtr(this.hr_1) < requiredCapacity) {
      var tmp = this;
      // Inline function 'kotlin.collections.copyOf' call
      var this_0 = this.hr_1;
      var newSize = coerceAtLeast(requiredCapacity, imul(_ULongArray___get_size__impl__ju6dtr(this.hr_1), 2));
      tmp.hr_1 = _ULongArray___init__impl__twm1l3_0(copyOf_2(_ULongArray___get_storage__impl__28e64j(this_0), newSize));
    }
  };
  protoOf(ULongArrayBuilder).jr = function (c) {
    this.pn();
    var tmp = this.hr_1;
    var tmp1 = this.ir_1;
    this.ir_1 = tmp1 + 1 | 0;
    ULongArray__set_impl_z19mvh(tmp, tmp1, c);
  };
  protoOf(ULongArrayBuilder).kt = function () {
    // Inline function 'kotlin.collections.copyOf' call
    var this_0 = this.hr_1;
    var newSize = this.ir_1;
    return _ULongArray___init__impl__twm1l3_0(copyOf_2(_ULongArray___get_storage__impl__28e64j(this_0), newSize));
  };
  protoOf(ULongArrayBuilder).kn = function () {
    return new ULongArray(this.kt());
  };
  function IntArrayBuilder(bufferWithData) {
    PrimitiveArrayBuilder.call(this);
    this.or_1 = bufferWithData;
    this.pr_1 = bufferWithData.length;
    this.k2(10);
  }
  protoOf(IntArrayBuilder).in = function () {
    return this.pr_1;
  };
  protoOf(IntArrayBuilder).k2 = function (requiredCapacity) {
    if (this.or_1.length < requiredCapacity)
      this.or_1 = copyOf_3(this.or_1, coerceAtLeast(requiredCapacity, imul(this.or_1.length, 2)));
  };
  protoOf(IntArrayBuilder).qr = function (c) {
    this.pn();
    var tmp = this.or_1;
    var tmp1 = this.pr_1;
    this.pr_1 = tmp1 + 1 | 0;
    tmp[tmp1] = c;
  };
  protoOf(IntArrayBuilder).kn = function () {
    return copyOf_3(this.or_1, this.pr_1);
  };
  function UIntArrayBuilder(bufferWithData) {
    PrimitiveArrayBuilder.call(this);
    this.wr_1 = bufferWithData;
    this.xr_1 = _UIntArray___get_size__impl__r6l8ci(bufferWithData);
    this.k2(10);
  }
  protoOf(UIntArrayBuilder).in = function () {
    return this.xr_1;
  };
  protoOf(UIntArrayBuilder).k2 = function (requiredCapacity) {
    if (_UIntArray___get_size__impl__r6l8ci(this.wr_1) < requiredCapacity) {
      var tmp = this;
      // Inline function 'kotlin.collections.copyOf' call
      var this_0 = this.wr_1;
      var newSize = coerceAtLeast(requiredCapacity, imul(_UIntArray___get_size__impl__r6l8ci(this.wr_1), 2));
      tmp.wr_1 = _UIntArray___init__impl__ghjpc6_0(copyOf_3(_UIntArray___get_storage__impl__92a0v0(this_0), newSize));
    }
  };
  protoOf(UIntArrayBuilder).yr = function (c) {
    this.pn();
    var tmp = this.wr_1;
    var tmp1 = this.xr_1;
    this.xr_1 = tmp1 + 1 | 0;
    UIntArray__set_impl_7f2zu2(tmp, tmp1, c);
  };
  protoOf(UIntArrayBuilder).lt = function () {
    // Inline function 'kotlin.collections.copyOf' call
    var this_0 = this.wr_1;
    var newSize = this.xr_1;
    return _UIntArray___init__impl__ghjpc6_0(copyOf_3(_UIntArray___get_storage__impl__92a0v0(this_0), newSize));
  };
  protoOf(UIntArrayBuilder).kn = function () {
    return new UIntArray(this.lt());
  };
  function ShortArrayBuilder(bufferWithData) {
    PrimitiveArrayBuilder.call(this);
    this.ds_1 = bufferWithData;
    this.es_1 = bufferWithData.length;
    this.k2(10);
  }
  protoOf(ShortArrayBuilder).in = function () {
    return this.es_1;
  };
  protoOf(ShortArrayBuilder).k2 = function (requiredCapacity) {
    if (this.ds_1.length < requiredCapacity)
      this.ds_1 = copyOf_4(this.ds_1, coerceAtLeast(requiredCapacity, imul(this.ds_1.length, 2)));
  };
  protoOf(ShortArrayBuilder).fs = function (c) {
    this.pn();
    var tmp = this.ds_1;
    var tmp1 = this.es_1;
    this.es_1 = tmp1 + 1 | 0;
    tmp[tmp1] = c;
  };
  protoOf(ShortArrayBuilder).kn = function () {
    return copyOf_4(this.ds_1, this.es_1);
  };
  function UShortArrayBuilder(bufferWithData) {
    PrimitiveArrayBuilder.call(this);
    this.ls_1 = bufferWithData;
    this.ms_1 = _UShortArray___get_size__impl__jqto1b(bufferWithData);
    this.k2(10);
  }
  protoOf(UShortArrayBuilder).in = function () {
    return this.ms_1;
  };
  protoOf(UShortArrayBuilder).k2 = function (requiredCapacity) {
    if (_UShortArray___get_size__impl__jqto1b(this.ls_1) < requiredCapacity) {
      var tmp = this;
      // Inline function 'kotlin.collections.copyOf' call
      var this_0 = this.ls_1;
      var newSize = coerceAtLeast(requiredCapacity, imul(_UShortArray___get_size__impl__jqto1b(this.ls_1), 2));
      tmp.ls_1 = _UShortArray___init__impl__9b26ef_0(copyOf_4(_UShortArray___get_storage__impl__t2jpv5(this_0), newSize));
    }
  };
  protoOf(UShortArrayBuilder).ns = function (c) {
    this.pn();
    var tmp = this.ls_1;
    var tmp1 = this.ms_1;
    this.ms_1 = tmp1 + 1 | 0;
    UShortArray__set_impl_6d8whp(tmp, tmp1, c);
  };
  protoOf(UShortArrayBuilder).mt = function () {
    // Inline function 'kotlin.collections.copyOf' call
    var this_0 = this.ls_1;
    var newSize = this.ms_1;
    return _UShortArray___init__impl__9b26ef_0(copyOf_4(_UShortArray___get_storage__impl__t2jpv5(this_0), newSize));
  };
  protoOf(UShortArrayBuilder).kn = function () {
    return new UShortArray(this.mt());
  };
  function ByteArrayBuilder(bufferWithData) {
    PrimitiveArrayBuilder.call(this);
    this.ss_1 = bufferWithData;
    this.ts_1 = bufferWithData.length;
    this.k2(10);
  }
  protoOf(ByteArrayBuilder).in = function () {
    return this.ts_1;
  };
  protoOf(ByteArrayBuilder).k2 = function (requiredCapacity) {
    if (this.ss_1.length < requiredCapacity)
      this.ss_1 = copyOf_5(this.ss_1, coerceAtLeast(requiredCapacity, imul(this.ss_1.length, 2)));
  };
  protoOf(ByteArrayBuilder).us = function (c) {
    this.pn();
    var tmp = this.ss_1;
    var tmp1 = this.ts_1;
    this.ts_1 = tmp1 + 1 | 0;
    tmp[tmp1] = c;
  };
  protoOf(ByteArrayBuilder).kn = function () {
    return copyOf_5(this.ss_1, this.ts_1);
  };
  function UByteArrayBuilder(bufferWithData) {
    PrimitiveArrayBuilder.call(this);
    this.at_1 = bufferWithData;
    this.bt_1 = _UByteArray___get_size__impl__h6pkdv(bufferWithData);
    this.k2(10);
  }
  protoOf(UByteArrayBuilder).in = function () {
    return this.bt_1;
  };
  protoOf(UByteArrayBuilder).k2 = function (requiredCapacity) {
    if (_UByteArray___get_size__impl__h6pkdv(this.at_1) < requiredCapacity) {
      var tmp = this;
      // Inline function 'kotlin.collections.copyOf' call
      var this_0 = this.at_1;
      var newSize = coerceAtLeast(requiredCapacity, imul(_UByteArray___get_size__impl__h6pkdv(this.at_1), 2));
      tmp.at_1 = _UByteArray___init__impl__ip4y9n_0(copyOf_5(_UByteArray___get_storage__impl__d4kctt(this_0), newSize));
    }
  };
  protoOf(UByteArrayBuilder).ct = function (c) {
    this.pn();
    var tmp = this.at_1;
    var tmp1 = this.bt_1;
    this.bt_1 = tmp1 + 1 | 0;
    UByteArray__set_impl_jvcicn(tmp, tmp1, c);
  };
  protoOf(UByteArrayBuilder).nt = function () {
    // Inline function 'kotlin.collections.copyOf' call
    var this_0 = this.at_1;
    var newSize = this.bt_1;
    return _UByteArray___init__impl__ip4y9n_0(copyOf_5(_UByteArray___get_storage__impl__d4kctt(this_0), newSize));
  };
  protoOf(UByteArrayBuilder).kn = function () {
    return new UByteArray(this.nt());
  };
  function BooleanArrayBuilder(bufferWithData) {
    PrimitiveArrayBuilder.call(this);
    this.ht_1 = bufferWithData;
    this.it_1 = bufferWithData.length;
    this.k2(10);
  }
  protoOf(BooleanArrayBuilder).in = function () {
    return this.it_1;
  };
  protoOf(BooleanArrayBuilder).k2 = function (requiredCapacity) {
    if (this.ht_1.length < requiredCapacity)
      this.ht_1 = copyOf_6(this.ht_1, coerceAtLeast(requiredCapacity, imul(this.ht_1.length, 2)));
  };
  protoOf(BooleanArrayBuilder).jt = function (c) {
    this.pn();
    var tmp = this.ht_1;
    var tmp1 = this.it_1;
    this.it_1 = tmp1 + 1 | 0;
    tmp[tmp1] = c;
  };
  protoOf(BooleanArrayBuilder).kn = function () {
    return copyOf_6(this.ht_1, this.it_1);
  };
  function get_BUILTIN_SERIALIZERS() {
    _init_properties_Primitives_kt__k0eto4();
    return BUILTIN_SERIALIZERS;
  }
  var BUILTIN_SERIALIZERS;
  function StringSerializer() {
    StringSerializer_instance = this;
    this.ot_1 = new PrimitiveSerialDescriptor_0('kotlin.String', STRING_getInstance());
  }
  protoOf(StringSerializer).fi = function () {
    return this.ot_1;
  };
  protoOf(StringSerializer).gi = function (decoder) {
    return decoder.mk();
  };
  var StringSerializer_instance;
  function StringSerializer_getInstance() {
    if (StringSerializer_instance == null)
      new StringSerializer();
    return StringSerializer_instance;
  }
  function CharSerializer() {
    CharSerializer_instance = this;
    this.pt_1 = new PrimitiveSerialDescriptor_0('kotlin.Char', CHAR_getInstance());
  }
  protoOf(CharSerializer).fi = function () {
    return this.pt_1;
  };
  protoOf(CharSerializer).qt = function (decoder) {
    return decoder.lk();
  };
  protoOf(CharSerializer).gi = function (decoder) {
    return new Char(this.qt(decoder));
  };
  var CharSerializer_instance;
  function CharSerializer_getInstance() {
    if (CharSerializer_instance == null)
      new CharSerializer();
    return CharSerializer_instance;
  }
  function DoubleSerializer() {
    DoubleSerializer_instance = this;
    this.rt_1 = new PrimitiveSerialDescriptor_0('kotlin.Double', DOUBLE_getInstance());
  }
  protoOf(DoubleSerializer).fi = function () {
    return this.rt_1;
  };
  protoOf(DoubleSerializer).gi = function (decoder) {
    return decoder.kk();
  };
  var DoubleSerializer_instance;
  function DoubleSerializer_getInstance() {
    if (DoubleSerializer_instance == null)
      new DoubleSerializer();
    return DoubleSerializer_instance;
  }
  function FloatSerializer() {
    FloatSerializer_instance = this;
    this.st_1 = new PrimitiveSerialDescriptor_0('kotlin.Float', FLOAT_getInstance());
  }
  protoOf(FloatSerializer).fi = function () {
    return this.st_1;
  };
  protoOf(FloatSerializer).gi = function (decoder) {
    return decoder.jk();
  };
  var FloatSerializer_instance;
  function FloatSerializer_getInstance() {
    if (FloatSerializer_instance == null)
      new FloatSerializer();
    return FloatSerializer_instance;
  }
  function LongSerializer() {
    LongSerializer_instance = this;
    this.tt_1 = new PrimitiveSerialDescriptor_0('kotlin.Long', LONG_getInstance());
  }
  protoOf(LongSerializer).fi = function () {
    return this.tt_1;
  };
  protoOf(LongSerializer).gi = function (decoder) {
    return decoder.ik();
  };
  var LongSerializer_instance;
  function LongSerializer_getInstance() {
    if (LongSerializer_instance == null)
      new LongSerializer();
    return LongSerializer_instance;
  }
  function IntSerializer() {
    IntSerializer_instance = this;
    this.ut_1 = new PrimitiveSerialDescriptor_0('kotlin.Int', INT_getInstance());
  }
  protoOf(IntSerializer).fi = function () {
    return this.ut_1;
  };
  protoOf(IntSerializer).gi = function (decoder) {
    return decoder.hk();
  };
  var IntSerializer_instance;
  function IntSerializer_getInstance() {
    if (IntSerializer_instance == null)
      new IntSerializer();
    return IntSerializer_instance;
  }
  function ShortSerializer() {
    ShortSerializer_instance = this;
    this.vt_1 = new PrimitiveSerialDescriptor_0('kotlin.Short', SHORT_getInstance());
  }
  protoOf(ShortSerializer).fi = function () {
    return this.vt_1;
  };
  protoOf(ShortSerializer).gi = function (decoder) {
    return decoder.gk();
  };
  var ShortSerializer_instance;
  function ShortSerializer_getInstance() {
    if (ShortSerializer_instance == null)
      new ShortSerializer();
    return ShortSerializer_instance;
  }
  function ByteSerializer() {
    ByteSerializer_instance = this;
    this.wt_1 = new PrimitiveSerialDescriptor_0('kotlin.Byte', BYTE_getInstance());
  }
  protoOf(ByteSerializer).fi = function () {
    return this.wt_1;
  };
  protoOf(ByteSerializer).gi = function (decoder) {
    return decoder.fk();
  };
  var ByteSerializer_instance;
  function ByteSerializer_getInstance() {
    if (ByteSerializer_instance == null)
      new ByteSerializer();
    return ByteSerializer_instance;
  }
  function BooleanSerializer() {
    BooleanSerializer_instance = this;
    this.xt_1 = new PrimitiveSerialDescriptor_0('kotlin.Boolean', BOOLEAN_getInstance());
  }
  protoOf(BooleanSerializer).fi = function () {
    return this.xt_1;
  };
  protoOf(BooleanSerializer).gi = function (decoder) {
    return decoder.ek();
  };
  var BooleanSerializer_instance;
  function BooleanSerializer_getInstance() {
    if (BooleanSerializer_instance == null)
      new BooleanSerializer();
    return BooleanSerializer_instance;
  }
  function UnitSerializer() {
    UnitSerializer_instance = this;
    this.yt_1 = new ObjectSerializer('kotlin.Unit', Unit_instance);
  }
  protoOf(UnitSerializer).fi = function () {
    return this.yt_1.fi();
  };
  protoOf(UnitSerializer).zt = function (decoder) {
    this.yt_1.gi(decoder);
  };
  protoOf(UnitSerializer).gi = function (decoder) {
    this.zt(decoder);
    return Unit_instance;
  };
  var UnitSerializer_instance;
  function UnitSerializer_getInstance() {
    if (UnitSerializer_instance == null)
      new UnitSerializer();
    return UnitSerializer_instance;
  }
  function error_0($this) {
    throw IllegalStateException_init_$Create$('Primitive descriptor does not have elements');
  }
  function PrimitiveSerialDescriptor_0(serialName, kind) {
    this.au_1 = serialName;
    this.bu_1 = kind;
  }
  protoOf(PrimitiveSerialDescriptor_0).oi = function () {
    return this.au_1;
  };
  protoOf(PrimitiveSerialDescriptor_0).pi = function () {
    return this.bu_1;
  };
  protoOf(PrimitiveSerialDescriptor_0).ri = function () {
    return 0;
  };
  protoOf(PrimitiveSerialDescriptor_0).ti = function (index) {
    error_0(this);
  };
  protoOf(PrimitiveSerialDescriptor_0).ui = function (name) {
    error_0(this);
  };
  protoOf(PrimitiveSerialDescriptor_0).xi = function (index) {
    error_0(this);
  };
  protoOf(PrimitiveSerialDescriptor_0).wi = function (index) {
    error_0(this);
  };
  protoOf(PrimitiveSerialDescriptor_0).vi = function (index) {
    error_0(this);
  };
  protoOf(PrimitiveSerialDescriptor_0).toString = function () {
    return 'PrimitiveDescriptor(' + this.au_1 + ')';
  };
  protoOf(PrimitiveSerialDescriptor_0).equals = function (other) {
    if (this === other)
      return true;
    if (!(other instanceof PrimitiveSerialDescriptor_0))
      return false;
    if (this.au_1 === other.au_1 ? equals(this.bu_1, other.bu_1) : false)
      return true;
    return false;
  };
  protoOf(PrimitiveSerialDescriptor_0).hashCode = function () {
    return getStringHashCode(this.au_1) + imul(31, this.bu_1.hashCode()) | 0;
  };
  function PrimitiveDescriptorSafe(serialName, kind) {
    _init_properties_Primitives_kt__k0eto4();
    checkName(serialName);
    return new PrimitiveSerialDescriptor_0(serialName, kind);
  }
  function checkName(serialName) {
    _init_properties_Primitives_kt__k0eto4();
    var keys = get_BUILTIN_SERIALIZERS().p1();
    var tmp0_iterator = keys.r();
    while (tmp0_iterator.s()) {
      var primitive = tmp0_iterator.t();
      var simpleName = capitalize(ensureNotNull(primitive.c6()));
      var qualifiedName = 'kotlin.' + simpleName;
      if (equals_0(serialName, qualifiedName, true) ? true : equals_0(serialName, simpleName, true)) {
        throw IllegalArgumentException_init_$Create$(trimIndent('\n                The name of serial descriptor should uniquely identify associated serializer.\n                For serial name ' + serialName + ' there already exist ' + capitalize(simpleName) + 'Serializer.\n                Please refer to SerialDescriptor documentation for additional information.\n            '));
      }
    }
  }
  function capitalize(_this__u8e3s4) {
    _init_properties_Primitives_kt__k0eto4();
    // Inline function 'kotlin.text.replaceFirstChar' call
    var tmp;
    // Inline function 'kotlin.text.isNotEmpty' call
    if (charSequenceLength(_this__u8e3s4) > 0) {
      // Inline function 'kotlinx.serialization.internal.capitalize.<anonymous>' call
      var it = charSequenceGet(_this__u8e3s4, 0);
      var tmp$ret$1 = isLowerCase(it) ? titlecase(it) : toString_0(it);
      var tmp_0 = toString(tmp$ret$1);
      // Inline function 'kotlin.text.substring' call
      // Inline function 'kotlin.js.asDynamic' call
      tmp = tmp_0 + _this__u8e3s4.substring(1);
    } else {
      tmp = _this__u8e3s4;
    }
    return tmp;
  }
  var properties_initialized_Primitives_kt_6dpii6;
  function _init_properties_Primitives_kt__k0eto4() {
    if (!properties_initialized_Primitives_kt_6dpii6) {
      properties_initialized_Primitives_kt_6dpii6 = true;
      BUILTIN_SERIALIZERS = mapOf([to(PrimitiveClasses_getInstance().w6(), serializer(StringCompanionObject_instance)), to(getKClass(Char), serializer_0(Companion_getInstance_1())), to(PrimitiveClasses_getInstance().z6(), CharArraySerializer()), to(PrimitiveClasses_getInstance().u6(), serializer_1(DoubleCompanionObject_instance)), to(PrimitiveClasses_getInstance().f7(), DoubleArraySerializer()), to(PrimitiveClasses_getInstance().t6(), serializer_2(FloatCompanionObject_instance)), to(PrimitiveClasses_getInstance().e7(), FloatArraySerializer()), to(getKClass(Long), serializer_3(Companion_getInstance_0())), to(PrimitiveClasses_getInstance().d7(), LongArraySerializer()), to(getKClass(ULong), serializer_4(Companion_getInstance_2())), to(getKClass(ULongArray), ULongArraySerializer()), to(PrimitiveClasses_getInstance().s6(), serializer_5(IntCompanionObject_instance)), to(PrimitiveClasses_getInstance().c7(), IntArraySerializer()), to(getKClass(UInt), serializer_6(Companion_getInstance_3())), to(getKClass(UIntArray), UIntArraySerializer()), to(PrimitiveClasses_getInstance().r6(), serializer_7(ShortCompanionObject_instance)), to(PrimitiveClasses_getInstance().b7(), ShortArraySerializer()), to(getKClass(UShort), serializer_8(Companion_getInstance_4())), to(getKClass(UShortArray), UShortArraySerializer()), to(PrimitiveClasses_getInstance().q6(), serializer_9(ByteCompanionObject_instance)), to(PrimitiveClasses_getInstance().a7(), ByteArraySerializer()), to(getKClass(UByte), serializer_10(Companion_getInstance_5())), to(getKClass(UByteArray), UByteArraySerializer()), to(PrimitiveClasses_getInstance().p6(), serializer_11(BooleanCompanionObject_instance)), to(PrimitiveClasses_getInstance().y6(), BooleanArraySerializer()), to(getKClass(Unit), serializer_12(Unit_instance)), to(PrimitiveClasses_getInstance().o6(), NothingSerializer()), to(getKClass(Duration), serializer_13(Companion_getInstance()))]);
    }
  }
  function NamedValueDecoder() {
    TaggedDecoder.call(this);
  }
  protoOf(NamedValueDecoder).eu = function (_this__u8e3s4, index) {
    return this.gu(this.fu(_this__u8e3s4, index));
  };
  protoOf(NamedValueDecoder).gu = function (nestedName) {
    var tmp0_elvis_lhs = this.ju();
    return this.ku(tmp0_elvis_lhs == null ? '' : tmp0_elvis_lhs, nestedName);
  };
  protoOf(NamedValueDecoder).fu = function (descriptor, index) {
    return descriptor.ti(index);
  };
  protoOf(NamedValueDecoder).ku = function (parentName, childName) {
    var tmp;
    // Inline function 'kotlin.text.isEmpty' call
    if (charSequenceLength(parentName) === 0) {
      tmp = childName;
    } else {
      tmp = parentName + '.' + childName;
    }
    return tmp;
  };
  function tagBlock($this, tag, block) {
    $this.yu(tag);
    var r = block();
    if (!$this.iu_1) {
      $this.zu();
    }
    $this.iu_1 = false;
    return r;
  }
  function TaggedDecoder$decodeSerializableElement$lambda(this$0, $deserializer, $previousValue) {
    return function () {
      return this$0.pk($deserializer, $previousValue);
    };
  }
  function TaggedDecoder$decodeNullableSerializableElement$lambda(this$0, $deserializer, $previousValue) {
    return function () {
      return this$0.ck() ? this$0.pk($deserializer, $previousValue) : this$0.dk();
    };
  }
  function TaggedDecoder() {
    var tmp = this;
    // Inline function 'kotlin.collections.arrayListOf' call
    tmp.hu_1 = ArrayList_init_$Create$();
    this.iu_1 = false;
  }
  protoOf(TaggedDecoder).fl = function () {
    return EmptySerializersModule_0();
  };
  protoOf(TaggedDecoder).lu = function (tag) {
    throw SerializationException_init_$Create$_0('' + getKClassFromExpression(this) + " can't retrieve untyped values");
  };
  protoOf(TaggedDecoder).mu = function (tag) {
    return true;
  };
  protoOf(TaggedDecoder).nu = function (tag) {
    var tmp = this.lu(tag);
    return typeof tmp === 'boolean' ? tmp : THROW_CCE();
  };
  protoOf(TaggedDecoder).ou = function (tag) {
    var tmp = this.lu(tag);
    return typeof tmp === 'number' ? tmp : THROW_CCE();
  };
  protoOf(TaggedDecoder).pu = function (tag) {
    var tmp = this.lu(tag);
    return typeof tmp === 'number' ? tmp : THROW_CCE();
  };
  protoOf(TaggedDecoder).qu = function (tag) {
    var tmp = this.lu(tag);
    return typeof tmp === 'number' ? tmp : THROW_CCE();
  };
  protoOf(TaggedDecoder).ru = function (tag) {
    var tmp = this.lu(tag);
    return tmp instanceof Long ? tmp : THROW_CCE();
  };
  protoOf(TaggedDecoder).su = function (tag) {
    var tmp = this.lu(tag);
    return typeof tmp === 'number' ? tmp : THROW_CCE();
  };
  protoOf(TaggedDecoder).tu = function (tag) {
    var tmp = this.lu(tag);
    return typeof tmp === 'number' ? tmp : THROW_CCE();
  };
  protoOf(TaggedDecoder).uu = function (tag) {
    var tmp = this.lu(tag);
    return tmp instanceof Char ? tmp.h9_1 : THROW_CCE();
  };
  protoOf(TaggedDecoder).vu = function (tag) {
    var tmp = this.lu(tag);
    return typeof tmp === 'string' ? tmp : THROW_CCE();
  };
  protoOf(TaggedDecoder).wu = function (tag, enumDescriptor) {
    var tmp = this.lu(tag);
    return typeof tmp === 'number' ? tmp : THROW_CCE();
  };
  protoOf(TaggedDecoder).xu = function (tag, inlineDescriptor) {
    // Inline function 'kotlin.apply' call
    // Inline function 'kotlin.contracts.contract' call
    // Inline function 'kotlinx.serialization.internal.TaggedDecoder.decodeTaggedInline.<anonymous>' call
    this.yu(tag);
    return this;
  };
  protoOf(TaggedDecoder).pk = function (deserializer, previousValue) {
    return this.qk(deserializer);
  };
  protoOf(TaggedDecoder).ok = function (descriptor) {
    return this.xu(this.zu(), descriptor);
  };
  protoOf(TaggedDecoder).ck = function () {
    var tmp0_elvis_lhs = this.ju();
    var tmp;
    if (tmp0_elvis_lhs == null) {
      return false;
    } else {
      tmp = tmp0_elvis_lhs;
    }
    var currentTag = tmp;
    return this.mu(currentTag);
  };
  protoOf(TaggedDecoder).dk = function () {
    return null;
  };
  protoOf(TaggedDecoder).ek = function () {
    return this.nu(this.zu());
  };
  protoOf(TaggedDecoder).fk = function () {
    return this.ou(this.zu());
  };
  protoOf(TaggedDecoder).gk = function () {
    return this.pu(this.zu());
  };
  protoOf(TaggedDecoder).hk = function () {
    return this.qu(this.zu());
  };
  protoOf(TaggedDecoder).ik = function () {
    return this.ru(this.zu());
  };
  protoOf(TaggedDecoder).jk = function () {
    return this.su(this.zu());
  };
  protoOf(TaggedDecoder).kk = function () {
    return this.tu(this.zu());
  };
  protoOf(TaggedDecoder).lk = function () {
    return this.uu(this.zu());
  };
  protoOf(TaggedDecoder).mk = function () {
    return this.vu(this.zu());
  };
  protoOf(TaggedDecoder).nk = function (enumDescriptor) {
    return this.wu(this.zu(), enumDescriptor);
  };
  protoOf(TaggedDecoder).rk = function (descriptor) {
    return this;
  };
  protoOf(TaggedDecoder).sk = function (descriptor) {
  };
  protoOf(TaggedDecoder).tk = function (descriptor, index) {
    return this.nu(this.eu(descriptor, index));
  };
  protoOf(TaggedDecoder).uk = function (descriptor, index) {
    return this.ou(this.eu(descriptor, index));
  };
  protoOf(TaggedDecoder).vk = function (descriptor, index) {
    return this.pu(this.eu(descriptor, index));
  };
  protoOf(TaggedDecoder).wk = function (descriptor, index) {
    return this.qu(this.eu(descriptor, index));
  };
  protoOf(TaggedDecoder).xk = function (descriptor, index) {
    return this.ru(this.eu(descriptor, index));
  };
  protoOf(TaggedDecoder).yk = function (descriptor, index) {
    return this.su(this.eu(descriptor, index));
  };
  protoOf(TaggedDecoder).zk = function (descriptor, index) {
    return this.tu(this.eu(descriptor, index));
  };
  protoOf(TaggedDecoder).al = function (descriptor, index) {
    return this.uu(this.eu(descriptor, index));
  };
  protoOf(TaggedDecoder).bl = function (descriptor, index) {
    return this.xu(this.eu(descriptor, index), descriptor.wi(index));
  };
  protoOf(TaggedDecoder).cl = function (descriptor, index, deserializer, previousValue) {
    var tmp = this.eu(descriptor, index);
    return tagBlock(this, tmp, TaggedDecoder$decodeSerializableElement$lambda(this, deserializer, previousValue));
  };
  protoOf(TaggedDecoder).el = function (descriptor, index, deserializer, previousValue) {
    var tmp = this.eu(descriptor, index);
    return tagBlock(this, tmp, TaggedDecoder$decodeNullableSerializableElement$lambda(this, deserializer, previousValue));
  };
  protoOf(TaggedDecoder).ju = function () {
    return lastOrNull(this.hu_1);
  };
  protoOf(TaggedDecoder).yu = function (name) {
    this.hu_1.o(name);
  };
  protoOf(TaggedDecoder).zu = function () {
    var r = this.hu_1.e1(get_lastIndex_0(this.hu_1));
    this.iu_1 = true;
    return r;
  };
  function ULongSerializer() {
    ULongSerializer_instance = this;
    this.av_1 = InlinePrimitiveDescriptor('kotlin.ULong', serializer_3(Companion_getInstance_0()));
  }
  protoOf(ULongSerializer).fi = function () {
    return this.av_1;
  };
  protoOf(ULongSerializer).bv = function (decoder) {
    // Inline function 'kotlin.toULong' call
    var this_0 = decoder.ok(this.av_1).ik();
    return _ULong___init__impl__c78o9k(this_0);
  };
  protoOf(ULongSerializer).gi = function (decoder) {
    return new ULong(this.bv(decoder));
  };
  var ULongSerializer_instance;
  function ULongSerializer_getInstance() {
    if (ULongSerializer_instance == null)
      new ULongSerializer();
    return ULongSerializer_instance;
  }
  function UIntSerializer() {
    UIntSerializer_instance = this;
    this.cv_1 = InlinePrimitiveDescriptor('kotlin.UInt', serializer_5(IntCompanionObject_instance));
  }
  protoOf(UIntSerializer).fi = function () {
    return this.cv_1;
  };
  protoOf(UIntSerializer).dv = function (decoder) {
    // Inline function 'kotlin.toUInt' call
    var this_0 = decoder.ok(this.cv_1).hk();
    return _UInt___init__impl__l7qpdl(this_0);
  };
  protoOf(UIntSerializer).gi = function (decoder) {
    return new UInt(this.dv(decoder));
  };
  var UIntSerializer_instance;
  function UIntSerializer_getInstance() {
    if (UIntSerializer_instance == null)
      new UIntSerializer();
    return UIntSerializer_instance;
  }
  function UShortSerializer() {
    UShortSerializer_instance = this;
    this.ev_1 = InlinePrimitiveDescriptor('kotlin.UShort', serializer_7(ShortCompanionObject_instance));
  }
  protoOf(UShortSerializer).fi = function () {
    return this.ev_1;
  };
  protoOf(UShortSerializer).fv = function (decoder) {
    // Inline function 'kotlin.toUShort' call
    var this_0 = decoder.ok(this.ev_1).gk();
    return _UShort___init__impl__jigrne(this_0);
  };
  protoOf(UShortSerializer).gi = function (decoder) {
    return new UShort(this.fv(decoder));
  };
  var UShortSerializer_instance;
  function UShortSerializer_getInstance() {
    if (UShortSerializer_instance == null)
      new UShortSerializer();
    return UShortSerializer_instance;
  }
  function UByteSerializer() {
    UByteSerializer_instance = this;
    this.gv_1 = InlinePrimitiveDescriptor('kotlin.UByte', serializer_9(ByteCompanionObject_instance));
  }
  protoOf(UByteSerializer).fi = function () {
    return this.gv_1;
  };
  protoOf(UByteSerializer).hv = function (decoder) {
    // Inline function 'kotlin.toUByte' call
    var this_0 = decoder.ok(this.gv_1).fk();
    return _UByte___init__impl__g9hnc4(this_0);
  };
  protoOf(UByteSerializer).gi = function (decoder) {
    return new UByte(this.hv(decoder));
  };
  var UByteSerializer_instance;
  function UByteSerializer_getInstance() {
    if (UByteSerializer_instance == null)
      new UByteSerializer();
    return UByteSerializer_instance;
  }
  function get_EmptySerializersModuleLegacyJs() {
    _init_properties_SerializersModule_kt__u78ha3();
    return EmptySerializersModule;
  }
  var EmptySerializersModule;
  function SerializersModule() {
  }
  protoOf(SerializersModule).ji = function (kClass, typeArgumentsSerializers, $super) {
    typeArgumentsSerializers = typeArgumentsSerializers === VOID ? emptyList() : typeArgumentsSerializers;
    return $super === VOID ? this.iv(kClass, typeArgumentsSerializers) : $super.iv.call(this, kClass, typeArgumentsSerializers);
  };
  function SerialModuleImpl(class2ContextualFactory, polyBase2Serializers, polyBase2DefaultSerializerProvider, polyBase2NamedSerializers, polyBase2DefaultDeserializerProvider) {
    SerializersModule.call(this);
    this.jv_1 = class2ContextualFactory;
    this.kv_1 = polyBase2Serializers;
    this.lv_1 = polyBase2DefaultSerializerProvider;
    this.mv_1 = polyBase2NamedSerializers;
    this.nv_1 = polyBase2DefaultDeserializerProvider;
  }
  protoOf(SerialModuleImpl).nl = function (baseClass, serializedClassName) {
    var tmp0_safe_receiver = this.mv_1.b2(baseClass);
    var tmp;
    if (tmp0_safe_receiver == null) {
      tmp = null;
    } else {
      // Inline function 'kotlin.collections.get' call
      tmp = (isInterface(tmp0_safe_receiver, Map) ? tmp0_safe_receiver : THROW_CCE()).b2(serializedClassName);
    }
    var tmp_0 = tmp;
    var registered = (!(tmp_0 == null) ? isInterface(tmp_0, KSerializer) : false) ? tmp_0 : null;
    if (!(registered == null))
      return registered;
    var tmp_1 = this.nv_1.b2(baseClass);
    var tmp1_safe_receiver = (!(tmp_1 == null) ? typeof tmp_1 === 'function' : false) ? tmp_1 : null;
    return tmp1_safe_receiver == null ? null : tmp1_safe_receiver(serializedClassName);
  };
  protoOf(SerialModuleImpl).iv = function (kClass, typeArgumentsSerializers) {
    var tmp0_safe_receiver = this.jv_1.b2(kClass);
    var tmp = tmp0_safe_receiver == null ? null : tmp0_safe_receiver.ov(typeArgumentsSerializers);
    return (tmp == null ? true : isInterface(tmp, KSerializer)) ? tmp : null;
  };
  var properties_initialized_SerializersModule_kt_fjigjn;
  function _init_properties_SerializersModule_kt__u78ha3() {
    if (!properties_initialized_SerializersModule_kt_fjigjn) {
      properties_initialized_SerializersModule_kt_fjigjn = true;
      EmptySerializersModule = new SerialModuleImpl(emptyMap(), emptyMap(), emptyMap(), emptyMap(), emptyMap());
    }
  }
  function EmptySerializersModule_0() {
    return get_EmptySerializersModuleLegacyJs();
  }
  function SerializableWith() {
  }
  function getChecked(_this__u8e3s4, index) {
    if (!(0 <= index ? index <= (_this__u8e3s4.length - 1 | 0) : false))
      throw IndexOutOfBoundsException_init_$Create$('Index ' + index + ' out of bounds ' + get_indices(_this__u8e3s4));
    return _this__u8e3s4[index];
  }
  function getChecked_0(_this__u8e3s4, index) {
    if (!(0 <= index ? index <= (_this__u8e3s4.length - 1 | 0) : false))
      throw IndexOutOfBoundsException_init_$Create$('Index ' + index + ' out of bounds ' + get_indices_0(_this__u8e3s4));
    return _this__u8e3s4[index];
  }
  //region block: post-declaration
  protoOf(SerialDescriptorImpl).ii = get_isNullable;
  protoOf(SerialDescriptorImpl).qi = get_isInline;
  protoOf(AbstractDecoder).dl = decodeSerializableElement$default;
  protoOf(AbstractDecoder).qk = decodeSerializableValue;
  protoOf(AbstractDecoder).gl = decodeSequentially;
  protoOf(AbstractDecoder).il = decodeCollectionSize;
  protoOf(ListLikeDescriptor).ii = get_isNullable;
  protoOf(ListLikeDescriptor).qi = get_isInline;
  protoOf(ListLikeDescriptor).si = get_annotations;
  protoOf(MapLikeDescriptor).ii = get_isNullable;
  protoOf(MapLikeDescriptor).qi = get_isInline;
  protoOf(MapLikeDescriptor).si = get_annotations;
  protoOf(PluginGeneratedSerialDescriptor).ii = get_isNullable;
  protoOf(PluginGeneratedSerialDescriptor).qi = get_isInline;
  protoOf(InlinePrimitiveDescriptor$1).sp = typeParametersSerializers;
  protoOf(NothingSerialDescriptor).ii = get_isNullable;
  protoOf(NothingSerialDescriptor).qi = get_isInline;
  protoOf(NothingSerialDescriptor).si = get_annotations;
  protoOf(PrimitiveSerialDescriptor_0).ii = get_isNullable;
  protoOf(PrimitiveSerialDescriptor_0).qi = get_isInline;
  protoOf(PrimitiveSerialDescriptor_0).si = get_annotations;
  protoOf(TaggedDecoder).dl = decodeSerializableElement$default;
  protoOf(TaggedDecoder).qk = decodeSerializableValue;
  protoOf(TaggedDecoder).gl = decodeSequentially;
  protoOf(TaggedDecoder).il = decodeCollectionSize;
  //endregion
  //region block: init
  Companion_instance = new Companion();
  //endregion
  //region block: exports
  _.$_$ = _.$_$ || {};
  _.$_$.a = SerializationException_init_$Init$_0;
  _.$_$.b = SerializationException_init_$Create$_0;
  _.$_$.c = UnknownFieldException_init_$Create$;
  _.$_$.d = SEALED_getInstance;
  _.$_$.e = STRING_getInstance;
  _.$_$.f = CONTEXTUAL_getInstance;
  _.$_$.g = ENUM_getInstance;
  _.$_$.h = CLASS_getInstance;
  _.$_$.i = LIST_getInstance;
  _.$_$.j = MAP_getInstance;
  _.$_$.k = Companion_instance;
  _.$_$.l = BooleanSerializer_getInstance;
  _.$_$.m = IntSerializer_getInstance;
  _.$_$.n = StringSerializer_getInstance;
  _.$_$.o = ListSerializer;
  _.$_$.p = MapSerializer;
  _.$_$.q = get_nullable;
  _.$_$.r = serializer;
  _.$_$.s = serializer_8;
  _.$_$.t = serializer_6;
  _.$_$.u = serializer_10;
  _.$_$.v = serializer_4;
  _.$_$.w = PolymorphicKind;
  _.$_$.x = PrimitiveKind;
  _.$_$.y = PrimitiveSerialDescriptor;
  _.$_$.z = get_annotations;
  _.$_$.a1 = get_isInline;
  _.$_$.b1 = get_isNullable;
  _.$_$.c1 = SerialDescriptor;
  _.$_$.d1 = buildSerialDescriptor;
  _.$_$.e1 = getContextualDescriptor;
  _.$_$.f1 = AbstractDecoder;
  _.$_$.g1 = CompositeDecoder;
  _.$_$.h1 = Decoder;
  _.$_$.i1 = AbstractPolymorphicSerializer;
  _.$_$.j1 = ElementMarker;
  _.$_$.k1 = typeParametersSerializers;
  _.$_$.l1 = GeneratedSerializer;
  _.$_$.m1 = InlinePrimitiveDescriptor;
  _.$_$.n1 = LinkedHashMapSerializer;
  _.$_$.o1 = NamedValueDecoder;
  _.$_$.p1 = PluginGeneratedSerialDescriptor;
  _.$_$.q1 = createSimpleEnumSerializer;
  _.$_$.r1 = jsonCachedSerialNames;
  _.$_$.s1 = throwMissingFieldException;
  _.$_$.t1 = EmptySerializersModule_0;
  _.$_$.u1 = DeserializationStrategy;
  _.$_$.v1 = KSerializer;
  _.$_$.w1 = MissingFieldException;
  _.$_$.x1 = SerializationException;
  //endregion
  return _;
}));
