// waltid-sd-jwt/kotlinx-serialization-kotlinx-serialization-json.js
(function (root, factory) {
  if (typeof define === 'function' && define.amd)
    define(['exports', './kotlinx-serialization-kotlinx-serialization-core.js', './kotlin-kotlin-stdlib.js'], factory);
  else if (typeof exports === 'object')
    factory(module.exports, require('./kotlinx-serialization-kotlinx-serialization-core.js'), require('./kotlin-kotlin-stdlib.js'));
  else {
    if (typeof this['kotlinx-serialization-kotlinx-serialization-core'] === 'undefined') {
      throw new Error("Error loading module 'kotlinx-serialization-kotlinx-serialization-json'. Its dependency 'kotlinx-serialization-kotlinx-serialization-core' was not found. Please, check whether 'kotlinx-serialization-kotlinx-serialization-core' is loaded prior to 'kotlinx-serialization-kotlinx-serialization-json'.");
    }
    if (typeof this['kotlin-kotlin-stdlib'] === 'undefined') {
      throw new Error("Error loading module 'kotlinx-serialization-kotlinx-serialization-json'. Its dependency 'kotlin-kotlin-stdlib' was not found. Please, check whether 'kotlin-kotlin-stdlib' is loaded prior to 'kotlinx-serialization-kotlinx-serialization-json'.");
    }
    root['kotlinx-serialization-kotlinx-serialization-json'] = factory(typeof this['kotlinx-serialization-kotlinx-serialization-json'] === 'undefined' ? {} : this['kotlinx-serialization-kotlinx-serialization-json'], this['kotlinx-serialization-kotlinx-serialization-core'], this['kotlin-kotlin-stdlib']);
  }
}(this, function (_, kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core, kotlin_kotlin) {
  'use strict';
  //region block: imports
  var imul = Math.imul;
  var EmptySerializersModule = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.t1;
  var protoOf = kotlin_kotlin.$_$.s7;
  var objectMeta = kotlin_kotlin.$_$.r7;
  var setMetadataFor = kotlin_kotlin.$_$.t7;
  var classMeta = kotlin_kotlin.$_$.t6;
  var VOID = kotlin_kotlin.$_$.c;
  var Decoder = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.h1;
  var CompositeDecoder = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.g1;
  var interfaceMeta = kotlin_kotlin.$_$.b7;
  var StringBuilder_init_$Create$ = kotlin_kotlin.$_$.v;
  var _Char___init__impl__6a9atx = kotlin_kotlin.$_$.m1;
  var Unit_instance = kotlin_kotlin.$_$.o3;
  var THROW_CCE = kotlin_kotlin.$_$.p9;
  var equals = kotlin_kotlin.$_$.v6;
  var hashCode = kotlin_kotlin.$_$.a7;
  var joinToString = kotlin_kotlin.$_$.t4;
  var Map = kotlin_kotlin.$_$.t3;
  var IllegalStateException_init_$Create$ = kotlin_kotlin.$_$.d1;
  var LazyThreadSafetyMode_PUBLICATION_getInstance = kotlin_kotlin.$_$.d;
  var lazy = kotlin_kotlin.$_$.ia;
  var List = kotlin_kotlin.$_$.s3;
  var NumberFormatException_init_$Create$ = kotlin_kotlin.$_$.g1;
  var IntCompanionObject_instance = kotlin_kotlin.$_$.c3;
  var numberRangeToNumber = kotlin_kotlin.$_$.m7;
  var ClosedRange = kotlin_kotlin.$_$.z7;
  var isInterface = kotlin_kotlin.$_$.i7;
  var contains = kotlin_kotlin.$_$.c8;
  var getKClassFromExpression = kotlin_kotlin.$_$.a;
  var IllegalArgumentException_init_$Create$ = kotlin_kotlin.$_$.b1;
  var toString = kotlin_kotlin.$_$.x7;
  var getBooleanHashCode = kotlin_kotlin.$_$.x6;
  var getStringHashCode = kotlin_kotlin.$_$.z6;
  var toDouble = kotlin_kotlin.$_$.v8;
  var StringCompanionObject_instance = kotlin_kotlin.$_$.e3;
  var serializer = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.r;
  var InlinePrimitiveDescriptor = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.m1;
  var LinkedHashMap_init_$Create$ = kotlin_kotlin.$_$.o;
  var ArrayList_init_$Create$ = kotlin_kotlin.$_$.h;
  var MapSerializer = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.p;
  var SerialDescriptor = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.c1;
  var KSerializer = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.v1;
  var SEALED_getInstance = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.d;
  var buildSerialDescriptor = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.d1;
  var STRING_getInstance = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.e;
  var ENUM_getInstance = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.g;
  var ListSerializer = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.o;
  var PrimitiveSerialDescriptor = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.y;
  var lazy_0 = kotlin_kotlin.$_$.ja;
  var get_isNullable = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.b1;
  var get_isInline = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.a1;
  var get_annotations = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.z;
  var KProperty1 = kotlin_kotlin.$_$.g8;
  var getPropertyCallableRef = kotlin_kotlin.$_$.y6;
  var ElementMarker = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.j1;
  var captureStack = kotlin_kotlin.$_$.n6;
  var SerializationException = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.x1;
  var SerializationException_init_$Init$ = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.a;
  var charSequenceLength = kotlin_kotlin.$_$.r6;
  var charSequenceSubSequence = kotlin_kotlin.$_$.s6;
  var coerceAtLeast = kotlin_kotlin.$_$.a8;
  var coerceAtMost = kotlin_kotlin.$_$.b8;
  var Companion_instance = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.k;
  var SerializationException_init_$Create$ = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.b;
  var CLASS_getInstance = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.h;
  var singleOrNull = kotlin_kotlin.$_$.k5;
  var emptyMap = kotlin_kotlin.$_$.n4;
  var getValue = kotlin_kotlin.$_$.q4;
  var copyOf = kotlin_kotlin.$_$.j4;
  var copyOf_0 = kotlin_kotlin.$_$.k4;
  var fillArrayVal = kotlin_kotlin.$_$.w6;
  var LIST_getInstance = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.i;
  var DeepRecursiveFunction = kotlin_kotlin.$_$.g9;
  var invoke = kotlin_kotlin.$_$.fa;
  var CoroutineImpl = kotlin_kotlin.$_$.i6;
  var DeepRecursiveScope = kotlin_kotlin.$_$.h9;
  var Unit = kotlin_kotlin.$_$.z9;
  var get_COROUTINE_SUSPENDED = kotlin_kotlin.$_$.u5;
  var AbstractPolymorphicSerializer = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.i1;
  var getKClass = kotlin_kotlin.$_$.b;
  var DeserializationStrategy = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.u1;
  var AbstractDecoder = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.f1;
  var ensureNotNull = kotlin_kotlin.$_$.ea;
  var contains_0 = kotlin_kotlin.$_$.i8;
  var plus = kotlin_kotlin.$_$.la;
  var MissingFieldException = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.w1;
  var toLong = kotlin_kotlin.$_$.v7;
  var IllegalArgumentException = kotlin_kotlin.$_$.l9;
  var isFinite = kotlin_kotlin.$_$.ha;
  var isFinite_0 = kotlin_kotlin.$_$.ga;
  var charSequenceGet = kotlin_kotlin.$_$.q6;
  var toUInt = kotlin_kotlin.$_$.z8;
  var _UInt___get_data__impl__f0vqqw = kotlin_kotlin.$_$.c2;
  var toULong = kotlin_kotlin.$_$.a9;
  var _ULong___get_data__impl__fggpzb = kotlin_kotlin.$_$.j2;
  var toUByte = kotlin_kotlin.$_$.y8;
  var _UByte___get_data__impl__jof9qr = kotlin_kotlin.$_$.v1;
  var toUShort = kotlin_kotlin.$_$.b9;
  var _UShort___get_data__impl__g0245 = kotlin_kotlin.$_$.q2;
  var Companion_getInstance = kotlin_kotlin.$_$.l3;
  var serializer_0 = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.t;
  var Companion_getInstance_0 = kotlin_kotlin.$_$.m3;
  var serializer_1 = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.v;
  var Companion_getInstance_1 = kotlin_kotlin.$_$.k3;
  var serializer_2 = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.u;
  var Companion_getInstance_2 = kotlin_kotlin.$_$.n3;
  var serializer_3 = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.s;
  var setOf = kotlin_kotlin.$_$.j5;
  var Char__toInt_impl_vasixd = kotlin_kotlin.$_$.p1;
  var numberToChar = kotlin_kotlin.$_$.n7;
  var equals_0 = kotlin_kotlin.$_$.l8;
  var toString_0 = kotlin_kotlin.$_$.q1;
  var NamedValueDecoder = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.o1;
  var PolymorphicKind = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.w;
  var PrimitiveKind = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.x;
  var MAP_getInstance = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.j;
  var IllegalArgumentException_init_$Create$_0 = kotlin_kotlin.$_$.z;
  var ByteCompanionObject_instance = kotlin_kotlin.$_$.z2;
  var toByte = kotlin_kotlin.$_$.u7;
  var ShortCompanionObject_instance = kotlin_kotlin.$_$.d3;
  var toShort = kotlin_kotlin.$_$.w7;
  var single = kotlin_kotlin.$_$.q8;
  var Char = kotlin_kotlin.$_$.f9;
  var jsonCachedSerialNames = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.r1;
  var emptySet = kotlin_kotlin.$_$.o4;
  var plus_0 = kotlin_kotlin.$_$.f5;
  var toInt = kotlin_kotlin.$_$.x8;
  var toList = kotlin_kotlin.$_$.n5;
  var Enum = kotlin_kotlin.$_$.i9;
  var getContextualDescriptor = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.e1;
  var CONTEXTUAL_getInstance = kotlin_org_jetbrains_kotlinx_kotlinx_serialization_core.$_$.f;
  var noWhenBranchMatchedException = kotlin_kotlin.$_$.ka;
  var last = kotlin_kotlin.$_$.y4;
  var removeLast = kotlin_kotlin.$_$.i5;
  var lastIndexOf = kotlin_kotlin.$_$.p8;
  var Long = kotlin_kotlin.$_$.m9;
  var Char__minus_impl_a2frrh = kotlin_kotlin.$_$.n1;
  var Companion_getInstance_3 = kotlin_kotlin.$_$.i3;
  var numberToLong = kotlin_kotlin.$_$.p7;
  var charArray = kotlin_kotlin.$_$.p6;
  var indexOf = kotlin_kotlin.$_$.m8;
  var HashMap_init_$Create$ = kotlin_kotlin.$_$.j;
  //endregion
  //region block: pre-declaration
  setMetadataFor(Json, 'Json', classMeta);
  setMetadataFor(Default, 'Default', objectMeta, Json);
  setMetadataFor(JsonClassDiscriminator, 'JsonClassDiscriminator', classMeta);
  setMetadataFor(JsonNames, 'JsonNames', classMeta);
  setMetadataFor(JsonConfiguration, 'JsonConfiguration', classMeta);
  setMetadataFor(JsonDecoder, 'JsonDecoder', interfaceMeta, VOID, [Decoder, CompositeDecoder]);
  setMetadataFor(Companion, 'Companion', objectMeta);
  setMetadataFor(JsonElement, 'JsonElement', classMeta, VOID, VOID, VOID, VOID, {0: JsonElementSerializer_getInstance});
  setMetadataFor(JsonObject, 'JsonObject', classMeta, JsonElement, [JsonElement, Map], VOID, VOID, {0: JsonObjectSerializer_getInstance});
  setMetadataFor(Companion_0, 'Companion', objectMeta);
  setMetadataFor(Companion_1, 'Companion', objectMeta);
  setMetadataFor(JsonPrimitive, 'JsonPrimitive', classMeta, JsonElement, VOID, VOID, VOID, {0: JsonPrimitiveSerializer_getInstance});
  setMetadataFor(JsonNull, 'JsonNull', objectMeta, JsonPrimitive, VOID, VOID, VOID, {0: JsonNull_getInstance});
  setMetadataFor(Companion_2, 'Companion', objectMeta);
  setMetadataFor(JsonArray, 'JsonArray', classMeta, JsonElement, [JsonElement, List], VOID, VOID, {0: JsonArraySerializer_getInstance});
  setMetadataFor(JsonLiteral, 'JsonLiteral', classMeta, JsonPrimitive);
  setMetadataFor(JsonObjectBuilder, 'JsonObjectBuilder', classMeta);
  setMetadataFor(JsonArrayBuilder, 'JsonArrayBuilder', classMeta);
  setMetadataFor(JsonObjectDescriptor, 'JsonObjectDescriptor', objectMeta, VOID, [SerialDescriptor]);
  setMetadataFor(JsonObjectSerializer, 'JsonObjectSerializer', objectMeta, VOID, [KSerializer]);
  setMetadataFor(JsonElementSerializer, 'JsonElementSerializer', objectMeta, VOID, [KSerializer]);
  setMetadataFor(JsonPrimitiveSerializer, 'JsonPrimitiveSerializer', objectMeta, VOID, [KSerializer]);
  setMetadataFor(JsonNullSerializer, 'JsonNullSerializer', objectMeta, VOID, [KSerializer]);
  setMetadataFor(JsonArrayDescriptor, 'JsonArrayDescriptor', objectMeta, VOID, [SerialDescriptor]);
  setMetadataFor(JsonArraySerializer, 'JsonArraySerializer', objectMeta, VOID, [KSerializer]);
  setMetadataFor(JsonLiteralSerializer, 'JsonLiteralSerializer', objectMeta, VOID, [KSerializer]);
  setMetadataFor(defer$1, VOID, classMeta, VOID, [SerialDescriptor]);
  setMetadataFor(JsonElementMarker, 'JsonElementMarker', classMeta);
  setMetadataFor(JsonException, 'JsonException', classMeta, SerializationException);
  setMetadataFor(JsonDecodingException, 'JsonDecodingException', classMeta, JsonException);
  setMetadataFor(JsonEncodingException, 'JsonEncodingException', classMeta, JsonException);
  setMetadataFor(Tombstone, 'Tombstone', objectMeta);
  setMetadataFor(JsonPath, 'JsonPath', classMeta, VOID, VOID, JsonPath);
  setMetadataFor(JsonTreeReader$readDeepRecursive$slambda, 'JsonTreeReader$readDeepRecursive$slambda', classMeta, CoroutineImpl, VOID, VOID, VOID, VOID, [2]);
  setMetadataFor($readObjectCOROUTINE$0, '$readObjectCOROUTINE$0', classMeta, CoroutineImpl);
  setMetadataFor(JsonTreeReader, 'JsonTreeReader', classMeta, VOID, VOID, VOID, VOID, VOID, [0]);
  setMetadataFor(Key, 'Key', classMeta, VOID, VOID, Key);
  setMetadataFor(DescriptorSchemaCache, 'DescriptorSchemaCache', classMeta, VOID, VOID, DescriptorSchemaCache);
  setMetadataFor(DiscriminatorHolder, 'DiscriminatorHolder', classMeta);
  setMetadataFor(StreamingJsonDecoder, 'StreamingJsonDecoder', classMeta, AbstractDecoder, [JsonDecoder, AbstractDecoder]);
  setMetadataFor(JsonDecoderForUnsignedTypes, 'JsonDecoderForUnsignedTypes', classMeta, AbstractDecoder);
  setMetadataFor(AbstractJsonTreeDecoder, 'AbstractJsonTreeDecoder', classMeta, NamedValueDecoder, [NamedValueDecoder, JsonDecoder]);
  setMetadataFor(JsonTreeDecoder, 'JsonTreeDecoder', classMeta, AbstractJsonTreeDecoder);
  setMetadataFor(JsonTreeListDecoder, 'JsonTreeListDecoder', classMeta, AbstractJsonTreeDecoder);
  setMetadataFor(JsonPrimitiveDecoder, 'JsonPrimitiveDecoder', classMeta, AbstractJsonTreeDecoder);
  setMetadataFor(JsonTreeMapDecoder, 'JsonTreeMapDecoder', classMeta, JsonTreeDecoder);
  setMetadataFor(WriteMode, 'WriteMode', classMeta, Enum);
  setMetadataFor(AbstractJsonLexer, 'AbstractJsonLexer', classMeta);
  setMetadataFor(CharMappings, 'CharMappings', objectMeta);
  setMetadataFor(StringJsonLexer, 'StringJsonLexer', classMeta, AbstractJsonLexer);
  //endregion
  function Default() {
    Default_instance = this;
    Json.call(this, new JsonConfiguration(), EmptySerializersModule());
  }
  var Default_instance;
  function Default_getInstance() {
    if (Default_instance == null)
      new Default();
    return Default_instance;
  }
  function Json(configuration, serializersModule) {
    Default_getInstance();
    this.pv_1 = configuration;
    this.qv_1 = serializersModule;
    this.rv_1 = new DescriptorSchemaCache();
  }
  protoOf(Json).fl = function () {
    return this.qv_1;
  };
  protoOf(Json).sv = function (deserializer, string) {
    var lexer = new StringJsonLexer(string);
    var input = new StreamingJsonDecoder(this, WriteMode_OBJ_getInstance(), lexer, deserializer.fi(), null);
    var result = input.qk(deserializer);
    lexer.gw();
    return result;
  };
  protoOf(Json).tv = function (string) {
    return this.sv(JsonElementSerializer_getInstance(), string);
  };
  function JsonClassDiscriminator() {
  }
  function JsonNames() {
  }
  function JsonConfiguration(encodeDefaults, ignoreUnknownKeys, isLenient, allowStructuredMapKeys, prettyPrint, explicitNulls, prettyPrintIndent, coerceInputValues, useArrayPolymorphism, classDiscriminator, allowSpecialFloatingPointValues, useAlternativeNames, namingStrategy, decodeEnumsCaseInsensitive) {
    encodeDefaults = encodeDefaults === VOID ? false : encodeDefaults;
    ignoreUnknownKeys = ignoreUnknownKeys === VOID ? false : ignoreUnknownKeys;
    isLenient = isLenient === VOID ? false : isLenient;
    allowStructuredMapKeys = allowStructuredMapKeys === VOID ? false : allowStructuredMapKeys;
    prettyPrint = prettyPrint === VOID ? false : prettyPrint;
    explicitNulls = explicitNulls === VOID ? true : explicitNulls;
    prettyPrintIndent = prettyPrintIndent === VOID ? '    ' : prettyPrintIndent;
    coerceInputValues = coerceInputValues === VOID ? false : coerceInputValues;
    useArrayPolymorphism = useArrayPolymorphism === VOID ? false : useArrayPolymorphism;
    classDiscriminator = classDiscriminator === VOID ? 'type' : classDiscriminator;
    allowSpecialFloatingPointValues = allowSpecialFloatingPointValues === VOID ? false : allowSpecialFloatingPointValues;
    useAlternativeNames = useAlternativeNames === VOID ? true : useAlternativeNames;
    namingStrategy = namingStrategy === VOID ? null : namingStrategy;
    decodeEnumsCaseInsensitive = decodeEnumsCaseInsensitive === VOID ? false : decodeEnumsCaseInsensitive;
    this.hw_1 = encodeDefaults;
    this.iw_1 = ignoreUnknownKeys;
    this.jw_1 = isLenient;
    this.kw_1 = allowStructuredMapKeys;
    this.lw_1 = prettyPrint;
    this.mw_1 = explicitNulls;
    this.nw_1 = prettyPrintIndent;
    this.ow_1 = coerceInputValues;
    this.pw_1 = useArrayPolymorphism;
    this.qw_1 = classDiscriminator;
    this.rw_1 = allowSpecialFloatingPointValues;
    this.sw_1 = useAlternativeNames;
    this.tw_1 = namingStrategy;
    this.uw_1 = decodeEnumsCaseInsensitive;
  }
  protoOf(JsonConfiguration).toString = function () {
    return 'JsonConfiguration(encodeDefaults=' + this.hw_1 + ', ignoreUnknownKeys=' + this.iw_1 + ', isLenient=' + this.jw_1 + ', ' + ('allowStructuredMapKeys=' + this.kw_1 + ', prettyPrint=' + this.lw_1 + ', explicitNulls=' + this.mw_1 + ', ') + ("prettyPrintIndent='" + this.nw_1 + "', coerceInputValues=" + this.ow_1 + ', useArrayPolymorphism=' + this.pw_1 + ', ') + ("classDiscriminator='" + this.qw_1 + "', allowSpecialFloatingPointValues=" + this.rw_1 + ', useAlternativeNames=' + this.sw_1 + ', ') + ('namingStrategy=' + this.tw_1 + ', decodeEnumsCaseInsensitive=' + this.uw_1 + ')');
  };
  function JsonDecoder() {
  }
  var jsonUnquotedLiteralDescriptor;
  function Companion() {
  }
  var Companion_instance_0;
  function Companion_getInstance_4() {
    return Companion_instance_0;
  }
  function JsonObject$toString$lambda(_name_for_destructuring_parameter_0__wldtmu) {
    // Inline function 'kotlin.collections.component1' call
    var k = _name_for_destructuring_parameter_0__wldtmu.u1();
    // Inline function 'kotlin.collections.component2' call
    var v = _name_for_destructuring_parameter_0__wldtmu.v1();
    // Inline function 'kotlin.text.buildString' call
    // Inline function 'kotlin.contracts.contract' call
    // Inline function 'kotlin.apply' call
    var this_0 = StringBuilder_init_$Create$();
    // Inline function 'kotlin.contracts.contract' call
    // Inline function 'kotlinx.serialization.json.JsonObject.toString.<anonymous>.<anonymous>' call
    printQuoted(this_0, k);
    this_0.a5(_Char___init__impl__6a9atx(58));
    this_0.y4(v);
    return this_0.toString();
  }
  function JsonObject(content) {
    JsonElement.call(this);
    this.xw_1 = content;
  }
  protoOf(JsonObject).t1 = function () {
    return this.xw_1.t1();
  };
  protoOf(JsonObject).p1 = function () {
    return this.xw_1.p1();
  };
  protoOf(JsonObject).u = function () {
    return this.xw_1.u();
  };
  protoOf(JsonObject).q1 = function () {
    return this.xw_1.q1();
  };
  protoOf(JsonObject).yw = function (key) {
    return this.xw_1.y1(key);
  };
  protoOf(JsonObject).y1 = function (key) {
    if (!(!(key == null) ? typeof key === 'string' : false))
      return false;
    return this.yw((!(key == null) ? typeof key === 'string' : false) ? key : THROW_CCE());
  };
  protoOf(JsonObject).e8 = function (key) {
    return this.xw_1.b2(key);
  };
  protoOf(JsonObject).b2 = function (key) {
    if (!(!(key == null) ? typeof key === 'string' : false))
      return null;
    return this.e8((!(key == null) ? typeof key === 'string' : false) ? key : THROW_CCE());
  };
  protoOf(JsonObject).x = function () {
    return this.xw_1.x();
  };
  protoOf(JsonObject).equals = function (other) {
    return equals(this.xw_1, other);
  };
  protoOf(JsonObject).hashCode = function () {
    return hashCode(this.xw_1);
  };
  protoOf(JsonObject).toString = function () {
    var tmp = this.xw_1.t1();
    return joinToString(tmp, ',', '{', '}', VOID, VOID, JsonObject$toString$lambda);
  };
  function Companion_0() {
  }
  var Companion_instance_1;
  function Companion_getInstance_5() {
    return Companion_instance_1;
  }
  function JsonElement() {
  }
  function Companion_1() {
  }
  var Companion_instance_2;
  function Companion_getInstance_6() {
    return Companion_instance_2;
  }
  function JsonPrimitive() {
    JsonElement.call(this);
  }
  protoOf(JsonPrimitive).toString = function () {
    return this.zw();
  };
  function get_jsonObject(_this__u8e3s4) {
    _init_properties_JsonElement_kt__7cbdc2();
    var tmp0_elvis_lhs = _this__u8e3s4 instanceof JsonObject ? _this__u8e3s4 : null;
    var tmp;
    if (tmp0_elvis_lhs == null) {
      error(_this__u8e3s4, 'JsonObject');
    } else {
      tmp = tmp0_elvis_lhs;
    }
    return tmp;
  }
  function get_jsonPrimitive(_this__u8e3s4) {
    _init_properties_JsonElement_kt__7cbdc2();
    var tmp0_elvis_lhs = _this__u8e3s4 instanceof JsonPrimitive ? _this__u8e3s4 : null;
    var tmp;
    if (tmp0_elvis_lhs == null) {
      error(_this__u8e3s4, 'JsonPrimitive');
    } else {
      tmp = tmp0_elvis_lhs;
    }
    return tmp;
  }
  function get_boolean(_this__u8e3s4) {
    _init_properties_JsonElement_kt__7cbdc2();
    var tmp0_elvis_lhs = toBooleanStrictOrNull(_this__u8e3s4.zw());
    var tmp;
    if (tmp0_elvis_lhs == null) {
      throw IllegalStateException_init_$Create$('' + _this__u8e3s4 + ' does not represent a Boolean');
    } else {
      tmp = tmp0_elvis_lhs;
    }
    return tmp;
  }
  function JsonNull$_anonymous__enib48() {
    return JsonNullSerializer_getInstance();
  }
  function JsonNull() {
    JsonNull_instance = this;
    JsonPrimitive.call(this);
    this.ax_1 = 'null';
    var tmp = this;
    var tmp_0 = LazyThreadSafetyMode_PUBLICATION_getInstance();
    tmp.bx_1 = lazy(tmp_0, JsonNull$_anonymous__enib48);
  }
  protoOf(JsonNull).zw = function () {
    return this.ax_1;
  };
  var JsonNull_instance;
  function JsonNull_getInstance() {
    if (JsonNull_instance == null)
      new JsonNull();
    return JsonNull_instance;
  }
  function get_contentOrNull(_this__u8e3s4) {
    _init_properties_JsonElement_kt__7cbdc2();
    var tmp;
    if (_this__u8e3s4 instanceof JsonNull) {
      tmp = null;
    } else {
      tmp = _this__u8e3s4.zw();
    }
    return tmp;
  }
  function Companion_2() {
  }
  var Companion_instance_3;
  function Companion_getInstance_7() {
    return Companion_instance_3;
  }
  function JsonArray(content) {
    JsonElement.call(this);
    this.cx_1 = content;
  }
  protoOf(JsonArray).u = function () {
    return this.cx_1.u();
  };
  protoOf(JsonArray).dx = function (elements) {
    return this.cx_1.w(elements);
  };
  protoOf(JsonArray).w = function (elements) {
    return this.dx(elements);
  };
  protoOf(JsonArray).b1 = function (index) {
    return this.cx_1.b1(index);
  };
  protoOf(JsonArray).x = function () {
    return this.cx_1.x();
  };
  protoOf(JsonArray).r = function () {
    return this.cx_1.r();
  };
  protoOf(JsonArray).equals = function (other) {
    return equals(this.cx_1, other);
  };
  protoOf(JsonArray).hashCode = function () {
    return hashCode(this.cx_1);
  };
  protoOf(JsonArray).toString = function () {
    return joinToString(this.cx_1, ',', '[', ']');
  };
  function get_jsonArray(_this__u8e3s4) {
    _init_properties_JsonElement_kt__7cbdc2();
    var tmp0_elvis_lhs = _this__u8e3s4 instanceof JsonArray ? _this__u8e3s4 : null;
    var tmp;
    if (tmp0_elvis_lhs == null) {
      error(_this__u8e3s4, 'JsonArray');
    } else {
      tmp = tmp0_elvis_lhs;
    }
    return tmp;
  }
  function get_int(_this__u8e3s4) {
    _init_properties_JsonElement_kt__7cbdc2();
    // Inline function 'kotlinx.serialization.json.mapExceptions' call
    var tmp;
    try {
      // Inline function 'kotlinx.serialization.json.<get-int>.<anonymous>' call
      tmp = (new StringJsonLexer(_this__u8e3s4.zw())).ex();
    } catch ($p) {
      var tmp_0;
      if ($p instanceof JsonDecodingException) {
        var e = $p;
        throw NumberFormatException_init_$Create$(e.message);
      } else {
        throw $p;
      }
    }
    var result = tmp;
    // Inline function 'kotlin.ranges.contains' call
    var this_0 = numberRangeToNumber(IntCompanionObject_instance.MIN_VALUE, IntCompanionObject_instance.MAX_VALUE);
    if (!contains(isInterface(this_0, ClosedRange) ? this_0 : THROW_CCE(), result))
      throw NumberFormatException_init_$Create$(_this__u8e3s4.zw() + ' is not an Int');
    return result.f9();
  }
  function error(_this__u8e3s4, element) {
    _init_properties_JsonElement_kt__7cbdc2();
    throw IllegalArgumentException_init_$Create$('Element ' + getKClassFromExpression(_this__u8e3s4) + ' is not a ' + element);
  }
  function JsonPrimitive_0(value) {
    _init_properties_JsonElement_kt__7cbdc2();
    if (value == null)
      return JsonNull_getInstance();
    return new JsonLiteral(value, false);
  }
  function JsonPrimitive_1(value) {
    _init_properties_JsonElement_kt__7cbdc2();
    if (value == null)
      return JsonNull_getInstance();
    return new JsonLiteral(value, true);
  }
  function JsonPrimitive_2(value) {
    _init_properties_JsonElement_kt__7cbdc2();
    if (value == null)
      return JsonNull_getInstance();
    return new JsonLiteral(value, false);
  }
  function JsonLiteral(body, isString, coerceToInlineType) {
    coerceToInlineType = coerceToInlineType === VOID ? null : coerceToInlineType;
    JsonPrimitive.call(this);
    this.fx_1 = isString;
    this.gx_1 = coerceToInlineType;
    this.hx_1 = toString(body);
    if (!(this.gx_1 == null)) {
      // Inline function 'kotlin.require' call
      // Inline function 'kotlin.contracts.contract' call
      // Inline function 'kotlin.require' call
      // Inline function 'kotlin.contracts.contract' call
      if (!this.gx_1.qi()) {
        // Inline function 'kotlin.require.<anonymous>' call
        var message = 'Failed requirement.';
        throw IllegalArgumentException_init_$Create$(toString(message));
      }
    }
  }
  protoOf(JsonLiteral).zw = function () {
    return this.hx_1;
  };
  protoOf(JsonLiteral).toString = function () {
    var tmp;
    if (this.fx_1) {
      // Inline function 'kotlin.text.buildString' call
      // Inline function 'kotlin.contracts.contract' call
      // Inline function 'kotlin.apply' call
      var this_0 = StringBuilder_init_$Create$();
      // Inline function 'kotlin.contracts.contract' call
      // Inline function 'kotlinx.serialization.json.JsonLiteral.toString.<anonymous>' call
      printQuoted(this_0, this.hx_1);
      tmp = this_0.toString();
    } else {
      tmp = this.hx_1;
    }
    return tmp;
  };
  protoOf(JsonLiteral).equals = function (other) {
    if (this === other)
      return true;
    if (other == null ? true : !getKClassFromExpression(this).equals(getKClassFromExpression(other)))
      return false;
    if (!(other instanceof JsonLiteral))
      THROW_CCE();
    if (!(this.fx_1 === other.fx_1))
      return false;
    if (!(this.hx_1 === other.hx_1))
      return false;
    return true;
  };
  protoOf(JsonLiteral).hashCode = function () {
    var result = getBooleanHashCode(this.fx_1);
    result = imul(31, result) + getStringHashCode(this.hx_1) | 0;
    return result;
  };
  function get_booleanOrNull(_this__u8e3s4) {
    _init_properties_JsonElement_kt__7cbdc2();
    return toBooleanStrictOrNull(_this__u8e3s4.zw());
  }
  function get_long(_this__u8e3s4) {
    _init_properties_JsonElement_kt__7cbdc2();
    // Inline function 'kotlinx.serialization.json.mapExceptions' call
    var tmp;
    try {
      // Inline function 'kotlinx.serialization.json.<get-long>.<anonymous>' call
      tmp = (new StringJsonLexer(_this__u8e3s4.zw())).ex();
    } catch ($p) {
      var tmp_0;
      if ($p instanceof JsonDecodingException) {
        var e = $p;
        throw NumberFormatException_init_$Create$(e.message);
      } else {
        throw $p;
      }
    }
    return tmp;
  }
  function get_float(_this__u8e3s4) {
    _init_properties_JsonElement_kt__7cbdc2();
    // Inline function 'kotlin.text.toFloat' call
    var this_0 = _this__u8e3s4.zw();
    // Inline function 'kotlin.js.unsafeCast' call
    // Inline function 'kotlin.js.asDynamic' call
    return toDouble(this_0);
  }
  function get_double(_this__u8e3s4) {
    _init_properties_JsonElement_kt__7cbdc2();
    return toDouble(_this__u8e3s4.zw());
  }
  var properties_initialized_JsonElement_kt_abxy8s;
  function _init_properties_JsonElement_kt__7cbdc2() {
    if (!properties_initialized_JsonElement_kt_abxy8s) {
      properties_initialized_JsonElement_kt_abxy8s = true;
      jsonUnquotedLiteralDescriptor = InlinePrimitiveDescriptor('kotlinx.serialization.json.JsonUnquotedLiteral', serializer(StringCompanionObject_instance));
    }
  }
  function JsonObjectBuilder() {
    var tmp = this;
    // Inline function 'kotlin.collections.linkedMapOf' call
    tmp.ix_1 = LinkedHashMap_init_$Create$();
  }
  protoOf(JsonObjectBuilder).jx = function (key, element) {
    return this.ix_1.r1(key, element);
  };
  protoOf(JsonObjectBuilder).kn = function () {
    return new JsonObject(this.ix_1);
  };
  function put(_this__u8e3s4, key, value) {
    return _this__u8e3s4.jx(key, JsonPrimitive_0(value));
  }
  function put_0(_this__u8e3s4, key, value) {
    return _this__u8e3s4.jx(key, JsonPrimitive_1(value));
  }
  function put_1(_this__u8e3s4, key, value) {
    return _this__u8e3s4.jx(key, JsonPrimitive_2(value));
  }
  function JsonArrayBuilder() {
    var tmp = this;
    // Inline function 'kotlin.collections.mutableListOf' call
    tmp.kx_1 = ArrayList_init_$Create$();
  }
  protoOf(JsonArrayBuilder).lx = function (element) {
    // Inline function 'kotlin.collections.plusAssign' call
    this.kx_1.o(element);
    return true;
  };
  protoOf(JsonArrayBuilder).kn = function () {
    return new JsonArray(this.kx_1);
  };
  function add(_this__u8e3s4, value) {
    return _this__u8e3s4.lx(JsonPrimitive_1(value));
  }
  function JsonObjectDescriptor() {
    JsonObjectDescriptor_instance = this;
    this.mx_1 = MapSerializer(serializer(StringCompanionObject_instance), JsonElementSerializer_getInstance()).fi();
    this.nx_1 = 'kotlinx.serialization.json.JsonObject';
  }
  protoOf(JsonObjectDescriptor).si = function () {
    return this.mx_1.si();
  };
  protoOf(JsonObjectDescriptor).ri = function () {
    return this.mx_1.ri();
  };
  protoOf(JsonObjectDescriptor).qi = function () {
    return this.mx_1.qi();
  };
  protoOf(JsonObjectDescriptor).ii = function () {
    return this.mx_1.ii();
  };
  protoOf(JsonObjectDescriptor).pi = function () {
    return this.mx_1.pi();
  };
  protoOf(JsonObjectDescriptor).vi = function (index) {
    return this.mx_1.vi(index);
  };
  protoOf(JsonObjectDescriptor).wi = function (index) {
    return this.mx_1.wi(index);
  };
  protoOf(JsonObjectDescriptor).ui = function (name) {
    return this.mx_1.ui(name);
  };
  protoOf(JsonObjectDescriptor).ti = function (index) {
    return this.mx_1.ti(index);
  };
  protoOf(JsonObjectDescriptor).xi = function (index) {
    return this.mx_1.xi(index);
  };
  protoOf(JsonObjectDescriptor).oi = function () {
    return this.nx_1;
  };
  var JsonObjectDescriptor_instance;
  function JsonObjectDescriptor_getInstance() {
    if (JsonObjectDescriptor_instance == null)
      new JsonObjectDescriptor();
    return JsonObjectDescriptor_instance;
  }
  function JsonObjectSerializer() {
    JsonObjectSerializer_instance = this;
    this.ox_1 = JsonObjectDescriptor_getInstance();
  }
  protoOf(JsonObjectSerializer).fi = function () {
    return this.ox_1;
  };
  protoOf(JsonObjectSerializer).gi = function (decoder) {
    verify(decoder);
    return new JsonObject(MapSerializer(serializer(StringCompanionObject_instance), JsonElementSerializer_getInstance()).gi(decoder));
  };
  var JsonObjectSerializer_instance;
  function JsonObjectSerializer_getInstance() {
    if (JsonObjectSerializer_instance == null)
      new JsonObjectSerializer();
    return JsonObjectSerializer_instance;
  }
  function JsonElementSerializer$descriptor$lambda($this$buildSerialDescriptor) {
    $this$buildSerialDescriptor.nj('JsonPrimitive', defer(JsonElementSerializer$descriptor$lambda$lambda));
    $this$buildSerialDescriptor.nj('JsonNull', defer(JsonElementSerializer$descriptor$lambda$lambda_0));
    $this$buildSerialDescriptor.nj('JsonLiteral', defer(JsonElementSerializer$descriptor$lambda$lambda_1));
    $this$buildSerialDescriptor.nj('JsonObject', defer(JsonElementSerializer$descriptor$lambda$lambda_2));
    $this$buildSerialDescriptor.nj('JsonArray', defer(JsonElementSerializer$descriptor$lambda$lambda_3));
    return Unit_instance;
  }
  function JsonElementSerializer$descriptor$lambda$lambda() {
    return JsonPrimitiveSerializer_getInstance().px_1;
  }
  function JsonElementSerializer$descriptor$lambda$lambda_0() {
    return JsonNullSerializer_getInstance().qx_1;
  }
  function JsonElementSerializer$descriptor$lambda$lambda_1() {
    return JsonLiteralSerializer_getInstance().rx_1;
  }
  function JsonElementSerializer$descriptor$lambda$lambda_2() {
    return JsonObjectSerializer_getInstance().ox_1;
  }
  function JsonElementSerializer$descriptor$lambda$lambda_3() {
    return JsonArraySerializer_getInstance().sx_1;
  }
  function JsonElementSerializer() {
    JsonElementSerializer_instance = this;
    var tmp = this;
    var tmp_0 = SEALED_getInstance();
    tmp.tx_1 = buildSerialDescriptor('kotlinx.serialization.json.JsonElement', tmp_0, [], JsonElementSerializer$descriptor$lambda);
  }
  protoOf(JsonElementSerializer).fi = function () {
    return this.tx_1;
  };
  protoOf(JsonElementSerializer).gi = function (decoder) {
    var input = asJsonDecoder(decoder);
    return input.ww();
  };
  var JsonElementSerializer_instance;
  function JsonElementSerializer_getInstance() {
    if (JsonElementSerializer_instance == null)
      new JsonElementSerializer();
    return JsonElementSerializer_instance;
  }
  function JsonPrimitiveSerializer() {
    JsonPrimitiveSerializer_instance = this;
    this.px_1 = buildSerialDescriptor('kotlinx.serialization.json.JsonPrimitive', STRING_getInstance(), []);
  }
  protoOf(JsonPrimitiveSerializer).fi = function () {
    return this.px_1;
  };
  protoOf(JsonPrimitiveSerializer).gi = function (decoder) {
    var result = asJsonDecoder(decoder).ww();
    if (!(result instanceof JsonPrimitive))
      throw JsonDecodingException_0(-1, 'Unexpected JSON element, expected JsonPrimitive, had ' + getKClassFromExpression(result), toString(result));
    return result;
  };
  var JsonPrimitiveSerializer_instance;
  function JsonPrimitiveSerializer_getInstance() {
    if (JsonPrimitiveSerializer_instance == null)
      new JsonPrimitiveSerializer();
    return JsonPrimitiveSerializer_instance;
  }
  function JsonNullSerializer() {
    JsonNullSerializer_instance = this;
    this.qx_1 = buildSerialDescriptor('kotlinx.serialization.json.JsonNull', ENUM_getInstance(), []);
  }
  protoOf(JsonNullSerializer).fi = function () {
    return this.qx_1;
  };
  protoOf(JsonNullSerializer).gi = function (decoder) {
    verify(decoder);
    if (decoder.ck()) {
      throw new JsonDecodingException("Expected 'null' literal");
    }
    decoder.dk();
    return JsonNull_getInstance();
  };
  var JsonNullSerializer_instance;
  function JsonNullSerializer_getInstance() {
    if (JsonNullSerializer_instance == null)
      new JsonNullSerializer();
    return JsonNullSerializer_instance;
  }
  function JsonArrayDescriptor() {
    JsonArrayDescriptor_instance = this;
    this.ux_1 = ListSerializer(JsonElementSerializer_getInstance()).fi();
    this.vx_1 = 'kotlinx.serialization.json.JsonArray';
  }
  protoOf(JsonArrayDescriptor).si = function () {
    return this.ux_1.si();
  };
  protoOf(JsonArrayDescriptor).ri = function () {
    return this.ux_1.ri();
  };
  protoOf(JsonArrayDescriptor).qi = function () {
    return this.ux_1.qi();
  };
  protoOf(JsonArrayDescriptor).ii = function () {
    return this.ux_1.ii();
  };
  protoOf(JsonArrayDescriptor).pi = function () {
    return this.ux_1.pi();
  };
  protoOf(JsonArrayDescriptor).vi = function (index) {
    return this.ux_1.vi(index);
  };
  protoOf(JsonArrayDescriptor).wi = function (index) {
    return this.ux_1.wi(index);
  };
  protoOf(JsonArrayDescriptor).ui = function (name) {
    return this.ux_1.ui(name);
  };
  protoOf(JsonArrayDescriptor).ti = function (index) {
    return this.ux_1.ti(index);
  };
  protoOf(JsonArrayDescriptor).xi = function (index) {
    return this.ux_1.xi(index);
  };
  protoOf(JsonArrayDescriptor).oi = function () {
    return this.vx_1;
  };
  var JsonArrayDescriptor_instance;
  function JsonArrayDescriptor_getInstance() {
    if (JsonArrayDescriptor_instance == null)
      new JsonArrayDescriptor();
    return JsonArrayDescriptor_instance;
  }
  function JsonArraySerializer() {
    JsonArraySerializer_instance = this;
    this.sx_1 = JsonArrayDescriptor_getInstance();
  }
  protoOf(JsonArraySerializer).fi = function () {
    return this.sx_1;
  };
  protoOf(JsonArraySerializer).gi = function (decoder) {
    verify(decoder);
    return new JsonArray(ListSerializer(JsonElementSerializer_getInstance()).gi(decoder));
  };
  var JsonArraySerializer_instance;
  function JsonArraySerializer_getInstance() {
    if (JsonArraySerializer_instance == null)
      new JsonArraySerializer();
    return JsonArraySerializer_instance;
  }
  function verify(decoder) {
    asJsonDecoder(decoder);
  }
  function defer(deferred) {
    return new defer$1(deferred);
  }
  function JsonLiteralSerializer() {
    JsonLiteralSerializer_instance = this;
    this.rx_1 = PrimitiveSerialDescriptor('kotlinx.serialization.json.JsonLiteral', STRING_getInstance());
  }
  protoOf(JsonLiteralSerializer).fi = function () {
    return this.rx_1;
  };
  protoOf(JsonLiteralSerializer).gi = function (decoder) {
    var result = asJsonDecoder(decoder).ww();
    if (!(result instanceof JsonLiteral))
      throw JsonDecodingException_0(-1, 'Unexpected JSON element, expected JsonLiteral, had ' + getKClassFromExpression(result), toString(result));
    return result;
  };
  var JsonLiteralSerializer_instance;
  function JsonLiteralSerializer_getInstance() {
    if (JsonLiteralSerializer_instance == null)
      new JsonLiteralSerializer();
    return JsonLiteralSerializer_instance;
  }
  function asJsonDecoder(_this__u8e3s4) {
    var tmp0_elvis_lhs = isInterface(_this__u8e3s4, JsonDecoder) ? _this__u8e3s4 : null;
    var tmp;
    if (tmp0_elvis_lhs == null) {
      throw IllegalStateException_init_$Create$('This serializer can be used only with Json format.' + ('Expected Decoder to be JsonDecoder, got ' + getKClassFromExpression(_this__u8e3s4)));
    } else {
      tmp = tmp0_elvis_lhs;
    }
    return tmp;
  }
  function _get_original__l7ku1m($this) {
    // Inline function 'kotlin.getValue' call
    var this_0 = $this.wx_1;
    original$factory();
    return this_0.v1();
  }
  function defer$1($deferred) {
    this.wx_1 = lazy_0($deferred);
  }
  protoOf(defer$1).oi = function () {
    return _get_original__l7ku1m(this).oi();
  };
  protoOf(defer$1).pi = function () {
    return _get_original__l7ku1m(this).pi();
  };
  protoOf(defer$1).ri = function () {
    return _get_original__l7ku1m(this).ri();
  };
  protoOf(defer$1).ti = function (index) {
    return _get_original__l7ku1m(this).ti(index);
  };
  protoOf(defer$1).ui = function (name) {
    return _get_original__l7ku1m(this).ui(name);
  };
  protoOf(defer$1).vi = function (index) {
    return _get_original__l7ku1m(this).vi(index);
  };
  protoOf(defer$1).wi = function (index) {
    return _get_original__l7ku1m(this).wi(index);
  };
  protoOf(defer$1).xi = function (index) {
    return _get_original__l7ku1m(this).xi(index);
  };
  function original$factory() {
    return getPropertyCallableRef('original', 1, KProperty1, function (receiver) {
      return _get_original__l7ku1m(receiver);
    }, null);
  }
  function readIfAbsent($this, descriptor, index) {
    $this.yx_1 = !descriptor.xi(index) ? descriptor.wi(index).ii() : false;
    return $this.yx_1;
  }
  function JsonElementMarker$readIfAbsent$ref($boundThis) {
    var l = function (p0, p1) {
      return readIfAbsent($boundThis, p0, p1);
    };
    l.callableName = 'readIfAbsent';
    return l;
  }
  function JsonElementMarker(descriptor) {
    var tmp = this;
    tmp.xx_1 = new ElementMarker(descriptor, JsonElementMarker$readIfAbsent$ref(this));
    this.yx_1 = false;
  }
  protoOf(JsonElementMarker).vn = function (index) {
    this.xx_1.vn(index);
  };
  protoOf(JsonElementMarker).wn = function () {
    return this.xx_1.wn();
  };
  function JsonDecodingException(message) {
    JsonException.call(this, message);
    captureStack(this, JsonDecodingException);
  }
  function JsonDecodingException_0(offset, message, input) {
    return JsonDecodingException_1(offset, message + '\nJSON input: ' + minify(input, offset));
  }
  function JsonException(message) {
    SerializationException_init_$Init$(message, this);
    captureStack(this, JsonException);
  }
  function JsonDecodingException_1(offset, message) {
    return new JsonDecodingException(offset >= 0 ? 'Unexpected JSON token at offset ' + offset + ': ' + message : message);
  }
  function minify(_this__u8e3s4, offset) {
    offset = offset === VOID ? -1 : offset;
    if (charSequenceLength(_this__u8e3s4) < 200)
      return _this__u8e3s4;
    if (offset === -1) {
      var start = charSequenceLength(_this__u8e3s4) - 60 | 0;
      if (start <= 0)
        return _this__u8e3s4;
      // Inline function 'kotlin.text.substring' call
      var endIndex = charSequenceLength(_this__u8e3s4);
      return '.....' + toString(charSequenceSubSequence(_this__u8e3s4, start, endIndex));
    }
    var start_0 = offset - 30 | 0;
    var end = offset + 30 | 0;
    var prefix = start_0 <= 0 ? '' : '.....';
    var suffix = end >= charSequenceLength(_this__u8e3s4) ? '' : '.....';
    // Inline function 'kotlin.text.substring' call
    var startIndex = coerceAtLeast(start_0, 0);
    var endIndex_0 = coerceAtMost(end, charSequenceLength(_this__u8e3s4));
    return prefix + toString(charSequenceSubSequence(_this__u8e3s4, startIndex, endIndex_0)) + suffix;
  }
  function throwInvalidFloatingPointDecoded(_this__u8e3s4, result) {
    _this__u8e3s4.zx('Unexpected special floating-point value ' + toString(result) + '. By default, ' + 'non-finite floating point values are prohibited because they do not conform JSON specification', VOID, get_specialFlowingValuesHint());
  }
  function JsonEncodingException(message) {
    JsonException.call(this, message);
    captureStack(this, JsonEncodingException);
  }
  function InvalidKeyKindException(keyDescriptor) {
    return new JsonEncodingException("Value of type '" + keyDescriptor.oi() + "' can't be used in JSON as a key in the map. " + ("It should have either primitive or enum kind, but its kind is '" + keyDescriptor.pi() + "'.\n") + get_allowStructuredMapKeysHint());
  }
  function InvalidFloatingPointDecoded(value, key, output) {
    return JsonDecodingException_1(-1, unexpectedFpErrorMessage(value, key, output));
  }
  function UnknownKeyException(key, input) {
    return JsonDecodingException_1(-1, "Encountered an unknown key '" + key + "'.\n" + (get_ignoreUnknownKeysHint() + '\n') + ('Current input: ' + minify(input)));
  }
  function unexpectedFpErrorMessage(value, key, output) {
    return 'Unexpected special floating-point value ' + toString(value) + ' with key ' + key + '. By default, ' + 'non-finite floating point values are prohibited because they do not conform JSON specification. ' + (get_specialFlowingValuesHint() + '\n') + ('Current output: ' + minify(output));
  }
  function get_JsonDeserializationNamesKey() {
    _init_properties_JsonNamesMap_kt__cbbp0k();
    return JsonDeserializationNamesKey;
  }
  var JsonDeserializationNamesKey;
  var JsonSerializationNamesKey;
  function getJsonNameIndex(_this__u8e3s4, json, name) {
    _init_properties_JsonNamesMap_kt__cbbp0k();
    if (decodeCaseInsensitive(json, _this__u8e3s4)) {
      // Inline function 'kotlin.text.lowercase' call
      // Inline function 'kotlin.js.asDynamic' call
      var tmp$ret$1 = name.toLowerCase();
      return getJsonNameIndexSlowPath(_this__u8e3s4, json, tmp$ret$1);
    }
    var strategy = namingStrategy(_this__u8e3s4, json);
    if (!(strategy == null))
      return getJsonNameIndexSlowPath(_this__u8e3s4, json, name);
    var index = _this__u8e3s4.ui(name);
    if (!(index === -3))
      return index;
    if (!json.pv_1.sw_1)
      return index;
    return getJsonNameIndexSlowPath(_this__u8e3s4, json, name);
  }
  function getJsonNameIndexOrThrow(_this__u8e3s4, json, name, suffix) {
    suffix = suffix === VOID ? '' : suffix;
    _init_properties_JsonNamesMap_kt__cbbp0k();
    var index = getJsonNameIndex(_this__u8e3s4, json, name);
    if (index === -3)
      throw SerializationException_init_$Create$(_this__u8e3s4.oi() + " does not contain element with name '" + name + "'" + suffix);
    return index;
  }
  function namingStrategy(_this__u8e3s4, json) {
    _init_properties_JsonNamesMap_kt__cbbp0k();
    return equals(_this__u8e3s4.pi(), CLASS_getInstance()) ? json.pv_1.tw_1 : null;
  }
  function deserializationNamesMap(_this__u8e3s4, descriptor) {
    _init_properties_JsonNamesMap_kt__cbbp0k();
    var tmp = get_schemaCache(_this__u8e3s4);
    var tmp_0 = get_JsonDeserializationNamesKey();
    return tmp.by(descriptor, tmp_0, deserializationNamesMap$lambda(descriptor, _this__u8e3s4));
  }
  function decodeCaseInsensitive(_this__u8e3s4, descriptor) {
    _init_properties_JsonNamesMap_kt__cbbp0k();
    return _this__u8e3s4.pv_1.uw_1 ? equals(descriptor.pi(), ENUM_getInstance()) : false;
  }
  function getJsonNameIndexSlowPath(_this__u8e3s4, json, name) {
    _init_properties_JsonNamesMap_kt__cbbp0k();
    var tmp0_elvis_lhs = deserializationNamesMap(json, _this__u8e3s4).b2(name);
    var tmp;
    if (tmp0_elvis_lhs == null) {
      tmp = -3;
    } else {
      tmp = tmp0_elvis_lhs;
    }
    return tmp;
  }
  function buildDeserializationNamesMap(_this__u8e3s4, json) {
    _init_properties_JsonNamesMap_kt__cbbp0k();
    // Inline function 'kotlin.collections.mutableMapOf' call
    var builder = LinkedHashMap_init_$Create$();
    var useLowercaseEnums = decodeCaseInsensitive(json, _this__u8e3s4);
    var strategyForClasses = namingStrategy(_this__u8e3s4, json);
    var inductionVariable = 0;
    var last = _this__u8e3s4.ri();
    if (inductionVariable < last)
      do {
        var i = inductionVariable;
        inductionVariable = inductionVariable + 1 | 0;
        // Inline function 'kotlin.collections.filterIsInstance' call
        // Inline function 'kotlin.collections.filterIsInstanceTo' call
        var this_0 = _this__u8e3s4.vi(i);
        var destination = ArrayList_init_$Create$();
        var tmp0_iterator = this_0.r();
        while (tmp0_iterator.s()) {
          var element = tmp0_iterator.t();
          if (element instanceof JsonNames) {
            destination.o(element);
          }
        }
        var tmp1_safe_receiver = singleOrNull(destination);
        var tmp2_safe_receiver = tmp1_safe_receiver == null ? null : tmp1_safe_receiver.cy_1;
        if (tmp2_safe_receiver == null)
          null;
        else {
          // Inline function 'kotlin.collections.forEach' call
          var inductionVariable_0 = 0;
          var last_0 = tmp2_safe_receiver.length;
          while (inductionVariable_0 < last_0) {
            var element_0 = tmp2_safe_receiver[inductionVariable_0];
            inductionVariable_0 = inductionVariable_0 + 1 | 0;
            // Inline function 'kotlinx.serialization.json.internal.buildDeserializationNamesMap.<anonymous>' call
            var tmp;
            if (useLowercaseEnums) {
              // Inline function 'kotlin.text.lowercase' call
              // Inline function 'kotlin.js.asDynamic' call
              tmp = element_0.toLowerCase();
            } else {
              tmp = element_0;
            }
            buildDeserializationNamesMap$putOrThrow(builder, _this__u8e3s4, tmp, i);
          }
        }
        var tmp_0;
        if (useLowercaseEnums) {
          // Inline function 'kotlin.text.lowercase' call
          // Inline function 'kotlin.js.asDynamic' call
          tmp_0 = _this__u8e3s4.ti(i).toLowerCase();
        } else if (!(strategyForClasses == null)) {
          tmp_0 = strategyForClasses.dy(_this__u8e3s4, i, _this__u8e3s4.ti(i));
        } else {
          tmp_0 = null;
        }
        var nameToPut = tmp_0;
        if (nameToPut == null)
          null;
        else {
          // Inline function 'kotlin.let' call
          // Inline function 'kotlin.contracts.contract' call
          buildDeserializationNamesMap$putOrThrow(builder, _this__u8e3s4, nameToPut, i);
        }
      }
       while (inductionVariable < last);
    // Inline function 'kotlin.collections.ifEmpty' call
    var tmp_1;
    if (builder.x()) {
      // Inline function 'kotlinx.serialization.json.internal.buildDeserializationNamesMap.<anonymous>' call
      tmp_1 = emptyMap();
    } else {
      tmp_1 = builder;
    }
    return tmp_1;
  }
  function buildDeserializationNamesMap$putOrThrow(_this__u8e3s4, $this_buildDeserializationNamesMap, name, index) {
    var entity = equals($this_buildDeserializationNamesMap.pi(), ENUM_getInstance()) ? 'enum value' : 'property';
    // Inline function 'kotlin.collections.contains' call
    // Inline function 'kotlin.collections.containsKey' call
    if ((isInterface(_this__u8e3s4, Map) ? _this__u8e3s4 : THROW_CCE()).y1(name)) {
      throw new JsonException("The suggested name '" + name + "' for " + entity + ' ' + $this_buildDeserializationNamesMap.ti(index) + ' is already one of the names for ' + entity + ' ' + ($this_buildDeserializationNamesMap.ti(getValue(_this__u8e3s4, name)) + ' in ' + $this_buildDeserializationNamesMap));
    }
    // Inline function 'kotlin.collections.set' call
    _this__u8e3s4.r1(name, index);
  }
  function deserializationNamesMap$lambda($descriptor, $this_deserializationNamesMap) {
    return function () {
      return buildDeserializationNamesMap($descriptor, $this_deserializationNamesMap);
    };
  }
  var properties_initialized_JsonNamesMap_kt_ljpf42;
  function _init_properties_JsonNamesMap_kt__cbbp0k() {
    if (!properties_initialized_JsonNamesMap_kt_ljpf42) {
      properties_initialized_JsonNamesMap_kt_ljpf42 = true;
      JsonDeserializationNamesKey = new Key();
      JsonSerializationNamesKey = new Key();
    }
  }
  function Tombstone() {
  }
  var Tombstone_instance;
  function Tombstone_getInstance() {
    return Tombstone_instance;
  }
  function resize($this) {
    var newSize = imul($this.gy_1, 2);
    $this.ey_1 = copyOf($this.ey_1, newSize);
    $this.fy_1 = copyOf_0($this.fy_1, newSize);
  }
  function JsonPath() {
    var tmp = this;
    // Inline function 'kotlin.arrayOfNulls' call
    tmp.ey_1 = fillArrayVal(Array(8), null);
    var tmp_0 = this;
    var tmp_1 = 0;
    var tmp_2 = new Int32Array(8);
    while (tmp_1 < 8) {
      tmp_2[tmp_1] = -1;
      tmp_1 = tmp_1 + 1 | 0;
    }
    tmp_0.fy_1 = tmp_2;
    this.gy_1 = -1;
  }
  protoOf(JsonPath).hy = function (sd) {
    this.gy_1 = this.gy_1 + 1 | 0;
    var depth = this.gy_1;
    if (depth === this.ey_1.length) {
      resize(this);
    }
    this.ey_1[depth] = sd;
  };
  protoOf(JsonPath).iy = function (index) {
    this.fy_1[this.gy_1] = index;
  };
  protoOf(JsonPath).jy = function (key) {
    var tmp;
    if (!(this.fy_1[this.gy_1] === -2)) {
      this.gy_1 = this.gy_1 + 1 | 0;
      tmp = this.gy_1 === this.ey_1.length;
    } else {
      tmp = false;
    }
    if (tmp) {
      resize(this);
    }
    this.ey_1[this.gy_1] = key;
    this.fy_1[this.gy_1] = -2;
  };
  protoOf(JsonPath).ky = function () {
    if (this.fy_1[this.gy_1] === -2) {
      this.ey_1[this.gy_1] = Tombstone_instance;
    }
  };
  protoOf(JsonPath).ly = function () {
    var depth = this.gy_1;
    if (this.fy_1[depth] === -2) {
      this.fy_1[depth] = -1;
      this.gy_1 = this.gy_1 - 1 | 0;
    }
    if (!(this.gy_1 === -1)) {
      this.gy_1 = this.gy_1 - 1 | 0;
    }
  };
  protoOf(JsonPath).my = function () {
    // Inline function 'kotlin.text.buildString' call
    // Inline function 'kotlin.contracts.contract' call
    // Inline function 'kotlin.apply' call
    var this_0 = StringBuilder_init_$Create$();
    // Inline function 'kotlin.contracts.contract' call
    // Inline function 'kotlinx.serialization.json.internal.JsonPath.getPath.<anonymous>' call
    this_0.z4('$');
    // Inline function 'kotlin.repeat' call
    var times = this.gy_1 + 1 | 0;
    // Inline function 'kotlin.contracts.contract' call
    var inductionVariable = 0;
    if (inductionVariable < times)
      do {
        var index = inductionVariable;
        inductionVariable = inductionVariable + 1 | 0;
        // Inline function 'kotlinx.serialization.json.internal.JsonPath.getPath.<anonymous>.<anonymous>' call
        var element = this.ey_1[index];
        if (!(element == null) ? isInterface(element, SerialDescriptor) : false) {
          if (equals(element.pi(), LIST_getInstance())) {
            if (!(this.fy_1[index] === -1)) {
              this_0.z4('[');
              this_0.y4(this.fy_1[index]);
              this_0.z4(']');
            }
          } else {
            var idx = this.fy_1[index];
            if (idx >= 0) {
              this_0.z4('.');
              this_0.z4(element.ti(idx));
            }
          }
        } else {
          if (!(element === Tombstone_instance)) {
            this_0.z4('[');
            this_0.z4("'");
            this_0.y4(element);
            this_0.z4("'");
            this_0.z4(']');
          }
        }
      }
       while (inductionVariable < times);
    return this_0.toString();
  };
  protoOf(JsonPath).toString = function () {
    return this.my();
  };
  function readObject($this) {
    // Inline function 'kotlinx.serialization.json.internal.JsonTreeReader.readObjectImpl' call
    var lastToken = $this.ny_1.qy(get_TC_BEGIN_OBJ());
    if ($this.ny_1.ry() === get_TC_COMMA()) {
      $this.ny_1.zx('Unexpected leading comma');
    }
    // Inline function 'kotlin.collections.linkedMapOf' call
    var result = LinkedHashMap_init_$Create$();
    $l$loop: while ($this.ny_1.sy()) {
      var key = $this.oy_1 ? $this.ny_1.uy() : $this.ny_1.ty();
      $this.ny_1.qy(get_TC_COLON());
      // Inline function 'kotlinx.serialization.json.internal.JsonTreeReader.readObject.<anonymous>' call
      var element = $this.vy();
      // Inline function 'kotlin.collections.set' call
      result.r1(key, element);
      lastToken = $this.ny_1.wy();
      var tmp0_subject = lastToken;
      if (tmp0_subject !== get_TC_COMMA())
        if (tmp0_subject === get_TC_END_OBJ())
          break $l$loop;
        else {
          $this.ny_1.zx('Expected end of the object or comma');
        }
    }
    if (lastToken === get_TC_BEGIN_OBJ()) {
      $this.ny_1.qy(get_TC_END_OBJ());
    } else if (lastToken === get_TC_COMMA()) {
      $this.ny_1.zx('Unexpected trailing comma');
    }
    return new JsonObject(result);
  }
  function readObject_0(_this__u8e3s4, $this, $completion) {
    var tmp = new $readObjectCOROUTINE$0($this, _this__u8e3s4, $completion);
    tmp.xa_1 = Unit_instance;
    tmp.ya_1 = null;
    return tmp.rb();
  }
  function readArray($this) {
    var lastToken = $this.ny_1.wy();
    if ($this.ny_1.ry() === get_TC_COMMA()) {
      $this.ny_1.zx('Unexpected leading comma');
    }
    // Inline function 'kotlin.collections.arrayListOf' call
    var result = ArrayList_init_$Create$();
    while ($this.ny_1.sy()) {
      var element = $this.vy();
      result.o(element);
      lastToken = $this.ny_1.wy();
      if (!(lastToken === get_TC_COMMA())) {
        // Inline function 'kotlinx.serialization.json.internal.AbstractJsonLexer.require' call
        var this_0 = $this.ny_1;
        var condition = lastToken === get_TC_END_LIST();
        var position = this_0.cw_1;
        if (!condition) {
          // Inline function 'kotlinx.serialization.json.internal.JsonTreeReader.readArray.<anonymous>' call
          var tmp$ret$1 = 'Expected end of the array or comma';
          this_0.zx(tmp$ret$1, position);
        }
      }
    }
    if (lastToken === get_TC_BEGIN_LIST()) {
      $this.ny_1.qy(get_TC_END_LIST());
    } else if (lastToken === get_TC_COMMA()) {
      $this.ny_1.zx('Unexpected trailing comma');
    }
    return new JsonArray(result);
  }
  function readValue($this, isString) {
    var tmp;
    if ($this.oy_1 ? true : !isString) {
      tmp = $this.ny_1.uy();
    } else {
      tmp = $this.ny_1.ty();
    }
    var string = tmp;
    if (!isString ? string === get_NULL() : false)
      return JsonNull_getInstance();
    return new JsonLiteral(string, isString);
  }
  function readDeepRecursive($this) {
    return invoke(new DeepRecursiveFunction(JsonTreeReader$readDeepRecursive$slambda_0($this, null)), Unit_instance);
  }
  function JsonTreeReader$readDeepRecursive$slambda(this$0, resultContinuation) {
    this.tz_1 = this$0;
    CoroutineImpl.call(this, resultContinuation);
  }
  protoOf(JsonTreeReader$readDeepRecursive$slambda).yz = function ($this$$receiver, it, $completion) {
    var tmp = this.zz($this$$receiver, it, $completion);
    tmp.xa_1 = Unit_instance;
    tmp.ya_1 = null;
    return tmp.rb();
  };
  protoOf(JsonTreeReader$readDeepRecursive$slambda).eg = function (p1, p2, $completion) {
    var tmp = p1 instanceof DeepRecursiveScope ? p1 : THROW_CCE();
    return this.yz(tmp, p2 instanceof Unit ? p2 : THROW_CCE(), $completion);
  };
  protoOf(JsonTreeReader$readDeepRecursive$slambda).rb = function () {
    var suspendResult = this.xa_1;
    $sm: do
      try {
        var tmp = this.va_1;
        switch (tmp) {
          case 0:
            this.wa_1 = 3;
            this.wz_1 = this.tz_1.ny_1.ry();
            if (this.wz_1 === get_TC_STRING()) {
              this.xz_1 = readValue(this.tz_1, true);
              this.va_1 = 2;
              continue $sm;
            } else {
              if (this.wz_1 === get_TC_OTHER()) {
                this.xz_1 = readValue(this.tz_1, false);
                this.va_1 = 2;
                continue $sm;
              } else {
                if (this.wz_1 === get_TC_BEGIN_OBJ()) {
                  this.va_1 = 1;
                  suspendResult = readObject_0(this.uz_1, this.tz_1, this);
                  if (suspendResult === get_COROUTINE_SUSPENDED()) {
                    return suspendResult;
                  }
                  continue $sm;
                } else {
                  if (this.wz_1 === get_TC_BEGIN_LIST()) {
                    this.xz_1 = readArray(this.tz_1);
                    this.va_1 = 2;
                    continue $sm;
                  } else {
                    var tmp_0 = this;
                    this.tz_1.ny_1.zx("Can't begin reading element, unexpected token");
                  }
                }
              }
            }

            break;
          case 1:
            this.xz_1 = suspendResult;
            this.va_1 = 2;
            continue $sm;
          case 2:
            return this.xz_1;
          case 3:
            throw this.ya_1;
        }
      } catch ($p) {
        var e = $p;
        if (this.wa_1 === 3) {
          throw e;
        } else {
          this.va_1 = this.wa_1;
          this.ya_1 = e;
        }
      }
     while (true);
  };
  protoOf(JsonTreeReader$readDeepRecursive$slambda).zz = function ($this$$receiver, it, completion) {
    var i = new JsonTreeReader$readDeepRecursive$slambda(this.tz_1, completion);
    i.uz_1 = $this$$receiver;
    i.vz_1 = it;
    return i;
  };
  function JsonTreeReader$readDeepRecursive$slambda_0(this$0, resultContinuation) {
    var i = new JsonTreeReader$readDeepRecursive$slambda(this$0, resultContinuation);
    var l = function ($this$$receiver, it, $completion) {
      return i.yz($this$$receiver, it, $completion);
    };
    l.$arity = 2;
    return l;
  }
  function $readObjectCOROUTINE$0(_this__u8e3s4, _this__u8e3s4_0, resultContinuation) {
    CoroutineImpl.call(this, resultContinuation);
    this.fz_1 = _this__u8e3s4;
    this.gz_1 = _this__u8e3s4_0;
  }
  protoOf($readObjectCOROUTINE$0).rb = function () {
    var suspendResult = this.xa_1;
    $sm: do
      try {
        var tmp = this.va_1;
        switch (tmp) {
          case 0:
            this.wa_1 = 5;
            var tmp_0 = this;
            tmp_0.hz_1 = this.fz_1;
            this.iz_1 = this.hz_1.ny_1.qy(get_TC_BEGIN_OBJ());
            if (this.hz_1.ny_1.ry() === get_TC_COMMA()) {
              this.hz_1.ny_1.zx('Unexpected leading comma');
            }

            var tmp_1 = this;
            tmp_1.jz_1 = LinkedHashMap_init_$Create$();
            this.va_1 = 1;
            continue $sm;
          case 1:
            if (!this.hz_1.ny_1.sy()) {
              this.va_1 = 4;
              continue $sm;
            }

            this.kz_1 = this.hz_1.oy_1 ? this.hz_1.ny_1.uy() : this.hz_1.ny_1.ty();
            this.hz_1.ny_1.qy(get_TC_COLON());
            this.va_1 = 2;
            suspendResult = this.gz_1.wf(Unit_instance, this);
            if (suspendResult === get_COROUTINE_SUSPENDED()) {
              return suspendResult;
            }

            continue $sm;
          case 2:
            var element = suspendResult;
            var this_0 = this.jz_1;
            var key = this.kz_1;
            this_0.r1(key, element);
            this.iz_1 = this.hz_1.ny_1.wy();
            var tmp0_subject = this.iz_1;
            if (tmp0_subject === get_TC_COMMA()) {
              this.va_1 = 3;
              continue $sm;
            } else {
              if (tmp0_subject === get_TC_END_OBJ()) {
                this.va_1 = 4;
                continue $sm;
              } else {
                this.hz_1.ny_1.zx('Expected end of the object or comma');
              }
            }

            break;
          case 3:
            this.va_1 = 1;
            continue $sm;
          case 4:
            if (this.iz_1 === get_TC_BEGIN_OBJ()) {
              this.hz_1.ny_1.qy(get_TC_END_OBJ());
            } else if (this.iz_1 === get_TC_COMMA()) {
              this.hz_1.ny_1.zx('Unexpected trailing comma');
            }

            return new JsonObject(this.jz_1);
          case 5:
            throw this.ya_1;
        }
      } catch ($p) {
        var e = $p;
        if (this.wa_1 === 5) {
          throw e;
        } else {
          this.va_1 = this.wa_1;
          this.ya_1 = e;
        }
      }
     while (true);
  };
  function JsonTreeReader(configuration, lexer) {
    this.ny_1 = lexer;
    this.oy_1 = configuration.jw_1;
    this.py_1 = 0;
  }
  protoOf(JsonTreeReader).vy = function () {
    var token = this.ny_1.ry();
    var tmp;
    if (token === get_TC_STRING()) {
      tmp = readValue(this, true);
    } else if (token === get_TC_OTHER()) {
      tmp = readValue(this, false);
    } else if (token === get_TC_BEGIN_OBJ()) {
      var tmp_0;
      this.py_1 = this.py_1 + 1 | 0;
      if (this.py_1 === 200) {
        tmp_0 = readDeepRecursive(this);
      } else {
        tmp_0 = readObject(this);
      }
      var result = tmp_0;
      this.py_1 = this.py_1 - 1 | 0;
      tmp = result;
    } else if (token === get_TC_BEGIN_LIST()) {
      tmp = readArray(this);
    } else {
      this.ny_1.zx('Cannot read Json element because of unexpected ' + tokenDescription(token));
    }
    return tmp;
  };
  function classDiscriminator(_this__u8e3s4, json) {
    var tmp0_iterator = _this__u8e3s4.si().r();
    while (tmp0_iterator.s()) {
      var annotation = tmp0_iterator.t();
      if (annotation instanceof JsonClassDiscriminator)
        return annotation.a10_1;
    }
    return json.pv_1.qw_1;
  }
  function decodeSerializableValuePolymorphic(_this__u8e3s4, deserializer) {
    var tmp;
    if (!(deserializer instanceof AbstractPolymorphicSerializer)) {
      tmp = true;
    } else {
      tmp = _this__u8e3s4.vw().pv_1.pw_1;
    }
    if (tmp) {
      return deserializer.gi(_this__u8e3s4);
    }
    var discriminator = classDiscriminator(deserializer.fi(), _this__u8e3s4.vw());
    // Inline function 'kotlinx.serialization.json.internal.cast' call
    var value = _this__u8e3s4.ww();
    var descriptor = deserializer.fi();
    if (!(value instanceof JsonObject)) {
      throw JsonDecodingException_1(-1, 'Expected ' + getKClass(JsonObject) + ' as the serialized body of ' + descriptor.oi() + ', but had ' + getKClassFromExpression(value));
    }
    var jsonTree = value;
    var tmp0_safe_receiver = jsonTree.e8(discriminator);
    var tmp1_safe_receiver = tmp0_safe_receiver == null ? null : get_jsonPrimitive(tmp0_safe_receiver);
    var type = tmp1_safe_receiver == null ? null : tmp1_safe_receiver.zw();
    var tmp2_elvis_lhs = deserializer.ml(_this__u8e3s4, type);
    var tmp_0;
    if (tmp2_elvis_lhs == null) {
      throwSerializerNotFound(type, jsonTree);
    } else {
      tmp_0 = tmp2_elvis_lhs;
    }
    var actualSerializer = tmp_0;
    var tmp_1 = _this__u8e3s4.vw();
    return readPolymorphicJson(tmp_1, discriminator, jsonTree, isInterface(actualSerializer, DeserializationStrategy) ? actualSerializer : THROW_CCE());
  }
  function throwSerializerNotFound(type, jsonTree) {
    var suffix = type == null ? "missing class discriminator ('null')" : "class discriminator '" + type + "'";
    throw JsonDecodingException_0(-1, 'Polymorphic serializer was not found for ' + suffix, jsonTree.toString());
  }
  function Key() {
  }
  function DescriptorSchemaCache() {
    this.ay_1 = createMapForCache(16);
  }
  protoOf(DescriptorSchemaCache).b10 = function (descriptor, key, value) {
    // Inline function 'kotlin.collections.set' call
    // Inline function 'kotlin.collections.getOrPut' call
    var this_0 = this.ay_1;
    var value_0 = this_0.b2(descriptor);
    var tmp;
    if (value_0 == null) {
      // Inline function 'kotlinx.serialization.json.internal.DescriptorSchemaCache.set.<anonymous>' call
      var answer = createMapForCache(2);
      this_0.r1(descriptor, answer);
      tmp = answer;
    } else {
      tmp = value_0;
    }
    var this_1 = tmp;
    var key_0 = key instanceof Key ? key : THROW_CCE();
    var value_1 = !(value == null) ? value : THROW_CCE();
    this_1.r1(key_0, value_1);
  };
  protoOf(DescriptorSchemaCache).by = function (descriptor, key, defaultValue) {
    var tmp0_safe_receiver = this.c10(descriptor, key);
    if (tmp0_safe_receiver == null)
      null;
    else {
      // Inline function 'kotlin.let' call
      // Inline function 'kotlin.contracts.contract' call
      return tmp0_safe_receiver;
    }
    var value = defaultValue();
    this.b10(descriptor, key, value);
    return value;
  };
  protoOf(DescriptorSchemaCache).c10 = function (descriptor, key) {
    var tmp0_safe_receiver = this.ay_1.b2(descriptor);
    var tmp;
    if (tmp0_safe_receiver == null) {
      tmp = null;
    } else {
      tmp = tmp0_safe_receiver.b2(key instanceof Key ? key : THROW_CCE());
    }
    var tmp_0 = tmp;
    return !(tmp_0 == null) ? tmp_0 : null;
  };
  function DiscriminatorHolder(discriminatorToSkip) {
    this.d10_1 = discriminatorToSkip;
  }
  function trySkip(_this__u8e3s4, $this, unknownKey) {
    if (_this__u8e3s4 == null)
      return false;
    if (_this__u8e3s4.d10_1 === unknownKey) {
      _this__u8e3s4.d10_1 = null;
      return true;
    }
    return false;
  }
  function skipLeftoverElements($this, descriptor) {
    $l$loop: while (true) {
      var tmp = $this.hl(descriptor);
      if (!!(tmp === -1)) {
        break $l$loop;
      }
    }
  }
  function checkLeadingComma($this) {
    if ($this.wv_1.ry() === get_TC_COMMA()) {
      $this.wv_1.zx('Unexpected leading comma');
    }
  }
  function decodeMapIndex($this) {
    var hasComma = false;
    var decodingKey = !(($this.yv_1 % 2 | 0) === 0);
    if (decodingKey) {
      if (!($this.yv_1 === -1)) {
        hasComma = $this.wv_1.f10();
      }
    } else {
      $this.wv_1.e10(get_COLON());
    }
    var tmp;
    if ($this.wv_1.sy()) {
      if (decodingKey) {
        if ($this.yv_1 === -1) {
          // Inline function 'kotlinx.serialization.json.internal.AbstractJsonLexer.require' call
          var this_0 = $this.wv_1;
          var condition = !hasComma;
          var position = this_0.cw_1;
          if (!condition) {
            // Inline function 'kotlinx.serialization.json.internal.StreamingJsonDecoder.decodeMapIndex.<anonymous>' call
            var tmp$ret$0 = 'Unexpected trailing comma';
            this_0.zx(tmp$ret$0, position);
          }
        } else {
          // Inline function 'kotlinx.serialization.json.internal.AbstractJsonLexer.require' call
          var this_1 = $this.wv_1;
          var condition_0 = hasComma;
          var position_0 = this_1.cw_1;
          if (!condition_0) {
            // Inline function 'kotlinx.serialization.json.internal.StreamingJsonDecoder.decodeMapIndex.<anonymous>' call
            var tmp$ret$1 = 'Expected comma after the key-value pair';
            this_1.zx(tmp$ret$1, position_0);
          }
        }
      }
      $this.yv_1 = $this.yv_1 + 1 | 0;
      tmp = $this.yv_1;
    } else {
      if (hasComma) {
        $this.wv_1.zx("Expected '}', but had ',' instead");
      }
      tmp = -1;
    }
    return tmp;
  }
  function coerceInputValue($this, descriptor, index) {
    var tmp$ret$1;
    $l$block_2: {
      // Inline function 'kotlinx.serialization.json.internal.tryCoerceValue' call
      var this_0 = $this.uv_1;
      var elementDescriptor = descriptor.wi(index);
      var tmp;
      if (!elementDescriptor.ii()) {
        // Inline function 'kotlinx.serialization.json.internal.StreamingJsonDecoder.coerceInputValue.<anonymous>' call
        tmp = $this.wv_1.g10(true);
      } else {
        tmp = false;
      }
      if (tmp) {
        tmp$ret$1 = true;
        break $l$block_2;
      }
      if (equals(elementDescriptor.pi(), ENUM_getInstance())) {
        var tmp_0;
        if (elementDescriptor.ii()) {
          // Inline function 'kotlinx.serialization.json.internal.StreamingJsonDecoder.coerceInputValue.<anonymous>' call
          tmp_0 = $this.wv_1.g10(false);
        } else {
          tmp_0 = false;
        }
        if (tmp_0) {
          tmp$ret$1 = false;
          break $l$block_2;
        }
        // Inline function 'kotlinx.serialization.json.internal.StreamingJsonDecoder.coerceInputValue.<anonymous>' call
        var tmp0_elvis_lhs = $this.wv_1.h10($this.aw_1.jw_1);
        var tmp_1;
        if (tmp0_elvis_lhs == null) {
          tmp$ret$1 = false;
          break $l$block_2;
        } else {
          tmp_1 = tmp0_elvis_lhs;
        }
        var enumValue = tmp_1;
        var enumIndex = getJsonNameIndex(elementDescriptor, this_0, enumValue);
        if (enumIndex === -3) {
          // Inline function 'kotlinx.serialization.json.internal.StreamingJsonDecoder.coerceInputValue.<anonymous>' call
          $this.wv_1.ty();
          tmp$ret$1 = true;
          break $l$block_2;
        }
      }
      tmp$ret$1 = false;
    }
    return tmp$ret$1;
  }
  function decodeObjectIndex($this, descriptor) {
    var hasComma = $this.wv_1.f10();
    while ($this.wv_1.sy()) {
      hasComma = false;
      var key = decodeStringKey($this);
      $this.wv_1.e10(get_COLON());
      var index = getJsonNameIndex(descriptor, $this.uv_1, key);
      var tmp;
      if (!(index === -3)) {
        var tmp_0;
        if ($this.aw_1.ow_1 ? coerceInputValue($this, descriptor, index) : false) {
          hasComma = $this.wv_1.f10();
          tmp_0 = false;
        } else {
          var tmp0_safe_receiver = $this.bw_1;
          if (tmp0_safe_receiver == null)
            null;
          else {
            tmp0_safe_receiver.vn(index);
          }
          return index;
        }
        tmp = tmp_0;
      } else {
        tmp = true;
      }
      var isUnknown = tmp;
      if (isUnknown) {
        hasComma = handleUnknown($this, key);
      }
    }
    if (hasComma) {
      $this.wv_1.zx('Unexpected trailing comma');
    }
    var tmp1_safe_receiver = $this.bw_1;
    var tmp2_elvis_lhs = tmp1_safe_receiver == null ? null : tmp1_safe_receiver.wn();
    var tmp_1;
    if (tmp2_elvis_lhs == null) {
      tmp_1 = -1;
    } else {
      tmp_1 = tmp2_elvis_lhs;
    }
    return tmp_1;
  }
  function handleUnknown($this, key) {
    if ($this.aw_1.iw_1 ? true : trySkip($this.zv_1, $this, key)) {
      $this.wv_1.j10($this.aw_1.jw_1);
    } else {
      $this.wv_1.i10(key);
    }
    return $this.wv_1.f10();
  }
  function decodeListIndex($this) {
    var hasComma = $this.wv_1.f10();
    var tmp;
    if ($this.wv_1.sy()) {
      if (!($this.yv_1 === -1) ? !hasComma : false) {
        $this.wv_1.zx('Expected end of the array or comma');
      }
      $this.yv_1 = $this.yv_1 + 1 | 0;
      tmp = $this.yv_1;
    } else {
      if (hasComma) {
        $this.wv_1.zx('Unexpected trailing comma');
      }
      tmp = -1;
    }
    return tmp;
  }
  function decodeStringKey($this) {
    var tmp;
    if ($this.aw_1.jw_1) {
      tmp = $this.wv_1.l10();
    } else {
      tmp = $this.wv_1.k10();
    }
    return tmp;
  }
  function StreamingJsonDecoder(json, mode, lexer, descriptor, discriminatorHolder) {
    AbstractDecoder.call(this);
    this.uv_1 = json;
    this.vv_1 = mode;
    this.wv_1 = lexer;
    this.xv_1 = this.uv_1.fl();
    this.yv_1 = -1;
    this.zv_1 = discriminatorHolder;
    this.aw_1 = this.uv_1.pv_1;
    this.bw_1 = this.aw_1.mw_1 ? null : new JsonElementMarker(descriptor);
  }
  protoOf(StreamingJsonDecoder).vw = function () {
    return this.uv_1;
  };
  protoOf(StreamingJsonDecoder).fl = function () {
    return this.xv_1;
  };
  protoOf(StreamingJsonDecoder).ww = function () {
    return (new JsonTreeReader(this.uv_1.pv_1, this.wv_1)).vy();
  };
  protoOf(StreamingJsonDecoder).qk = function (deserializer) {
    try {
      var tmp;
      if (!(deserializer instanceof AbstractPolymorphicSerializer)) {
        tmp = true;
      } else {
        tmp = this.uv_1.pv_1.pw_1;
      }
      if (tmp) {
        return deserializer.gi(this);
      }
      var discriminator = classDiscriminator(deserializer.fi(), this.uv_1);
      var type = this.wv_1.m10(discriminator, this.aw_1.jw_1);
      var actualSerializer = null;
      if (!(type == null)) {
        actualSerializer = deserializer.ml(this, type);
      }
      if (actualSerializer == null) {
        return decodeSerializableValuePolymorphic(this, isInterface(deserializer, DeserializationStrategy) ? deserializer : THROW_CCE());
      }
      this.zv_1 = new DiscriminatorHolder(discriminator);
      var tmp_0 = actualSerializer.gi(this);
      var result = !(tmp_0 == null) ? tmp_0 : THROW_CCE();
      return result;
    } catch ($p) {
      if ($p instanceof MissingFieldException) {
        var e = $p;
        if (contains_0(ensureNotNull(e.message), 'at path'))
          throw e;
        throw new MissingFieldException(e.hi_1, plus(e.message, ' at path: ') + this.wv_1.dw_1.my(), e);
      } else {
        throw $p;
      }
    }
  };
  protoOf(StreamingJsonDecoder).rk = function (descriptor) {
    var newMode = switchMode(this.uv_1, descriptor);
    this.wv_1.dw_1.hy(descriptor);
    this.wv_1.e10(newMode.p10_1);
    checkLeadingComma(this);
    var tmp;
    switch (newMode.t8_1) {
      case 1:
      case 2:
      case 3:
        tmp = new StreamingJsonDecoder(this.uv_1, newMode, this.wv_1, descriptor, this.zv_1);
        break;
      default:
        var tmp_0;
        if (this.vv_1.equals(newMode) ? this.uv_1.pv_1.mw_1 : false) {
          tmp_0 = this;
        } else {
          tmp_0 = new StreamingJsonDecoder(this.uv_1, newMode, this.wv_1, descriptor, this.zv_1);
        }

        tmp = tmp_0;
        break;
    }
    return tmp;
  };
  protoOf(StreamingJsonDecoder).sk = function (descriptor) {
    if (this.uv_1.pv_1.iw_1 ? descriptor.ri() === 0 : false) {
      skipLeftoverElements(this, descriptor);
    }
    this.wv_1.e10(this.vv_1.q10_1);
    this.wv_1.dw_1.ly();
  };
  protoOf(StreamingJsonDecoder).ck = function () {
    var tmp;
    var tmp0_safe_receiver = this.bw_1;
    var tmp1_elvis_lhs = tmp0_safe_receiver == null ? null : tmp0_safe_receiver.yx_1;
    if (!(tmp1_elvis_lhs == null ? false : tmp1_elvis_lhs)) {
      tmp = !this.wv_1.r10();
    } else {
      tmp = false;
    }
    return tmp;
  };
  protoOf(StreamingJsonDecoder).dk = function () {
    return null;
  };
  protoOf(StreamingJsonDecoder).cl = function (descriptor, index, deserializer, previousValue) {
    var isMapKey = this.vv_1.equals(WriteMode_MAP_getInstance()) ? (index & 1) === 0 : false;
    if (isMapKey) {
      this.wv_1.dw_1.ky();
    }
    var value = protoOf(AbstractDecoder).cl.call(this, descriptor, index, deserializer, previousValue);
    if (isMapKey) {
      this.wv_1.dw_1.jy(value);
    }
    return value;
  };
  protoOf(StreamingJsonDecoder).hl = function (descriptor) {
    var index;
    switch (this.vv_1.t8_1) {
      case 0:
        index = decodeObjectIndex(this, descriptor);
        break;
      case 2:
        index = decodeMapIndex(this);
        break;
      default:
        index = decodeListIndex(this);
        break;
    }
    if (!this.vv_1.equals(WriteMode_MAP_getInstance())) {
      this.wv_1.dw_1.iy(index);
    }
    return index;
  };
  protoOf(StreamingJsonDecoder).ek = function () {
    var tmp;
    if (this.aw_1.jw_1) {
      tmp = this.wv_1.t10();
    } else {
      tmp = this.wv_1.s10();
    }
    return tmp;
  };
  protoOf(StreamingJsonDecoder).fk = function () {
    var value = this.wv_1.ex();
    if (!value.equals(toLong(value.ma()))) {
      this.wv_1.zx("Failed to parse byte for input '" + value.toString() + "'");
    }
    return value.ma();
  };
  protoOf(StreamingJsonDecoder).gk = function () {
    var value = this.wv_1.ex();
    if (!value.equals(toLong(value.na()))) {
      this.wv_1.zx("Failed to parse short for input '" + value.toString() + "'");
    }
    return value.na();
  };
  protoOf(StreamingJsonDecoder).hk = function () {
    var value = this.wv_1.ex();
    if (!value.equals(toLong(value.f9()))) {
      this.wv_1.zx("Failed to parse int for input '" + value.toString() + "'");
    }
    return value.f9();
  };
  protoOf(StreamingJsonDecoder).ik = function () {
    return this.wv_1.ex();
  };
  protoOf(StreamingJsonDecoder).jk = function () {
    var tmp$ret$4;
    $l$block: {
      // Inline function 'kotlinx.serialization.json.internal.parseString' call
      var this_0 = this.wv_1;
      var input = this_0.uy();
      try {
        // Inline function 'kotlinx.serialization.json.internal.StreamingJsonDecoder.decodeFloat.<anonymous>' call
        // Inline function 'kotlin.text.toFloat' call
        // Inline function 'kotlin.js.unsafeCast' call
        // Inline function 'kotlin.js.asDynamic' call
        tmp$ret$4 = toDouble(input);
        break $l$block;
      } catch ($p) {
        if ($p instanceof IllegalArgumentException) {
          var e = $p;
          this_0.zx("Failed to parse type '" + 'float' + "' for input '" + input + "'");
        } else {
          throw $p;
        }
      }
    }
    var result = tmp$ret$4;
    var specialFp = this.uv_1.pv_1.rw_1;
    if (specialFp ? true : isFinite(result))
      return result;
    throwInvalidFloatingPointDecoded(this.wv_1, result);
  };
  protoOf(StreamingJsonDecoder).kk = function () {
    var tmp$ret$1;
    $l$block: {
      // Inline function 'kotlinx.serialization.json.internal.parseString' call
      var this_0 = this.wv_1;
      var input = this_0.uy();
      try {
        // Inline function 'kotlinx.serialization.json.internal.StreamingJsonDecoder.decodeDouble.<anonymous>' call
        tmp$ret$1 = toDouble(input);
        break $l$block;
      } catch ($p) {
        if ($p instanceof IllegalArgumentException) {
          var e = $p;
          this_0.zx("Failed to parse type '" + 'double' + "' for input '" + input + "'");
        } else {
          throw $p;
        }
      }
    }
    var result = tmp$ret$1;
    var specialFp = this.uv_1.pv_1.rw_1;
    if (specialFp ? true : isFinite_0(result))
      return result;
    throwInvalidFloatingPointDecoded(this.wv_1, result);
  };
  protoOf(StreamingJsonDecoder).lk = function () {
    var string = this.wv_1.uy();
    if (!(string.length === 1)) {
      this.wv_1.zx("Expected single char, but got '" + string + "'");
    }
    return charSequenceGet(string, 0);
  };
  protoOf(StreamingJsonDecoder).mk = function () {
    var tmp;
    if (this.aw_1.jw_1) {
      tmp = this.wv_1.l10();
    } else {
      tmp = this.wv_1.ty();
    }
    return tmp;
  };
  protoOf(StreamingJsonDecoder).ok = function (descriptor) {
    return get_isUnsignedNumber(descriptor) ? new JsonDecoderForUnsignedTypes(this.wv_1, this.uv_1) : protoOf(AbstractDecoder).ok.call(this, descriptor);
  };
  protoOf(StreamingJsonDecoder).nk = function (enumDescriptor) {
    return getJsonNameIndexOrThrow(enumDescriptor, this.uv_1, this.mk(), ' at path ' + this.wv_1.dw_1.my());
  };
  function JsonDecoderForUnsignedTypes(lexer, json) {
    AbstractDecoder.call(this);
    this.u10_1 = lexer;
    this.v10_1 = json.fl();
  }
  protoOf(JsonDecoderForUnsignedTypes).fl = function () {
    return this.v10_1;
  };
  protoOf(JsonDecoderForUnsignedTypes).hl = function (descriptor) {
    var message = 'unsupported';
    throw IllegalStateException_init_$Create$(toString(message));
  };
  protoOf(JsonDecoderForUnsignedTypes).hk = function () {
    var tmp$ret$2;
    $l$block: {
      // Inline function 'kotlinx.serialization.json.internal.parseString' call
      var this_0 = this.u10_1;
      var input = this_0.uy();
      try {
        // Inline function 'kotlinx.serialization.json.internal.JsonDecoderForUnsignedTypes.decodeInt.<anonymous>' call
        // Inline function 'kotlin.UInt.toInt' call
        var this_1 = toUInt(input);
        tmp$ret$2 = _UInt___get_data__impl__f0vqqw(this_1);
        break $l$block;
      } catch ($p) {
        if ($p instanceof IllegalArgumentException) {
          var e = $p;
          this_0.zx("Failed to parse type '" + 'UInt' + "' for input '" + input + "'");
        } else {
          throw $p;
        }
      }
    }
    return tmp$ret$2;
  };
  protoOf(JsonDecoderForUnsignedTypes).ik = function () {
    var tmp$ret$2;
    $l$block: {
      // Inline function 'kotlinx.serialization.json.internal.parseString' call
      var this_0 = this.u10_1;
      var input = this_0.uy();
      try {
        // Inline function 'kotlinx.serialization.json.internal.JsonDecoderForUnsignedTypes.decodeLong.<anonymous>' call
        // Inline function 'kotlin.ULong.toLong' call
        var this_1 = toULong(input);
        tmp$ret$2 = _ULong___get_data__impl__fggpzb(this_1);
        break $l$block;
      } catch ($p) {
        if ($p instanceof IllegalArgumentException) {
          var e = $p;
          this_0.zx("Failed to parse type '" + 'ULong' + "' for input '" + input + "'");
        } else {
          throw $p;
        }
      }
    }
    return tmp$ret$2;
  };
  protoOf(JsonDecoderForUnsignedTypes).fk = function () {
    var tmp$ret$2;
    $l$block: {
      // Inline function 'kotlinx.serialization.json.internal.parseString' call
      var this_0 = this.u10_1;
      var input = this_0.uy();
      try {
        // Inline function 'kotlinx.serialization.json.internal.JsonDecoderForUnsignedTypes.decodeByte.<anonymous>' call
        // Inline function 'kotlin.UByte.toByte' call
        var this_1 = toUByte(input);
        tmp$ret$2 = _UByte___get_data__impl__jof9qr(this_1);
        break $l$block;
      } catch ($p) {
        if ($p instanceof IllegalArgumentException) {
          var e = $p;
          this_0.zx("Failed to parse type '" + 'UByte' + "' for input '" + input + "'");
        } else {
          throw $p;
        }
      }
    }
    return tmp$ret$2;
  };
  protoOf(JsonDecoderForUnsignedTypes).gk = function () {
    var tmp$ret$2;
    $l$block: {
      // Inline function 'kotlinx.serialization.json.internal.parseString' call
      var this_0 = this.u10_1;
      var input = this_0.uy();
      try {
        // Inline function 'kotlinx.serialization.json.internal.JsonDecoderForUnsignedTypes.decodeShort.<anonymous>' call
        // Inline function 'kotlin.UShort.toShort' call
        var this_1 = toUShort(input);
        tmp$ret$2 = _UShort___get_data__impl__g0245(this_1);
        break $l$block;
      } catch ($p) {
        if ($p instanceof IllegalArgumentException) {
          var e = $p;
          this_0.zx("Failed to parse type '" + 'UShort' + "' for input '" + input + "'");
        } else {
          throw $p;
        }
      }
    }
    return tmp$ret$2;
  };
  function get_unsignedNumberDescriptors() {
    _init_properties_StreamingJsonEncoder_kt__pn1bsi();
    return unsignedNumberDescriptors;
  }
  var unsignedNumberDescriptors;
  function get_isUnsignedNumber(_this__u8e3s4) {
    _init_properties_StreamingJsonEncoder_kt__pn1bsi();
    return _this__u8e3s4.qi() ? get_unsignedNumberDescriptors().v(_this__u8e3s4) : false;
  }
  var properties_initialized_StreamingJsonEncoder_kt_6ifwwk;
  function _init_properties_StreamingJsonEncoder_kt__pn1bsi() {
    if (!properties_initialized_StreamingJsonEncoder_kt_6ifwwk) {
      properties_initialized_StreamingJsonEncoder_kt_6ifwwk = true;
      unsignedNumberDescriptors = setOf([serializer_0(Companion_getInstance()).fi(), serializer_1(Companion_getInstance_0()).fi(), serializer_2(Companion_getInstance_1()).fi(), serializer_3(Companion_getInstance_2()).fi()]);
    }
  }
  function get_ESCAPE_STRINGS() {
    _init_properties_StringOps_kt__fcy1db();
    return ESCAPE_STRINGS;
  }
  var ESCAPE_STRINGS;
  var ESCAPE_MARKERS;
  function toHexChar(i) {
    _init_properties_StringOps_kt__fcy1db();
    var d = i & 15;
    var tmp;
    if (d < 10) {
      // Inline function 'kotlin.code' call
      var this_0 = _Char___init__impl__6a9atx(48);
      var tmp$ret$0 = Char__toInt_impl_vasixd(this_0);
      tmp = numberToChar(d + tmp$ret$0 | 0);
    } else {
      var tmp_0 = d - 10 | 0;
      // Inline function 'kotlin.code' call
      var this_1 = _Char___init__impl__6a9atx(97);
      var tmp$ret$1 = Char__toInt_impl_vasixd(this_1);
      tmp = numberToChar(tmp_0 + tmp$ret$1 | 0);
    }
    return tmp;
  }
  function printQuoted(_this__u8e3s4, value) {
    _init_properties_StringOps_kt__fcy1db();
    _this__u8e3s4.a5(get_STRING());
    var lastPos = 0;
    var inductionVariable = 0;
    var last = charSequenceLength(value) - 1 | 0;
    if (inductionVariable <= last)
      do {
        var i = inductionVariable;
        inductionVariable = inductionVariable + 1 | 0;
        // Inline function 'kotlin.code' call
        var this_0 = charSequenceGet(value, i);
        var c = Char__toInt_impl_vasixd(this_0);
        if (c < get_ESCAPE_STRINGS().length ? !(get_ESCAPE_STRINGS()[c] == null) : false) {
          _this__u8e3s4.h7(value, lastPos, i);
          _this__u8e3s4.z4(get_ESCAPE_STRINGS()[c]);
          lastPos = i + 1 | 0;
        }
      }
       while (inductionVariable <= last);
    if (!(lastPos === 0)) {
      _this__u8e3s4.h7(value, lastPos, value.length);
    } else {
      _this__u8e3s4.z4(value);
    }
    _this__u8e3s4.a5(get_STRING());
  }
  function toBooleanStrictOrNull(_this__u8e3s4) {
    _init_properties_StringOps_kt__fcy1db();
    return equals_0(_this__u8e3s4, 'true', true) ? true : equals_0(_this__u8e3s4, 'false', true) ? false : null;
  }
  var properties_initialized_StringOps_kt_wzaea7;
  function _init_properties_StringOps_kt__fcy1db() {
    if (!properties_initialized_StringOps_kt_wzaea7) {
      properties_initialized_StringOps_kt_wzaea7 = true;
      // Inline function 'kotlin.apply' call
      // Inline function 'kotlin.arrayOfNulls' call
      var this_0 = fillArrayVal(Array(93), null);
      // Inline function 'kotlin.contracts.contract' call
      // Inline function 'kotlinx.serialization.json.internal.ESCAPE_STRINGS.<anonymous>' call
      var inductionVariable = 0;
      if (inductionVariable <= 31)
        do {
          var c = inductionVariable;
          inductionVariable = inductionVariable + 1 | 0;
          var c1 = toHexChar(c >> 12);
          var c2 = toHexChar(c >> 8);
          var c3 = toHexChar(c >> 4);
          var c4 = toHexChar(c);
          this_0[c] = '\\u' + toString_0(c1) + toString_0(c2) + toString_0(c3) + toString_0(c4);
        }
         while (inductionVariable <= 31);
      // Inline function 'kotlin.code' call
      var this_1 = _Char___init__impl__6a9atx(34);
      this_0[Char__toInt_impl_vasixd(this_1)] = '\\"';
      // Inline function 'kotlin.code' call
      var this_2 = _Char___init__impl__6a9atx(92);
      this_0[Char__toInt_impl_vasixd(this_2)] = '\\\\';
      // Inline function 'kotlin.code' call
      var this_3 = _Char___init__impl__6a9atx(9);
      this_0[Char__toInt_impl_vasixd(this_3)] = '\\t';
      // Inline function 'kotlin.code' call
      var this_4 = _Char___init__impl__6a9atx(8);
      this_0[Char__toInt_impl_vasixd(this_4)] = '\\b';
      // Inline function 'kotlin.code' call
      var this_5 = _Char___init__impl__6a9atx(10);
      this_0[Char__toInt_impl_vasixd(this_5)] = '\\n';
      // Inline function 'kotlin.code' call
      var this_6 = _Char___init__impl__6a9atx(13);
      this_0[Char__toInt_impl_vasixd(this_6)] = '\\r';
      this_0[12] = '\\f';
      ESCAPE_STRINGS = this_0;
      // Inline function 'kotlin.apply' call
      var this_7 = new Int8Array(93);
      // Inline function 'kotlin.contracts.contract' call
      // Inline function 'kotlinx.serialization.json.internal.ESCAPE_MARKERS.<anonymous>' call
      var inductionVariable_0 = 0;
      if (inductionVariable_0 <= 31)
        do {
          var c_0 = inductionVariable_0;
          inductionVariable_0 = inductionVariable_0 + 1 | 0;
          this_7[c_0] = 1;
        }
         while (inductionVariable_0 <= 31);
      // Inline function 'kotlin.code' call
      var this_8 = _Char___init__impl__6a9atx(34);
      this_7[Char__toInt_impl_vasixd(this_8)] = 34;
      // Inline function 'kotlin.code' call
      var this_9 = _Char___init__impl__6a9atx(92);
      this_7[Char__toInt_impl_vasixd(this_9)] = 92;
      // Inline function 'kotlin.code' call
      var this_10 = _Char___init__impl__6a9atx(9);
      this_7[Char__toInt_impl_vasixd(this_10)] = 116;
      // Inline function 'kotlin.code' call
      var this_11 = _Char___init__impl__6a9atx(8);
      this_7[Char__toInt_impl_vasixd(this_11)] = 98;
      // Inline function 'kotlin.code' call
      var this_12 = _Char___init__impl__6a9atx(10);
      this_7[Char__toInt_impl_vasixd(this_12)] = 110;
      // Inline function 'kotlin.code' call
      var this_13 = _Char___init__impl__6a9atx(13);
      this_7[Char__toInt_impl_vasixd(this_13)] = 114;
      this_7[12] = 102;
      ESCAPE_MARKERS = this_7;
    }
  }
  function unparsedPrimitive($this, primitive) {
    throw JsonDecodingException_0(-1, "Failed to parse literal as '" + primitive + "' value", toString($this.b11()));
  }
  function asLiteral(_this__u8e3s4, $this, type) {
    var tmp0_elvis_lhs = _this__u8e3s4 instanceof JsonLiteral ? _this__u8e3s4 : null;
    var tmp;
    if (tmp0_elvis_lhs == null) {
      throw JsonDecodingException_1(-1, "Unexpected 'null' literal when non-nullable " + type + ' was expected');
    } else {
      tmp = tmp0_elvis_lhs;
    }
    return tmp;
  }
  function AbstractJsonTreeDecoder(json, value) {
    NamedValueDecoder.call(this);
    this.y10_1 = json;
    this.z10_1 = value;
    this.a11_1 = this.vw().pv_1;
  }
  protoOf(AbstractJsonTreeDecoder).vw = function () {
    return this.y10_1;
  };
  protoOf(AbstractJsonTreeDecoder).v1 = function () {
    return this.z10_1;
  };
  protoOf(AbstractJsonTreeDecoder).fl = function () {
    return this.vw().fl();
  };
  protoOf(AbstractJsonTreeDecoder).b11 = function () {
    var tmp0_safe_receiver = this.ju();
    var tmp;
    if (tmp0_safe_receiver == null) {
      tmp = null;
    } else {
      // Inline function 'kotlin.let' call
      // Inline function 'kotlin.contracts.contract' call
      // Inline function 'kotlinx.serialization.json.internal.AbstractJsonTreeDecoder.currentObject.<anonymous>' call
      tmp = this.c11(tmp0_safe_receiver);
    }
    var tmp1_elvis_lhs = tmp;
    return tmp1_elvis_lhs == null ? this.v1() : tmp1_elvis_lhs;
  };
  protoOf(AbstractJsonTreeDecoder).ww = function () {
    return this.b11();
  };
  protoOf(AbstractJsonTreeDecoder).qk = function (deserializer) {
    return decodeSerializableValuePolymorphic(this, deserializer);
  };
  protoOf(AbstractJsonTreeDecoder).ku = function (parentName, childName) {
    return childName;
  };
  protoOf(AbstractJsonTreeDecoder).rk = function (descriptor) {
    var currentObject = this.b11();
    var tmp0_subject = descriptor.pi();
    var tmp;
    var tmp_0;
    if (equals(tmp0_subject, LIST_getInstance())) {
      tmp_0 = true;
    } else {
      tmp_0 = tmp0_subject instanceof PolymorphicKind;
    }
    if (tmp_0) {
      var tmp_1 = this.vw();
      // Inline function 'kotlinx.serialization.json.internal.cast' call
      if (!(currentObject instanceof JsonArray)) {
        throw JsonDecodingException_1(-1, 'Expected ' + getKClass(JsonArray) + ' as the serialized body of ' + descriptor.oi() + ', but had ' + getKClassFromExpression(currentObject));
      }
      tmp = new JsonTreeListDecoder(tmp_1, currentObject);
    } else {
      if (equals(tmp0_subject, MAP_getInstance())) {
        // Inline function 'kotlinx.serialization.json.internal.selectMapMode' call
        var this_0 = this.vw();
        var keyDescriptor = carrierDescriptor(descriptor.wi(0), this_0.fl());
        var keyKind = keyDescriptor.pi();
        var tmp_2;
        var tmp_3;
        if (keyKind instanceof PrimitiveKind) {
          tmp_3 = true;
        } else {
          tmp_3 = equals(keyKind, ENUM_getInstance());
        }
        if (tmp_3) {
          // Inline function 'kotlinx.serialization.json.internal.AbstractJsonTreeDecoder.beginStructure.<anonymous>' call
          var tmp_4 = this.vw();
          // Inline function 'kotlinx.serialization.json.internal.cast' call
          if (!(currentObject instanceof JsonObject)) {
            throw JsonDecodingException_1(-1, 'Expected ' + getKClass(JsonObject) + ' as the serialized body of ' + descriptor.oi() + ', but had ' + getKClassFromExpression(currentObject));
          }
          tmp_2 = new JsonTreeMapDecoder(tmp_4, currentObject);
        } else {
          if (this_0.pv_1.kw_1) {
            // Inline function 'kotlinx.serialization.json.internal.AbstractJsonTreeDecoder.beginStructure.<anonymous>' call
            var tmp_5 = this.vw();
            // Inline function 'kotlinx.serialization.json.internal.cast' call
            if (!(currentObject instanceof JsonArray)) {
              throw JsonDecodingException_1(-1, 'Expected ' + getKClass(JsonArray) + ' as the serialized body of ' + descriptor.oi() + ', but had ' + getKClassFromExpression(currentObject));
            }
            tmp_2 = new JsonTreeListDecoder(tmp_5, currentObject);
          } else {
            throw InvalidKeyKindException(keyDescriptor);
          }
        }
        tmp = tmp_2;
      } else {
        var tmp_6 = this.vw();
        // Inline function 'kotlinx.serialization.json.internal.cast' call
        if (!(currentObject instanceof JsonObject)) {
          throw JsonDecodingException_1(-1, 'Expected ' + getKClass(JsonObject) + ' as the serialized body of ' + descriptor.oi() + ', but had ' + getKClassFromExpression(currentObject));
        }
        tmp = new JsonTreeDecoder(tmp_6, currentObject);
      }
    }
    return tmp;
  };
  protoOf(AbstractJsonTreeDecoder).sk = function (descriptor) {
  };
  protoOf(AbstractJsonTreeDecoder).ck = function () {
    var tmp = this.b11();
    return !(tmp instanceof JsonNull);
  };
  protoOf(AbstractJsonTreeDecoder).d11 = function (tag) {
    var currentElement = this.c11(tag);
    var tmp0_elvis_lhs = currentElement instanceof JsonPrimitive ? currentElement : null;
    var tmp;
    if (tmp0_elvis_lhs == null) {
      throw JsonDecodingException_0(-1, 'Expected JsonPrimitive at ' + tag + ', found ' + currentElement, toString(this.b11()));
    } else {
      tmp = tmp0_elvis_lhs;
    }
    return tmp;
  };
  protoOf(AbstractJsonTreeDecoder).e11 = function (tag, enumDescriptor) {
    return getJsonNameIndexOrThrow(enumDescriptor, this.vw(), this.d11(tag).zw());
  };
  protoOf(AbstractJsonTreeDecoder).wu = function (tag, enumDescriptor) {
    return this.e11((!(tag == null) ? typeof tag === 'string' : false) ? tag : THROW_CCE(), enumDescriptor);
  };
  protoOf(AbstractJsonTreeDecoder).f11 = function (tag) {
    return !(this.c11(tag) === JsonNull_getInstance());
  };
  protoOf(AbstractJsonTreeDecoder).mu = function (tag) {
    return this.f11((!(tag == null) ? typeof tag === 'string' : false) ? tag : THROW_CCE());
  };
  protoOf(AbstractJsonTreeDecoder).g11 = function (tag) {
    var value = this.d11(tag);
    if (!this.vw().pv_1.jw_1) {
      var literal = asLiteral(value, this, 'boolean');
      if (literal.fx_1)
        throw JsonDecodingException_0(-1, "Boolean literal for key '" + tag + "' should be unquoted.\n" + get_lenientHint(), toString(this.b11()));
    }
    var tmp$ret$1;
    $l$block: {
      // Inline function 'kotlinx.serialization.json.internal.AbstractJsonTreeDecoder.primitive' call
      try {
        // Inline function 'kotlinx.serialization.json.internal.AbstractJsonTreeDecoder.decodeTaggedBoolean.<anonymous>' call
        var tmp0_elvis_lhs = get_booleanOrNull(value);
        var tmp;
        if (tmp0_elvis_lhs == null) {
          throw IllegalArgumentException_init_$Create$_0();
        } else {
          tmp = tmp0_elvis_lhs;
        }
        var tmp0_elvis_lhs_0 = tmp;
        var tmp_0;
        if (tmp0_elvis_lhs_0 == null) {
          unparsedPrimitive(this, 'boolean');
        } else {
          tmp_0 = tmp0_elvis_lhs_0;
        }
        tmp$ret$1 = tmp_0;
        break $l$block;
      } catch ($p) {
        if ($p instanceof IllegalArgumentException) {
          var e = $p;
          unparsedPrimitive(this, 'boolean');
        } else {
          throw $p;
        }
      }
    }
    return tmp$ret$1;
  };
  protoOf(AbstractJsonTreeDecoder).nu = function (tag) {
    return this.g11((!(tag == null) ? typeof tag === 'string' : false) ? tag : THROW_CCE());
  };
  protoOf(AbstractJsonTreeDecoder).h11 = function (tag) {
    var tmp$ret$1;
    $l$block: {
      // Inline function 'kotlinx.serialization.json.internal.AbstractJsonTreeDecoder.primitive' call
      var this_0 = this.d11(tag);
      try {
        // Inline function 'kotlinx.serialization.json.internal.AbstractJsonTreeDecoder.decodeTaggedByte.<anonymous>' call
        var result = get_int(this_0);
        var tmp;
        var containsLower = ByteCompanionObject_instance.MIN_VALUE;
        if (result <= ByteCompanionObject_instance.MAX_VALUE ? containsLower <= result : false) {
          tmp = toByte(result);
        } else {
          tmp = null;
        }
        var tmp0_elvis_lhs = tmp;
        var tmp_0;
        if (tmp0_elvis_lhs == null) {
          unparsedPrimitive(this, 'byte');
        } else {
          tmp_0 = tmp0_elvis_lhs;
        }
        tmp$ret$1 = tmp_0;
        break $l$block;
      } catch ($p) {
        if ($p instanceof IllegalArgumentException) {
          var e = $p;
          unparsedPrimitive(this, 'byte');
        } else {
          throw $p;
        }
      }
    }
    return tmp$ret$1;
  };
  protoOf(AbstractJsonTreeDecoder).ou = function (tag) {
    return this.h11((!(tag == null) ? typeof tag === 'string' : false) ? tag : THROW_CCE());
  };
  protoOf(AbstractJsonTreeDecoder).i11 = function (tag) {
    var tmp$ret$1;
    $l$block: {
      // Inline function 'kotlinx.serialization.json.internal.AbstractJsonTreeDecoder.primitive' call
      var this_0 = this.d11(tag);
      try {
        // Inline function 'kotlinx.serialization.json.internal.AbstractJsonTreeDecoder.decodeTaggedShort.<anonymous>' call
        var result = get_int(this_0);
        var tmp;
        var containsLower = ShortCompanionObject_instance.MIN_VALUE;
        if (result <= ShortCompanionObject_instance.MAX_VALUE ? containsLower <= result : false) {
          tmp = toShort(result);
        } else {
          tmp = null;
        }
        var tmp0_elvis_lhs = tmp;
        var tmp_0;
        if (tmp0_elvis_lhs == null) {
          unparsedPrimitive(this, 'short');
        } else {
          tmp_0 = tmp0_elvis_lhs;
        }
        tmp$ret$1 = tmp_0;
        break $l$block;
      } catch ($p) {
        if ($p instanceof IllegalArgumentException) {
          var e = $p;
          unparsedPrimitive(this, 'short');
        } else {
          throw $p;
        }
      }
    }
    return tmp$ret$1;
  };
  protoOf(AbstractJsonTreeDecoder).pu = function (tag) {
    return this.i11((!(tag == null) ? typeof tag === 'string' : false) ? tag : THROW_CCE());
  };
  protoOf(AbstractJsonTreeDecoder).j11 = function (tag) {
    var tmp$ret$1;
    $l$block: {
      // Inline function 'kotlinx.serialization.json.internal.AbstractJsonTreeDecoder.primitive' call
      var this_0 = this.d11(tag);
      try {
        // Inline function 'kotlinx.serialization.json.internal.AbstractJsonTreeDecoder.decodeTaggedInt.<anonymous>' call
        var tmp0_elvis_lhs = get_int(this_0);
        var tmp;
        if (tmp0_elvis_lhs == null) {
          unparsedPrimitive(this, 'int');
        } else {
          tmp = tmp0_elvis_lhs;
        }
        tmp$ret$1 = tmp;
        break $l$block;
      } catch ($p) {
        if ($p instanceof IllegalArgumentException) {
          var e = $p;
          unparsedPrimitive(this, 'int');
        } else {
          throw $p;
        }
      }
    }
    return tmp$ret$1;
  };
  protoOf(AbstractJsonTreeDecoder).qu = function (tag) {
    return this.j11((!(tag == null) ? typeof tag === 'string' : false) ? tag : THROW_CCE());
  };
  protoOf(AbstractJsonTreeDecoder).k11 = function (tag) {
    var tmp$ret$1;
    $l$block: {
      // Inline function 'kotlinx.serialization.json.internal.AbstractJsonTreeDecoder.primitive' call
      var this_0 = this.d11(tag);
      try {
        // Inline function 'kotlinx.serialization.json.internal.AbstractJsonTreeDecoder.decodeTaggedLong.<anonymous>' call
        var tmp0_elvis_lhs = get_long(this_0);
        var tmp;
        if (tmp0_elvis_lhs == null) {
          unparsedPrimitive(this, 'long');
        } else {
          tmp = tmp0_elvis_lhs;
        }
        tmp$ret$1 = tmp;
        break $l$block;
      } catch ($p) {
        if ($p instanceof IllegalArgumentException) {
          var e = $p;
          unparsedPrimitive(this, 'long');
        } else {
          throw $p;
        }
      }
    }
    return tmp$ret$1;
  };
  protoOf(AbstractJsonTreeDecoder).ru = function (tag) {
    return this.k11((!(tag == null) ? typeof tag === 'string' : false) ? tag : THROW_CCE());
  };
  protoOf(AbstractJsonTreeDecoder).l11 = function (tag) {
    var tmp$ret$1;
    $l$block: {
      // Inline function 'kotlinx.serialization.json.internal.AbstractJsonTreeDecoder.primitive' call
      var this_0 = this.d11(tag);
      try {
        // Inline function 'kotlinx.serialization.json.internal.AbstractJsonTreeDecoder.decodeTaggedFloat.<anonymous>' call
        var tmp0_elvis_lhs = get_float(this_0);
        var tmp;
        if (tmp0_elvis_lhs == null) {
          unparsedPrimitive(this, 'float');
        } else {
          tmp = tmp0_elvis_lhs;
        }
        tmp$ret$1 = tmp;
        break $l$block;
      } catch ($p) {
        if ($p instanceof IllegalArgumentException) {
          var e = $p;
          unparsedPrimitive(this, 'float');
        } else {
          throw $p;
        }
      }
    }
    var result = tmp$ret$1;
    var specialFp = this.vw().pv_1.rw_1;
    if (specialFp ? true : isFinite(result))
      return result;
    throw InvalidFloatingPointDecoded(result, tag, toString(this.b11()));
  };
  protoOf(AbstractJsonTreeDecoder).su = function (tag) {
    return this.l11((!(tag == null) ? typeof tag === 'string' : false) ? tag : THROW_CCE());
  };
  protoOf(AbstractJsonTreeDecoder).m11 = function (tag) {
    var tmp$ret$1;
    $l$block: {
      // Inline function 'kotlinx.serialization.json.internal.AbstractJsonTreeDecoder.primitive' call
      var this_0 = this.d11(tag);
      try {
        // Inline function 'kotlinx.serialization.json.internal.AbstractJsonTreeDecoder.decodeTaggedDouble.<anonymous>' call
        var tmp0_elvis_lhs = get_double(this_0);
        var tmp;
        if (tmp0_elvis_lhs == null) {
          unparsedPrimitive(this, 'double');
        } else {
          tmp = tmp0_elvis_lhs;
        }
        tmp$ret$1 = tmp;
        break $l$block;
      } catch ($p) {
        if ($p instanceof IllegalArgumentException) {
          var e = $p;
          unparsedPrimitive(this, 'double');
        } else {
          throw $p;
        }
      }
    }
    var result = tmp$ret$1;
    var specialFp = this.vw().pv_1.rw_1;
    if (specialFp ? true : isFinite_0(result))
      return result;
    throw InvalidFloatingPointDecoded(result, tag, toString(this.b11()));
  };
  protoOf(AbstractJsonTreeDecoder).tu = function (tag) {
    return this.m11((!(tag == null) ? typeof tag === 'string' : false) ? tag : THROW_CCE());
  };
  protoOf(AbstractJsonTreeDecoder).n11 = function (tag) {
    var tmp$ret$1;
    $l$block: {
      // Inline function 'kotlinx.serialization.json.internal.AbstractJsonTreeDecoder.primitive' call
      var this_0 = this.d11(tag);
      try {
        // Inline function 'kotlinx.serialization.json.internal.AbstractJsonTreeDecoder.decodeTaggedChar.<anonymous>' call
        var tmp0_elvis_lhs = single(this_0.zw());
        var tmp;
        var tmp_0 = tmp0_elvis_lhs;
        if ((tmp_0 == null ? null : new Char(tmp_0)) == null) {
          unparsedPrimitive(this, 'char');
        } else {
          tmp = tmp0_elvis_lhs;
        }
        tmp$ret$1 = tmp;
        break $l$block;
      } catch ($p) {
        if ($p instanceof IllegalArgumentException) {
          var e = $p;
          unparsedPrimitive(this, 'char');
        } else {
          throw $p;
        }
      }
    }
    return tmp$ret$1;
  };
  protoOf(AbstractJsonTreeDecoder).uu = function (tag) {
    return this.n11((!(tag == null) ? typeof tag === 'string' : false) ? tag : THROW_CCE());
  };
  protoOf(AbstractJsonTreeDecoder).o11 = function (tag) {
    var value = this.d11(tag);
    if (!this.vw().pv_1.jw_1) {
      var literal = asLiteral(value, this, 'string');
      if (!literal.fx_1)
        throw JsonDecodingException_0(-1, "String literal for key '" + tag + "' should be quoted.\n" + get_lenientHint(), toString(this.b11()));
    }
    if (value instanceof JsonNull)
      throw JsonDecodingException_0(-1, "Unexpected 'null' value instead of string literal", toString(this.b11()));
    return value.zw();
  };
  protoOf(AbstractJsonTreeDecoder).vu = function (tag) {
    return this.o11((!(tag == null) ? typeof tag === 'string' : false) ? tag : THROW_CCE());
  };
  protoOf(AbstractJsonTreeDecoder).p11 = function (tag, inlineDescriptor) {
    return get_isUnsignedNumber(inlineDescriptor) ? new JsonDecoderForUnsignedTypes(new StringJsonLexer(this.d11(tag).zw()), this.vw()) : protoOf(NamedValueDecoder).xu.call(this, tag, inlineDescriptor);
  };
  protoOf(AbstractJsonTreeDecoder).xu = function (tag, inlineDescriptor) {
    return this.p11((!(tag == null) ? typeof tag === 'string' : false) ? tag : THROW_CCE(), inlineDescriptor);
  };
  protoOf(AbstractJsonTreeDecoder).ok = function (descriptor) {
    return !(this.ju() == null) ? protoOf(NamedValueDecoder).ok.call(this, descriptor) : (new JsonPrimitiveDecoder(this.vw(), this.v1())).ok(descriptor);
  };
  function coerceInputValue_0($this, descriptor, index, tag) {
    var tmp$ret$1;
    $l$block_2: {
      // Inline function 'kotlinx.serialization.json.internal.tryCoerceValue' call
      var this_0 = $this.vw();
      var elementDescriptor = descriptor.wi(index);
      var tmp;
      if (!elementDescriptor.ii()) {
        // Inline function 'kotlinx.serialization.json.internal.JsonTreeDecoder.coerceInputValue.<anonymous>' call
        var tmp_0 = $this.c11(tag);
        tmp = tmp_0 instanceof JsonNull;
      } else {
        tmp = false;
      }
      if (tmp) {
        tmp$ret$1 = true;
        break $l$block_2;
      }
      if (equals(elementDescriptor.pi(), ENUM_getInstance())) {
        var tmp_1;
        if (elementDescriptor.ii()) {
          // Inline function 'kotlinx.serialization.json.internal.JsonTreeDecoder.coerceInputValue.<anonymous>' call
          var tmp_2 = $this.c11(tag);
          tmp_1 = tmp_2 instanceof JsonNull;
        } else {
          tmp_1 = false;
        }
        if (tmp_1) {
          tmp$ret$1 = false;
          break $l$block_2;
        }
        // Inline function 'kotlinx.serialization.json.internal.JsonTreeDecoder.coerceInputValue.<anonymous>' call
        var tmp_3 = $this.c11(tag);
        var tmp0_safe_receiver = tmp_3 instanceof JsonPrimitive ? tmp_3 : null;
        var tmp0_elvis_lhs = tmp0_safe_receiver == null ? null : get_contentOrNull(tmp0_safe_receiver);
        var tmp_4;
        if (tmp0_elvis_lhs == null) {
          tmp$ret$1 = false;
          break $l$block_2;
        } else {
          tmp_4 = tmp0_elvis_lhs;
        }
        var enumValue = tmp_4;
        var enumIndex = getJsonNameIndex(elementDescriptor, this_0, enumValue);
        if (enumIndex === -3) {
          // Inline function 'kotlinx.serialization.json.internal.tryCoerceValue.<anonymous>' call
          tmp$ret$1 = true;
          break $l$block_2;
        }
      }
      tmp$ret$1 = false;
    }
    return tmp$ret$1;
  }
  function absenceIsNull($this, descriptor, index) {
    $this.z11_1 = (!$this.vw().pv_1.mw_1 ? !descriptor.xi(index) : false) ? descriptor.wi(index).ii() : false;
    return $this.z11_1;
  }
  function JsonTreeDecoder(json, value, polyDiscriminator, polyDescriptor) {
    polyDiscriminator = polyDiscriminator === VOID ? null : polyDiscriminator;
    polyDescriptor = polyDescriptor === VOID ? null : polyDescriptor;
    AbstractJsonTreeDecoder.call(this, json, value);
    this.v11_1 = value;
    this.w11_1 = polyDiscriminator;
    this.x11_1 = polyDescriptor;
    this.y11_1 = 0;
    this.z11_1 = false;
  }
  protoOf(JsonTreeDecoder).v1 = function () {
    return this.v11_1;
  };
  protoOf(JsonTreeDecoder).hl = function (descriptor) {
    while (this.y11_1 < descriptor.ri()) {
      var tmp1 = this.y11_1;
      this.y11_1 = tmp1 + 1 | 0;
      var name = this.eu(descriptor, tmp1);
      var index = this.y11_1 - 1 | 0;
      this.z11_1 = false;
      var tmp;
      var tmp_0;
      // Inline function 'kotlin.collections.contains' call
      // Inline function 'kotlin.collections.containsKey' call
      var this_0 = this.v1();
      if ((isInterface(this_0, Map) ? this_0 : THROW_CCE()).y1(name)) {
        tmp_0 = true;
      } else {
        tmp_0 = absenceIsNull(this, descriptor, index);
      }
      if (tmp_0) {
        tmp = !this.a11_1.ow_1 ? true : !coerceInputValue_0(this, descriptor, index, name);
      } else {
        tmp = false;
      }
      if (tmp) {
        return index;
      }
    }
    return -1;
  };
  protoOf(JsonTreeDecoder).ck = function () {
    return !this.z11_1 ? protoOf(AbstractJsonTreeDecoder).ck.call(this) : false;
  };
  protoOf(JsonTreeDecoder).fu = function (descriptor, index) {
    var strategy = namingStrategy(descriptor, this.vw());
    var baseName = descriptor.ti(index);
    if (strategy == null) {
      if (!this.a11_1.sw_1)
        return baseName;
      if (this.v1().p1().v(baseName))
        return baseName;
    }
    var deserializationNamesMap_0 = deserializationNamesMap(this.vw(), descriptor);
    // Inline function 'kotlin.collections.find' call
    var this_0 = this.v1().p1();
    var tmp$ret$1;
    $l$block: {
      // Inline function 'kotlin.collections.firstOrNull' call
      var tmp0_iterator = this_0.r();
      while (tmp0_iterator.s()) {
        var element = tmp0_iterator.t();
        // Inline function 'kotlinx.serialization.json.internal.JsonTreeDecoder.elementName.<anonymous>' call
        if (deserializationNamesMap_0.b2(element) === index) {
          tmp$ret$1 = element;
          break $l$block;
        }
      }
      tmp$ret$1 = null;
    }
    var tmp0_safe_receiver = tmp$ret$1;
    if (tmp0_safe_receiver == null)
      null;
    else {
      // Inline function 'kotlin.let' call
      // Inline function 'kotlin.contracts.contract' call
      return tmp0_safe_receiver;
    }
    var fallbackName = strategy == null ? null : strategy.dy(descriptor, index, baseName);
    return fallbackName == null ? baseName : fallbackName;
  };
  protoOf(JsonTreeDecoder).c11 = function (tag) {
    return getValue(this.v1(), tag);
  };
  protoOf(JsonTreeDecoder).rk = function (descriptor) {
    if (descriptor === this.x11_1) {
      var tmp = this.vw();
      // Inline function 'kotlinx.serialization.json.internal.cast' call
      var value = this.b11();
      var descriptor_0 = this.x11_1;
      if (!(value instanceof JsonObject)) {
        throw JsonDecodingException_1(-1, 'Expected ' + getKClass(JsonObject) + ' as the serialized body of ' + descriptor_0.oi() + ', but had ' + getKClassFromExpression(value));
      }
      return new JsonTreeDecoder(tmp, value, this.w11_1, this.x11_1);
    }
    return protoOf(AbstractJsonTreeDecoder).rk.call(this, descriptor);
  };
  protoOf(JsonTreeDecoder).sk = function (descriptor) {
    var tmp;
    if (this.a11_1.iw_1) {
      tmp = true;
    } else {
      var tmp_0 = descriptor.pi();
      tmp = tmp_0 instanceof PolymorphicKind;
    }
    if (tmp)
      return Unit_instance;
    var strategy = namingStrategy(descriptor, this.vw());
    var tmp_1;
    if (strategy == null ? !this.a11_1.sw_1 : false) {
      tmp_1 = jsonCachedSerialNames(descriptor);
    } else if (!(strategy == null)) {
      tmp_1 = deserializationNamesMap(this.vw(), descriptor).p1();
    } else {
      var tmp_2 = jsonCachedSerialNames(descriptor);
      // Inline function 'kotlin.collections.orEmpty' call
      var tmp0_safe_receiver = get_schemaCache(this.vw()).c10(descriptor, get_JsonDeserializationNamesKey());
      var tmp0_elvis_lhs = tmp0_safe_receiver == null ? null : tmp0_safe_receiver.p1();
      var tmp$ret$0 = tmp0_elvis_lhs == null ? emptySet() : tmp0_elvis_lhs;
      tmp_1 = plus_0(tmp_2, tmp$ret$0);
    }
    var names = tmp_1;
    var tmp1_iterator = this.v1().p1().r();
    while (tmp1_iterator.s()) {
      var key = tmp1_iterator.t();
      if (!names.v(key) ? !(key === this.w11_1) : false) {
        throw UnknownKeyException(key, this.v1().toString());
      }
    }
  };
  function JsonTreeListDecoder(json, value) {
    AbstractJsonTreeDecoder.call(this, json, value);
    this.f12_1 = value;
    this.g12_1 = this.f12_1.u();
    this.h12_1 = -1;
  }
  protoOf(JsonTreeListDecoder).v1 = function () {
    return this.f12_1;
  };
  protoOf(JsonTreeListDecoder).fu = function (descriptor, index) {
    return index.toString();
  };
  protoOf(JsonTreeListDecoder).c11 = function (tag) {
    return this.f12_1.b1(toInt(tag));
  };
  protoOf(JsonTreeListDecoder).hl = function (descriptor) {
    while (this.h12_1 < (this.g12_1 - 1 | 0)) {
      this.h12_1 = this.h12_1 + 1 | 0;
      return this.h12_1;
    }
    return -1;
  };
  function JsonPrimitiveDecoder(json, value) {
    AbstractJsonTreeDecoder.call(this, json, value);
    this.n12_1 = value;
    this.yu(get_PRIMITIVE_TAG());
  }
  protoOf(JsonPrimitiveDecoder).v1 = function () {
    return this.n12_1;
  };
  protoOf(JsonPrimitiveDecoder).hl = function (descriptor) {
    return 0;
  };
  protoOf(JsonPrimitiveDecoder).c11 = function (tag) {
    // Inline function 'kotlin.require' call
    // Inline function 'kotlin.contracts.contract' call
    if (!(tag === get_PRIMITIVE_TAG())) {
      // Inline function 'kotlinx.serialization.json.internal.JsonPrimitiveDecoder.currentElement.<anonymous>' call
      var message = "This input can only handle primitives with '" + get_PRIMITIVE_TAG() + "' tag";
      throw IllegalArgumentException_init_$Create$(toString(message));
    }
    return this.n12_1;
  };
  function JsonTreeMapDecoder(json, value) {
    JsonTreeDecoder.call(this, json, value);
    this.y12_1 = value;
    this.z12_1 = toList(this.y12_1.p1());
    this.a13_1 = imul(this.z12_1.u(), 2);
    this.b13_1 = -1;
  }
  protoOf(JsonTreeMapDecoder).v1 = function () {
    return this.y12_1;
  };
  protoOf(JsonTreeMapDecoder).fu = function (descriptor, index) {
    var i = index / 2 | 0;
    return this.z12_1.b1(i);
  };
  protoOf(JsonTreeMapDecoder).hl = function (descriptor) {
    while (this.b13_1 < (this.a13_1 - 1 | 0)) {
      this.b13_1 = this.b13_1 + 1 | 0;
      return this.b13_1;
    }
    return -1;
  };
  protoOf(JsonTreeMapDecoder).c11 = function (tag) {
    return (this.b13_1 % 2 | 0) === 0 ? JsonPrimitive_1(tag) : getValue(this.y12_1, tag);
  };
  protoOf(JsonTreeMapDecoder).sk = function (descriptor) {
  };
  function readPolymorphicJson(_this__u8e3s4, discriminator, element, deserializer) {
    return (new JsonTreeDecoder(_this__u8e3s4, element, discriminator, deserializer.fi())).qk(deserializer);
  }
  function get_PRIMITIVE_TAG() {
    return PRIMITIVE_TAG;
  }
  var PRIMITIVE_TAG;
  var WriteMode_OBJ_instance;
  var WriteMode_LIST_instance;
  var WriteMode_MAP_instance;
  var WriteMode_POLY_OBJ_instance;
  var WriteMode_entriesInitialized;
  function WriteMode_initEntries() {
    if (WriteMode_entriesInitialized)
      return Unit_instance;
    WriteMode_entriesInitialized = true;
    WriteMode_OBJ_instance = new WriteMode('OBJ', 0, get_BEGIN_OBJ(), get_END_OBJ());
    WriteMode_LIST_instance = new WriteMode('LIST', 1, get_BEGIN_LIST(), get_END_LIST());
    WriteMode_MAP_instance = new WriteMode('MAP', 2, get_BEGIN_OBJ(), get_END_OBJ());
    WriteMode_POLY_OBJ_instance = new WriteMode('POLY_OBJ', 3, get_BEGIN_LIST(), get_END_LIST());
  }
  function WriteMode(name, ordinal, begin, end) {
    Enum.call(this, name, ordinal);
    this.p10_1 = begin;
    this.q10_1 = end;
  }
  function switchMode(_this__u8e3s4, desc) {
    var tmp0_subject = desc.pi();
    var tmp;
    if (tmp0_subject instanceof PolymorphicKind) {
      tmp = WriteMode_POLY_OBJ_getInstance();
    } else {
      if (equals(tmp0_subject, LIST_getInstance())) {
        tmp = WriteMode_LIST_getInstance();
      } else {
        if (equals(tmp0_subject, MAP_getInstance())) {
          // Inline function 'kotlinx.serialization.json.internal.selectMapMode' call
          var keyDescriptor = carrierDescriptor(desc.wi(0), _this__u8e3s4.fl());
          var keyKind = keyDescriptor.pi();
          var tmp_0;
          var tmp_1;
          if (keyKind instanceof PrimitiveKind) {
            tmp_1 = true;
          } else {
            tmp_1 = equals(keyKind, ENUM_getInstance());
          }
          if (tmp_1) {
            // Inline function 'kotlinx.serialization.json.internal.switchMode.<anonymous>' call
            tmp_0 = WriteMode_MAP_getInstance();
          } else {
            if (_this__u8e3s4.pv_1.kw_1) {
              // Inline function 'kotlinx.serialization.json.internal.switchMode.<anonymous>' call
              tmp_0 = WriteMode_LIST_getInstance();
            } else {
              throw InvalidKeyKindException(keyDescriptor);
            }
          }
          tmp = tmp_0;
        } else {
          tmp = WriteMode_OBJ_getInstance();
        }
      }
    }
    return tmp;
  }
  function carrierDescriptor(_this__u8e3s4, module_0) {
    var tmp;
    if (equals(_this__u8e3s4.pi(), CONTEXTUAL_getInstance())) {
      var tmp0_safe_receiver = getContextualDescriptor(module_0, _this__u8e3s4);
      var tmp1_elvis_lhs = tmp0_safe_receiver == null ? null : carrierDescriptor(tmp0_safe_receiver, module_0);
      tmp = tmp1_elvis_lhs == null ? _this__u8e3s4 : tmp1_elvis_lhs;
    } else if (_this__u8e3s4.qi()) {
      tmp = carrierDescriptor(_this__u8e3s4.wi(0), module_0);
    } else {
      tmp = _this__u8e3s4;
    }
    return tmp;
  }
  function WriteMode_OBJ_getInstance() {
    WriteMode_initEntries();
    return WriteMode_OBJ_instance;
  }
  function WriteMode_LIST_getInstance() {
    WriteMode_initEntries();
    return WriteMode_LIST_instance;
  }
  function WriteMode_MAP_getInstance() {
    WriteMode_initEntries();
    return WriteMode_MAP_instance;
  }
  function WriteMode_POLY_OBJ_getInstance() {
    WriteMode_initEntries();
    return WriteMode_POLY_OBJ_instance;
  }
  function get_STRING() {
    return STRING;
  }
  var STRING;
  function appendEscape($this, lastPosition, current) {
    $this.c13(lastPosition, current);
    return appendEsc($this, current + 1 | 0);
  }
  function decodedString($this, lastPosition, currentPosition) {
    $this.c13(lastPosition, currentPosition);
    var result = $this.fw_1.toString();
    $this.fw_1.m7(0);
    return result;
  }
  function takePeeked($this) {
    // Inline function 'kotlin.also' call
    var this_0 = ensureNotNull($this.ew_1);
    // Inline function 'kotlin.contracts.contract' call
    // Inline function 'kotlinx.serialization.json.internal.AbstractJsonLexer.takePeeked.<anonymous>' call
    $this.ew_1 = null;
    return this_0;
  }
  function wasUnquotedString($this) {
    return !(charSequenceGet($this.d13(), $this.cw_1 - 1 | 0) === _Char___init__impl__6a9atx(34));
  }
  function appendEsc($this, startPosition) {
    var currentPosition = startPosition;
    currentPosition = $this.e13(currentPosition);
    if (currentPosition === -1) {
      $this.zx('Expected escape sequence to continue, got EOF');
    }
    var tmp = $this.d13();
    var tmp0 = currentPosition;
    currentPosition = tmp0 + 1 | 0;
    var currentChar = charSequenceGet(tmp, tmp0);
    if (currentChar === _Char___init__impl__6a9atx(117)) {
      return appendHex($this, $this.d13(), currentPosition);
    }
    // Inline function 'kotlin.code' call
    var tmp$ret$0 = Char__toInt_impl_vasixd(currentChar);
    var c = escapeToChar(tmp$ret$0);
    if (c === _Char___init__impl__6a9atx(0)) {
      $this.zx("Invalid escaped char '" + toString_0(currentChar) + "'");
    }
    $this.fw_1.a5(c);
    return currentPosition;
  }
  function appendHex($this, source, startPos) {
    if ((startPos + 4 | 0) >= charSequenceLength(source)) {
      $this.cw_1 = startPos;
      $this.f13();
      if (($this.cw_1 + 4 | 0) >= charSequenceLength(source)) {
        $this.zx('Unexpected EOF during unicode escape');
      }
      return appendHex($this, source, $this.cw_1);
    }
    $this.fw_1.a5(numberToChar((((fromHexChar($this, source, startPos) << 12) + (fromHexChar($this, source, startPos + 1 | 0) << 8) | 0) + (fromHexChar($this, source, startPos + 2 | 0) << 4) | 0) + fromHexChar($this, source, startPos + 3 | 0) | 0));
    return startPos + 4 | 0;
  }
  function fromHexChar($this, source, currentPosition) {
    var character = charSequenceGet(source, currentPosition);
    var tmp;
    if (_Char___init__impl__6a9atx(48) <= character ? character <= _Char___init__impl__6a9atx(57) : false) {
      // Inline function 'kotlin.code' call
      var tmp_0 = Char__toInt_impl_vasixd(character);
      // Inline function 'kotlin.code' call
      var this_0 = _Char___init__impl__6a9atx(48);
      tmp = tmp_0 - Char__toInt_impl_vasixd(this_0) | 0;
    } else if (_Char___init__impl__6a9atx(97) <= character ? character <= _Char___init__impl__6a9atx(102) : false) {
      // Inline function 'kotlin.code' call
      var tmp_1 = Char__toInt_impl_vasixd(character);
      // Inline function 'kotlin.code' call
      var this_1 = _Char___init__impl__6a9atx(97);
      tmp = (tmp_1 - Char__toInt_impl_vasixd(this_1) | 0) + 10 | 0;
    } else if (_Char___init__impl__6a9atx(65) <= character ? character <= _Char___init__impl__6a9atx(70) : false) {
      // Inline function 'kotlin.code' call
      var tmp_2 = Char__toInt_impl_vasixd(character);
      // Inline function 'kotlin.code' call
      var this_2 = _Char___init__impl__6a9atx(65);
      tmp = (tmp_2 - Char__toInt_impl_vasixd(this_2) | 0) + 10 | 0;
    } else {
      $this.zx("Invalid toHexChar char '" + toString_0(character) + "' in unicode escape");
    }
    return tmp;
  }
  function consumeBoolean($this, start) {
    var current = $this.e13(start);
    if (current >= charSequenceLength($this.d13()) ? true : current === -1) {
      $this.zx('EOF');
    }
    // Inline function 'kotlin.code' call
    var tmp = $this.d13();
    var tmp0 = current;
    current = tmp0 + 1 | 0;
    var this_0 = charSequenceGet(tmp, tmp0);
    var tmp1_subject = Char__toInt_impl_vasixd(this_0) | 32;
    var tmp_0;
    // Inline function 'kotlin.code' call
    var this_1 = _Char___init__impl__6a9atx(116);
    if (tmp1_subject === Char__toInt_impl_vasixd(this_1)) {
      consumeBooleanLiteral($this, 'rue', current);
      tmp_0 = true;
    } else {
      // Inline function 'kotlin.code' call
      var this_2 = _Char___init__impl__6a9atx(102);
      if (tmp1_subject === Char__toInt_impl_vasixd(this_2)) {
        consumeBooleanLiteral($this, 'alse', current);
        tmp_0 = false;
      } else {
        $this.zx("Expected valid boolean literal prefix, but had '" + $this.uy() + "'");
      }
    }
    return tmp_0;
  }
  function consumeBooleanLiteral($this, literalSuffix, current) {
    if ((charSequenceLength($this.d13()) - current | 0) < literalSuffix.length) {
      $this.zx('Unexpected end of boolean literal');
    }
    var inductionVariable = 0;
    var last = charSequenceLength(literalSuffix) - 1 | 0;
    if (inductionVariable <= last)
      do {
        var i = inductionVariable;
        inductionVariable = inductionVariable + 1 | 0;
        var expected = charSequenceGet(literalSuffix, i);
        var actual = charSequenceGet($this.d13(), current + i | 0);
        // Inline function 'kotlin.code' call
        var tmp = Char__toInt_impl_vasixd(expected);
        // Inline function 'kotlin.code' call
        if (!(tmp === (Char__toInt_impl_vasixd(actual) | 32))) {
          $this.zx("Expected valid boolean literal prefix, but had '" + $this.uy() + "'");
        }
      }
       while (inductionVariable <= last);
    $this.cw_1 = current + literalSuffix.length | 0;
  }
  function consumeNumericLiteral$calculateExponent(exponentAccumulator, isExponentPositive) {
    var tmp;
    switch (isExponentPositive) {
      case false:
        // Inline function 'kotlin.math.pow' call

        var x = -exponentAccumulator.a6();
        tmp = Math.pow(10.0, x);
        break;
      case true:
        // Inline function 'kotlin.math.pow' call

        var x_0 = exponentAccumulator.a6();
        tmp = Math.pow(10.0, x_0);
        break;
      default:
        noWhenBranchMatchedException();
        break;
    }
    return tmp;
  }
  function AbstractJsonLexer() {
    this.cw_1 = 0;
    this.dw_1 = new JsonPath();
    this.ew_1 = null;
    this.fw_1 = StringBuilder_init_$Create$();
  }
  protoOf(AbstractJsonLexer).f13 = function () {
  };
  protoOf(AbstractJsonLexer).g13 = function (c) {
    return (((c === _Char___init__impl__6a9atx(125) ? true : c === _Char___init__impl__6a9atx(93)) ? true : c === _Char___init__impl__6a9atx(58)) ? true : c === _Char___init__impl__6a9atx(44)) ? false : true;
  };
  protoOf(AbstractJsonLexer).gw = function () {
    var nextToken = this.wy();
    if (!(nextToken === 10)) {
      this.zx('Expected EOF after parsing, but had ' + toString_0(charSequenceGet(this.d13(), this.cw_1 - 1 | 0)) + ' instead');
    }
  };
  protoOf(AbstractJsonLexer).qy = function (expected) {
    var token = this.wy();
    if (!(token === expected)) {
      this.h13(expected);
    }
    return token;
  };
  protoOf(AbstractJsonLexer).e10 = function (expected) {
    this.f13();
    var source = this.d13();
    var cpos = this.cw_1;
    $l$loop_0: while (true) {
      cpos = this.e13(cpos);
      if (cpos === -1)
        break $l$loop_0;
      var tmp0 = cpos;
      cpos = tmp0 + 1 | 0;
      var c = charSequenceGet(source, tmp0);
      if (((c === _Char___init__impl__6a9atx(32) ? true : c === _Char___init__impl__6a9atx(10)) ? true : c === _Char___init__impl__6a9atx(13)) ? true : c === _Char___init__impl__6a9atx(9))
        continue $l$loop_0;
      this.cw_1 = cpos;
      if (c === expected)
        return Unit_instance;
      this.i13(expected);
    }
    this.cw_1 = cpos;
    this.i13(expected);
  };
  protoOf(AbstractJsonLexer).i13 = function (expected) {
    if (this.cw_1 > 0 ? expected === _Char___init__impl__6a9atx(34) : false) {
      var tmp$ret$1;
      $l$block: {
        // Inline function 'kotlinx.serialization.json.internal.AbstractJsonLexer.withPositionRollback' call
        var snapshot = this.cw_1;
        try {
          // Inline function 'kotlinx.serialization.json.internal.AbstractJsonLexer.unexpectedToken.<anonymous>' call
          this.cw_1 = this.cw_1 - 1 | 0;
          tmp$ret$1 = this.uy();
          break $l$block;
        }finally {
          this.cw_1 = snapshot;
        }
      }
      var inputLiteral = tmp$ret$1;
      if (inputLiteral === 'null') {
        this.j13("Expected string literal but 'null' literal was found", this.cw_1 - 1 | 0, "Use 'coerceInputValues = true' in 'Json {}' builder to coerce nulls to default values.");
      }
    }
    this.h13(charToTokenClass(expected));
  };
  protoOf(AbstractJsonLexer).k13 = function (expectedToken, wasConsumed) {
    var expected = tokenDescription(expectedToken);
    var position = wasConsumed ? this.cw_1 - 1 | 0 : this.cw_1;
    var s = (this.cw_1 === charSequenceLength(this.d13()) ? true : position < 0) ? 'EOF' : toString_0(charSequenceGet(this.d13(), position));
    this.zx('Expected ' + expected + ", but had '" + s + "' instead", position);
  };
  protoOf(AbstractJsonLexer).h13 = function (expectedToken, wasConsumed, $super) {
    wasConsumed = wasConsumed === VOID ? true : wasConsumed;
    return $super === VOID ? this.k13(expectedToken, wasConsumed) : $super.k13.call(this, expectedToken, wasConsumed);
  };
  protoOf(AbstractJsonLexer).ry = function () {
    var source = this.d13();
    var cpos = this.cw_1;
    $l$loop_0: while (true) {
      cpos = this.e13(cpos);
      if (cpos === -1)
        break $l$loop_0;
      var ch = charSequenceGet(source, cpos);
      if (((ch === _Char___init__impl__6a9atx(32) ? true : ch === _Char___init__impl__6a9atx(10)) ? true : ch === _Char___init__impl__6a9atx(13)) ? true : ch === _Char___init__impl__6a9atx(9)) {
        cpos = cpos + 1 | 0;
        continue $l$loop_0;
      }
      this.cw_1 = cpos;
      return charToTokenClass(ch);
    }
    this.cw_1 = cpos;
    return 10;
  };
  protoOf(AbstractJsonLexer).g10 = function (doConsume) {
    var current = this.l13();
    current = this.e13(current);
    var len = charSequenceLength(this.d13()) - current | 0;
    if (len < 4 ? true : current === -1)
      return false;
    var inductionVariable = 0;
    if (inductionVariable <= 3)
      do {
        var i = inductionVariable;
        inductionVariable = inductionVariable + 1 | 0;
        if (!(charSequenceGet('null', i) === charSequenceGet(this.d13(), current + i | 0)))
          return false;
      }
       while (inductionVariable <= 3);
    if (len > 4 ? charToTokenClass(charSequenceGet(this.d13(), current + 4 | 0)) === 0 : false)
      return false;
    if (doConsume) {
      this.cw_1 = current + 4 | 0;
    }
    return true;
  };
  protoOf(AbstractJsonLexer).r10 = function (doConsume, $super) {
    doConsume = doConsume === VOID ? true : doConsume;
    return $super === VOID ? this.g10(doConsume) : $super.g10.call(this, doConsume);
  };
  protoOf(AbstractJsonLexer).l13 = function () {
    var current = this.cw_1;
    $l$loop_0: while (true) {
      current = this.e13(current);
      if (current === -1)
        break $l$loop_0;
      var c = charSequenceGet(this.d13(), current);
      if (((c === _Char___init__impl__6a9atx(32) ? true : c === _Char___init__impl__6a9atx(10)) ? true : c === _Char___init__impl__6a9atx(13)) ? true : c === _Char___init__impl__6a9atx(9)) {
        current = current + 1 | 0;
      } else {
        break $l$loop_0;
      }
    }
    this.cw_1 = current;
    return current;
  };
  protoOf(AbstractJsonLexer).h10 = function (isLenient) {
    var token = this.ry();
    var tmp;
    if (isLenient) {
      if (!(token === 1) ? !(token === 0) : false)
        return null;
      tmp = this.uy();
    } else {
      if (!(token === 1))
        return null;
      tmp = this.ty();
    }
    var string = tmp;
    this.ew_1 = string;
    return string;
  };
  protoOf(AbstractJsonLexer).m13 = function () {
    this.ew_1 = null;
  };
  protoOf(AbstractJsonLexer).n13 = function (startPos, endPos) {
    // Inline function 'kotlin.text.substring' call
    var this_0 = this.d13();
    return toString(charSequenceSubSequence(this_0, startPos, endPos));
  };
  protoOf(AbstractJsonLexer).ty = function () {
    if (!(this.ew_1 == null)) {
      return takePeeked(this);
    }
    return this.k10();
  };
  protoOf(AbstractJsonLexer).consumeString2 = function (source, startPosition, current) {
    var currentPosition = current;
    var lastPosition = startPosition;
    var char = charSequenceGet(source, currentPosition);
    var usedAppend = false;
    while (!(char === _Char___init__impl__6a9atx(34))) {
      if (char === _Char___init__impl__6a9atx(92)) {
        usedAppend = true;
        currentPosition = this.e13(appendEscape(this, lastPosition, currentPosition));
        if (currentPosition === -1) {
          this.zx('Unexpected EOF', currentPosition);
        }
        lastPosition = currentPosition;
      } else {
        currentPosition = currentPosition + 1 | 0;
        if (currentPosition >= charSequenceLength(source)) {
          usedAppend = true;
          this.c13(lastPosition, currentPosition);
          currentPosition = this.e13(currentPosition);
          if (currentPosition === -1) {
            this.zx('Unexpected EOF', currentPosition);
          }
          lastPosition = currentPosition;
        }
      }
      char = charSequenceGet(source, currentPosition);
    }
    var tmp;
    if (!usedAppend) {
      tmp = this.n13(lastPosition, currentPosition);
    } else {
      tmp = decodedString(this, lastPosition, currentPosition);
    }
    var string = tmp;
    this.cw_1 = currentPosition + 1 | 0;
    return string;
  };
  protoOf(AbstractJsonLexer).l10 = function () {
    var result = this.uy();
    if (result === 'null' ? wasUnquotedString(this) : false) {
      this.zx("Unexpected 'null' value instead of string literal");
    }
    return result;
  };
  protoOf(AbstractJsonLexer).uy = function () {
    if (!(this.ew_1 == null)) {
      return takePeeked(this);
    }
    var current = this.l13();
    if (current >= charSequenceLength(this.d13()) ? true : current === -1) {
      this.zx('EOF', current);
    }
    var token = charToTokenClass(charSequenceGet(this.d13(), current));
    if (token === 1) {
      return this.ty();
    }
    if (!(token === 0)) {
      this.zx('Expected beginning of the string, but got ' + toString_0(charSequenceGet(this.d13(), current)));
    }
    var usedAppend = false;
    while (charToTokenClass(charSequenceGet(this.d13(), current)) === 0) {
      current = current + 1 | 0;
      if (current >= charSequenceLength(this.d13())) {
        usedAppend = true;
        this.c13(this.cw_1, current);
        var eof = this.e13(current);
        if (eof === -1) {
          this.cw_1 = current;
          return decodedString(this, 0, 0);
        } else {
          current = eof;
        }
      }
    }
    var tmp;
    if (!usedAppend) {
      tmp = this.n13(this.cw_1, current);
    } else {
      tmp = decodedString(this, this.cw_1, current);
    }
    var result = tmp;
    this.cw_1 = current;
    return result;
  };
  protoOf(AbstractJsonLexer).c13 = function (fromIndex, toIndex) {
    this.fw_1.h7(this.d13(), fromIndex, toIndex);
  };
  protoOf(AbstractJsonLexer).j10 = function (allowLenientStrings) {
    // Inline function 'kotlin.collections.mutableListOf' call
    var tokenStack = ArrayList_init_$Create$();
    var lastToken = this.ry();
    if (!(lastToken === 8) ? !(lastToken === 6) : false) {
      this.uy();
      return Unit_instance;
    }
    $l$loop: while (true) {
      lastToken = this.ry();
      if (lastToken === 1) {
        if (allowLenientStrings) {
          this.uy();
        } else {
          this.k10();
        }
        continue $l$loop;
      }
      var tmp0_subject = lastToken;
      if (tmp0_subject === 8 ? true : tmp0_subject === 6) {
        tokenStack.o(lastToken);
      } else if (tmp0_subject === 9) {
        if (!(last(tokenStack) === 8))
          throw JsonDecodingException_0(this.cw_1, 'found ] instead of } at path: ' + this.dw_1, this.d13());
        removeLast(tokenStack);
      } else if (tmp0_subject === 7) {
        if (!(last(tokenStack) === 6))
          throw JsonDecodingException_0(this.cw_1, 'found } instead of ] at path: ' + this.dw_1, this.d13());
        removeLast(tokenStack);
      } else if (tmp0_subject === 10) {
        this.zx('Unexpected end of input due to malformed JSON during ignoring unknown keys');
      }
      this.wy();
      if (tokenStack.u() === 0)
        return Unit_instance;
    }
  };
  protoOf(AbstractJsonLexer).toString = function () {
    return "JsonReader(source='" + this.d13() + "', currentPosition=" + this.cw_1 + ')';
  };
  protoOf(AbstractJsonLexer).i10 = function (key) {
    var processed = this.n13(0, this.cw_1);
    var lastIndexOf_0 = lastIndexOf(processed, key);
    this.j13("Encountered an unknown key '" + key + "'", lastIndexOf_0, "Use 'ignoreUnknownKeys = true' in 'Json {}' builder to ignore unknown keys.");
  };
  protoOf(AbstractJsonLexer).j13 = function (message, position, hint) {
    var tmp;
    // Inline function 'kotlin.text.isEmpty' call
    if (charSequenceLength(hint) === 0) {
      tmp = '';
    } else {
      tmp = '\n' + hint;
    }
    var hintMessage = tmp;
    throw JsonDecodingException_0(position, message + ' at path: ' + this.dw_1.my() + hintMessage, this.d13());
  };
  protoOf(AbstractJsonLexer).zx = function (message, position, hint, $super) {
    position = position === VOID ? this.cw_1 : position;
    hint = hint === VOID ? '' : hint;
    return $super === VOID ? this.j13(message, position, hint) : $super.j13.call(this, message, position, hint);
  };
  protoOf(AbstractJsonLexer).ex = function () {
    var current = this.l13();
    current = this.e13(current);
    if (current >= charSequenceLength(this.d13()) ? true : current === -1) {
      this.zx('EOF');
    }
    var tmp;
    if (charSequenceGet(this.d13(), current) === _Char___init__impl__6a9atx(34)) {
      current = current + 1 | 0;
      if (current === charSequenceLength(this.d13())) {
        this.zx('EOF');
      }
      tmp = true;
    } else {
      tmp = false;
    }
    var hasQuotation = tmp;
    var accumulator = new Long(0, 0);
    var exponentAccumulator = new Long(0, 0);
    var isNegative = false;
    var isExponentPositive = false;
    var hasExponent = false;
    var start = current;
    $l$loop_4: while (!(current === charSequenceLength(this.d13()))) {
      var ch = charSequenceGet(this.d13(), current);
      if ((ch === _Char___init__impl__6a9atx(101) ? true : ch === _Char___init__impl__6a9atx(69)) ? !hasExponent : false) {
        if (current === start) {
          this.zx('Unexpected symbol ' + toString_0(ch) + ' in numeric literal');
        }
        isExponentPositive = true;
        hasExponent = true;
        current = current + 1 | 0;
        continue $l$loop_4;
      }
      if (ch === _Char___init__impl__6a9atx(45) ? hasExponent : false) {
        if (current === start) {
          this.zx("Unexpected symbol '-' in numeric literal");
        }
        isExponentPositive = false;
        current = current + 1 | 0;
        continue $l$loop_4;
      }
      if (ch === _Char___init__impl__6a9atx(43) ? hasExponent : false) {
        if (current === start) {
          this.zx("Unexpected symbol '+' in numeric literal");
        }
        isExponentPositive = true;
        current = current + 1 | 0;
        continue $l$loop_4;
      }
      if (ch === _Char___init__impl__6a9atx(45)) {
        if (!(current === start)) {
          this.zx("Unexpected symbol '-' in numeric literal");
        }
        isNegative = true;
        current = current + 1 | 0;
        continue $l$loop_4;
      }
      var token = charToTokenClass(ch);
      if (!(token === 0))
        break $l$loop_4;
      current = current + 1 | 0;
      var digit = Char__minus_impl_a2frrh(ch, _Char___init__impl__6a9atx(48));
      if (!(0 <= digit ? digit <= 9 : false)) {
        this.zx("Unexpected symbol '" + toString_0(ch) + "' in numeric literal");
      }
      if (hasExponent) {
        // Inline function 'kotlin.Long.plus' call
        // Inline function 'kotlin.Long.times' call
        exponentAccumulator = exponentAccumulator.w8(toLong(10)).b6(toLong(digit));
        continue $l$loop_4;
      }
      // Inline function 'kotlin.Long.minus' call
      // Inline function 'kotlin.Long.times' call
      accumulator = accumulator.w8(toLong(10)).ca(toLong(digit));
      if (accumulator.x8(new Long(0, 0)) > 0) {
        this.zx('Numeric value overflow');
      }
    }
    var hasChars = !(current === start);
    if (start === current ? true : isNegative ? start === (current - 1 | 0) : false) {
      this.zx('Expected numeric literal');
    }
    if (hasQuotation) {
      if (!hasChars) {
        this.zx('EOF');
      }
      if (!(charSequenceGet(this.d13(), current) === _Char___init__impl__6a9atx(34))) {
        this.zx('Expected closing quotation mark');
      }
      current = current + 1 | 0;
    }
    this.cw_1 = current;
    if (hasExponent) {
      var doubleAccumulator = accumulator.a6() * consumeNumericLiteral$calculateExponent(exponentAccumulator, isExponentPositive);
      var tmp_0;
      Companion_getInstance_3();
      if (doubleAccumulator > (new Long(-1, 2147483647)).a6()) {
        tmp_0 = true;
      } else {
        Companion_getInstance_3();
        tmp_0 = doubleAccumulator < (new Long(0, -2147483648)).a6();
      }
      if (tmp_0) {
        this.zx('Numeric value overflow');
      }
      // Inline function 'kotlin.math.floor' call
      if (!(Math.floor(doubleAccumulator) === doubleAccumulator)) {
        this.zx("Can't convert " + doubleAccumulator + ' to Long');
      }
      accumulator = numberToLong(doubleAccumulator);
    }
    var tmp_1;
    if (isNegative) {
      tmp_1 = accumulator;
    } else {
      var tmp_2 = accumulator;
      Companion_getInstance_3();
      if (!tmp_2.equals(new Long(0, -2147483648))) {
        tmp_1 = accumulator.ea();
      } else {
        this.zx('Numeric value overflow');
      }
    }
    return tmp_1;
  };
  protoOf(AbstractJsonLexer).s10 = function () {
    return consumeBoolean(this, this.l13());
  };
  protoOf(AbstractJsonLexer).t10 = function () {
    var current = this.l13();
    if (current === charSequenceLength(this.d13())) {
      this.zx('EOF');
    }
    var tmp;
    if (charSequenceGet(this.d13(), current) === _Char___init__impl__6a9atx(34)) {
      current = current + 1 | 0;
      tmp = true;
    } else {
      tmp = false;
    }
    var hasQuotation = tmp;
    var result = consumeBoolean(this, current);
    if (hasQuotation) {
      if (this.cw_1 === charSequenceLength(this.d13())) {
        this.zx('EOF');
      }
      if (!(charSequenceGet(this.d13(), this.cw_1) === _Char___init__impl__6a9atx(34))) {
        this.zx('Expected closing quotation mark');
      }
      this.cw_1 = this.cw_1 + 1 | 0;
    }
    return result;
  };
  function charToTokenClass(c) {
    var tmp;
    // Inline function 'kotlin.code' call
    if (Char__toInt_impl_vasixd(c) < 126) {
      var tmp_0 = CharMappings_getInstance().p13_1;
      // Inline function 'kotlin.code' call
      tmp = tmp_0[Char__toInt_impl_vasixd(c)];
    } else {
      tmp = 0;
    }
    return tmp;
  }
  function get_TC_WHITESPACE() {
    return TC_WHITESPACE;
  }
  var TC_WHITESPACE;
  function get_TC_EOF() {
    return TC_EOF;
  }
  var TC_EOF;
  function get_TC_STRING() {
    return TC_STRING;
  }
  var TC_STRING;
  function get_STRING_ESC() {
    return STRING_ESC;
  }
  var STRING_ESC;
  function get_TC_BEGIN_OBJ() {
    return TC_BEGIN_OBJ;
  }
  var TC_BEGIN_OBJ;
  function get_TC_COLON() {
    return TC_COLON;
  }
  var TC_COLON;
  function get_TC_COMMA() {
    return TC_COMMA;
  }
  var TC_COMMA;
  function get_COLON() {
    return COLON;
  }
  var COLON;
  function get_BEGIN_OBJ() {
    return BEGIN_OBJ;
  }
  var BEGIN_OBJ;
  function get_END_OBJ() {
    return END_OBJ;
  }
  var END_OBJ;
  function get_BEGIN_LIST() {
    return BEGIN_LIST;
  }
  var BEGIN_LIST;
  function get_END_LIST() {
    return END_LIST;
  }
  var END_LIST;
  function get_lenientHint() {
    return lenientHint;
  }
  var lenientHint;
  function get_NULL() {
    return NULL;
  }
  var NULL;
  function tokenDescription(token) {
    return token === 1 ? "quotation mark '\"'" : token === 2 ? "string escape sequence '\\'" : token === 4 ? "comma ','" : token === 5 ? "colon ':'" : token === 6 ? "start of the object '{'" : token === 7 ? "end of the object '}'" : token === 8 ? "start of the array '['" : token === 9 ? "end of the array ']'" : token === 10 ? 'end of the input' : token === 127 ? 'invalid token' : 'valid token';
  }
  function get_TC_OTHER() {
    return TC_OTHER;
  }
  var TC_OTHER;
  function escapeToChar(c) {
    return c < 117 ? CharMappings_getInstance().o13_1[c] : _Char___init__impl__6a9atx(0);
  }
  function get_TC_BEGIN_LIST() {
    return TC_BEGIN_LIST;
  }
  var TC_BEGIN_LIST;
  function get_TC_END_LIST() {
    return TC_END_LIST;
  }
  var TC_END_LIST;
  function get_TC_END_OBJ() {
    return TC_END_OBJ;
  }
  var TC_END_OBJ;
  function get_ignoreUnknownKeysHint() {
    return ignoreUnknownKeysHint;
  }
  var ignoreUnknownKeysHint;
  function initEscape($this) {
    var inductionVariable = 0;
    if (inductionVariable <= 31)
      do {
        var i = inductionVariable;
        inductionVariable = inductionVariable + 1 | 0;
        initC2ESC($this, i, _Char___init__impl__6a9atx(117));
      }
       while (inductionVariable <= 31);
    initC2ESC($this, 8, _Char___init__impl__6a9atx(98));
    initC2ESC($this, 9, _Char___init__impl__6a9atx(116));
    initC2ESC($this, 10, _Char___init__impl__6a9atx(110));
    initC2ESC($this, 12, _Char___init__impl__6a9atx(102));
    initC2ESC($this, 13, _Char___init__impl__6a9atx(114));
    initC2ESC_0($this, _Char___init__impl__6a9atx(47), _Char___init__impl__6a9atx(47));
    initC2ESC_0($this, _Char___init__impl__6a9atx(34), _Char___init__impl__6a9atx(34));
    initC2ESC_0($this, _Char___init__impl__6a9atx(92), _Char___init__impl__6a9atx(92));
  }
  function initCharToToken($this) {
    var inductionVariable = 0;
    if (inductionVariable <= 32)
      do {
        var i = inductionVariable;
        inductionVariable = inductionVariable + 1 | 0;
        initC2TC($this, i, 127);
      }
       while (inductionVariable <= 32);
    initC2TC($this, 9, 3);
    initC2TC($this, 10, 3);
    initC2TC($this, 13, 3);
    initC2TC($this, 32, 3);
    initC2TC_0($this, _Char___init__impl__6a9atx(44), 4);
    initC2TC_0($this, _Char___init__impl__6a9atx(58), 5);
    initC2TC_0($this, _Char___init__impl__6a9atx(123), 6);
    initC2TC_0($this, _Char___init__impl__6a9atx(125), 7);
    initC2TC_0($this, _Char___init__impl__6a9atx(91), 8);
    initC2TC_0($this, _Char___init__impl__6a9atx(93), 9);
    initC2TC_0($this, _Char___init__impl__6a9atx(34), 1);
    initC2TC_0($this, _Char___init__impl__6a9atx(92), 2);
  }
  function initC2ESC($this, c, esc) {
    if (!(esc === _Char___init__impl__6a9atx(117))) {
      // Inline function 'kotlin.code' call
      var tmp$ret$0 = Char__toInt_impl_vasixd(esc);
      $this.o13_1[tmp$ret$0] = numberToChar(c);
    }
  }
  function initC2ESC_0($this, c, esc) {
    // Inline function 'kotlin.code' call
    var tmp$ret$0 = Char__toInt_impl_vasixd(c);
    return initC2ESC($this, tmp$ret$0, esc);
  }
  function initC2TC($this, c, cl) {
    $this.p13_1[c] = cl;
  }
  function initC2TC_0($this, c, cl) {
    // Inline function 'kotlin.code' call
    var tmp$ret$0 = Char__toInt_impl_vasixd(c);
    return initC2TC($this, tmp$ret$0, cl);
  }
  function CharMappings() {
    CharMappings_instance = this;
    this.o13_1 = charArray(117);
    this.p13_1 = new Int8Array(126);
    initEscape(this);
    initCharToToken(this);
  }
  var CharMappings_instance;
  function CharMappings_getInstance() {
    if (CharMappings_instance == null)
      new CharMappings();
    return CharMappings_instance;
  }
  function get_specialFlowingValuesHint() {
    return specialFlowingValuesHint;
  }
  var specialFlowingValuesHint;
  function get_allowStructuredMapKeysHint() {
    return allowStructuredMapKeysHint;
  }
  var allowStructuredMapKeysHint;
  function StringJsonLexer(source) {
    AbstractJsonLexer.call(this);
    this.u13_1 = source;
  }
  protoOf(StringJsonLexer).d13 = function () {
    return this.u13_1;
  };
  protoOf(StringJsonLexer).e13 = function (position) {
    return position < this.u13_1.length ? position : -1;
  };
  protoOf(StringJsonLexer).wy = function () {
    var source = this.u13_1;
    $l$loop: while (!(this.cw_1 === -1) ? this.cw_1 < source.length : false) {
      var tmp1 = this.cw_1;
      this.cw_1 = tmp1 + 1 | 0;
      var ch = charSequenceGet(source, tmp1);
      var tc = charToTokenClass(ch);
      var tmp;
      if (tc === get_TC_WHITESPACE()) {
        continue $l$loop;
      } else {
        tmp = tc;
      }
      return tmp;
    }
    return get_TC_EOF();
  };
  protoOf(StringJsonLexer).f10 = function () {
    var current = this.l13();
    if (current === this.u13_1.length ? true : current === -1)
      return false;
    if (charSequenceGet(this.u13_1, current) === _Char___init__impl__6a9atx(44)) {
      this.cw_1 = this.cw_1 + 1 | 0;
      return true;
    }
    return false;
  };
  protoOf(StringJsonLexer).sy = function () {
    var current = this.cw_1;
    if (current === -1)
      return false;
    $l$loop: while (current < this.u13_1.length) {
      var c = charSequenceGet(this.u13_1, current);
      if (((c === _Char___init__impl__6a9atx(32) ? true : c === _Char___init__impl__6a9atx(10)) ? true : c === _Char___init__impl__6a9atx(13)) ? true : c === _Char___init__impl__6a9atx(9)) {
        current = current + 1 | 0;
        continue $l$loop;
      }
      this.cw_1 = current;
      return this.g13(c);
    }
    this.cw_1 = current;
    return false;
  };
  protoOf(StringJsonLexer).l13 = function () {
    var current = this.cw_1;
    if (current === -1)
      return current;
    $l$loop: while (current < this.u13_1.length) {
      var c = charSequenceGet(this.u13_1, current);
      if (((c === _Char___init__impl__6a9atx(32) ? true : c === _Char___init__impl__6a9atx(10)) ? true : c === _Char___init__impl__6a9atx(13)) ? true : c === _Char___init__impl__6a9atx(9)) {
        current = current + 1 | 0;
      } else {
        break $l$loop;
      }
    }
    this.cw_1 = current;
    return current;
  };
  protoOf(StringJsonLexer).e10 = function (expected) {
    if (this.cw_1 === -1) {
      this.i13(expected);
    }
    var source = this.u13_1;
    $l$loop: while (this.cw_1 < source.length) {
      var tmp1 = this.cw_1;
      this.cw_1 = tmp1 + 1 | 0;
      var c = charSequenceGet(source, tmp1);
      if (((c === _Char___init__impl__6a9atx(32) ? true : c === _Char___init__impl__6a9atx(10)) ? true : c === _Char___init__impl__6a9atx(13)) ? true : c === _Char___init__impl__6a9atx(9))
        continue $l$loop;
      if (c === expected)
        return Unit_instance;
      this.i13(expected);
    }
    this.cw_1 = -1;
    this.i13(expected);
  };
  protoOf(StringJsonLexer).k10 = function () {
    this.e10(get_STRING());
    var current = this.cw_1;
    var closingQuote = indexOf(this.u13_1, _Char___init__impl__6a9atx(34), current);
    if (closingQuote === -1) {
      this.uy();
      this.k13(get_TC_STRING(), false);
    }
    var inductionVariable = current;
    if (inductionVariable < closingQuote)
      do {
        var i = inductionVariable;
        inductionVariable = inductionVariable + 1 | 0;
        if (charSequenceGet(this.u13_1, i) === get_STRING_ESC()) {
          return this.consumeString2(this.u13_1, this.cw_1, i);
        }
      }
       while (inductionVariable < closingQuote);
    this.cw_1 = closingQuote + 1 | 0;
    // Inline function 'kotlin.text.substring' call
    // Inline function 'kotlin.js.asDynamic' call
    return this.u13_1.substring(current, closingQuote);
  };
  protoOf(StringJsonLexer).m10 = function (keyToMatch, isLenient) {
    var positionSnapshot = this.cw_1;
    try {
      if (!(this.wy() === get_TC_BEGIN_OBJ()))
        return null;
      var firstKey = this.h10(isLenient);
      if (!(firstKey === keyToMatch))
        return null;
      this.m13();
      if (!(this.wy() === get_TC_COLON()))
        return null;
      return this.h10(isLenient);
    }finally {
      this.cw_1 = positionSnapshot;
      this.m13();
    }
  };
  function get_schemaCache(_this__u8e3s4) {
    return _this__u8e3s4.rv_1;
  }
  function createMapForCache(initialCapacity) {
    return HashMap_init_$Create$(initialCapacity);
  }
  //region block: post-declaration
  protoOf(defer$1).ii = get_isNullable;
  protoOf(defer$1).qi = get_isInline;
  protoOf(defer$1).si = get_annotations;
  //endregion
  //region block: init
  Companion_instance_0 = new Companion();
  Companion_instance_1 = new Companion_0();
  Companion_instance_2 = new Companion_1();
  Companion_instance_3 = new Companion_2();
  Tombstone_instance = new Tombstone();
  PRIMITIVE_TAG = 'primitive';
  STRING = _Char___init__impl__6a9atx(34);
  TC_WHITESPACE = 3;
  TC_EOF = 10;
  TC_STRING = 1;
  STRING_ESC = _Char___init__impl__6a9atx(92);
  TC_BEGIN_OBJ = 6;
  TC_COLON = 5;
  TC_COMMA = 4;
  COLON = _Char___init__impl__6a9atx(58);
  BEGIN_OBJ = _Char___init__impl__6a9atx(123);
  END_OBJ = _Char___init__impl__6a9atx(125);
  BEGIN_LIST = _Char___init__impl__6a9atx(91);
  END_LIST = _Char___init__impl__6a9atx(93);
  lenientHint = "Use 'isLenient = true' in 'Json {}' builder to accept non-compliant JSON.";
  NULL = 'null';
  TC_OTHER = 0;
  TC_BEGIN_LIST = 8;
  TC_END_LIST = 9;
  TC_END_OBJ = 7;
  ignoreUnknownKeysHint = "Use 'ignoreUnknownKeys = true' in 'Json {}' builder to ignore unknown keys.";
  specialFlowingValuesHint = "It is possible to deserialize them using 'JsonBuilder.allowSpecialFloatingPointValues = true'";
  allowStructuredMapKeysHint = "Use 'allowStructuredMapKeys = true' in 'Json {}' builder to convert such maps to [key1, value1, key2, value2,...] arrays.";
  //endregion
  //region block: exports
  _.$_$ = _.$_$ || {};
  _.$_$.a = Default_getInstance;
  _.$_$.b = JsonArrayBuilder;
  _.$_$.c = JsonArray;
  _.$_$.d = JsonNull;
  _.$_$.e = JsonObjectBuilder;
  _.$_$.f = JsonObject;
  _.$_$.g = add;
  _.$_$.h = get_boolean;
  _.$_$.i = get_contentOrNull;
  _.$_$.j = get_int;
  _.$_$.k = get_jsonArray;
  _.$_$.l = get_jsonObject;
  _.$_$.m = get_jsonPrimitive;
  _.$_$.n = put_1;
  _.$_$.o = put_0;
  _.$_$.p = put;
  //endregion
  return _;
}));
