// @sd-jwt/crypto-nodejs/dist/index.mjs
var __async = (__this, __arguments, generator) => {
  return new Promise((resolve, reject) => {
    var fulfilled = (value) => {
      try {
        step(generator.next(value));
      } catch (e) {
        reject(e);
      }
    };
    var rejected = (value) => {
      try {
        step(generator.throw(value));
      } catch (e) {
        reject(e);
      }
    };
    var step = (x) => x.done ? resolve(x.value) : Promise.resolve(x.value).then(fulfilled, rejected);
    step((generator = generator.apply(__this, __arguments)).next());
  });
};

// src/crypto.ts
import { createHash, randomBytes, subtle } from "crypto";
var generateSalt = (length) => {
  if (length <= 0) {
    return "";
  }
  const saltBytes = randomBytes(length);
  const salt = saltBytes.toString("hex");
  return salt.substring(0, length);
};
var digest = (data, algorithm = "SHA-256") => {
  const nodeAlg = toNodeCryptoAlg(algorithm);
  const hash = createHash(nodeAlg);
  hash.update(data);
  const hashBuffer = hash.digest();
  return new Uint8Array(hashBuffer);
};
var toNodeCryptoAlg = (hashAlg) => hashAlg.replace("-", "").toLowerCase();
var ES256 = {
  alg: "ES256",
  generateKeyPair() {
    return __async(this, null, function* () {
      const keyPair = yield subtle.generateKey(
        {
          name: "ECDSA",
          namedCurve: "P-256"
          // ES256
        },
        true,
        // whether the key is extractable (i.e., can be used in exportKey)
        ["sign", "verify"]
        // can be used to sign and verify signatures
      );
      const publicKeyJWK = yield subtle.exportKey("jwk", keyPair.publicKey);
      const privateKeyJWK = yield subtle.exportKey("jwk", keyPair.privateKey);
      return { publicKey: publicKeyJWK, privateKey: privateKeyJWK };
    });
  },
  getSigner(privateKeyJWK) {
    return __async(this, null, function* () {
      const privateKey = yield subtle.importKey(
        "jwk",
        privateKeyJWK,
        {
          name: "ECDSA",
          namedCurve: "P-256"
          // Must match the curve used to generate the key
        },
        true,
        // whether the key is extractable (i.e., can be used in exportKey)
        ["sign"]
      );
      return (data) => __async(this, null, function* () {
        const encoder = new TextEncoder();
        const signature = yield subtle.sign(
          {
            name: "ECDSA",
            hash: { name: "SHA-256" }
            // Required for ES256
          },
          privateKey,
          encoder.encode(data)
        );
        return btoa(String.fromCharCode(...new Uint8Array(signature))).replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/, "");
      });
    });
  },
  getVerifier(publicKeyJWK) {
    return __async(this, null, function* () {
      const publicKey = yield subtle.importKey(
        "jwk",
        publicKeyJWK,
        {
          name: "ECDSA",
          namedCurve: "P-256"
          // Must match the curve used to generate the key
        },
        true,
        // whether the key is extractable (i.e., can be used in exportKey)
        ["verify"]
      );
      return (data, signatureBase64url) => __async(this, null, function* () {
        const encoder = new TextEncoder();
        const signature = Uint8Array.from(
          atob(signatureBase64url.replace(/-/g, "+").replace(/_/g, "/")),
          (c) => c.charCodeAt(0)
        );
        const isValid = yield subtle.verify(
          {
            name: "ECDSA",
            hash: { name: "SHA-256" }
            // Required for ES256
          },
          publicKey,
          signature,
          encoder.encode(data)
        );
        return isValid;
      });
    });
  }
};
export {
  ES256,
  digest,
  generateSalt
};
